{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   6   \n",
      " ┏━┻━┓ \n",
      " 5   ┃ \n",
      "┏┻┓  ┃ \n",
      "┃ ┃  4 \n",
      "┃ ┃ ┏┻┓\n",
      "0 1 2 3\n",
      "\n",
      "   6   \n",
      " ┏━┻━┓ \n",
      " ┃   5 \n",
      " ┃  ┏┻┓\n",
      " 4  ┃ ┃\n",
      "┏┻┓ ┃ ┃\n",
      "0 1 2 3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#conda install ipykernel -y put this in the terminal to fix the module missing error \n",
    "\n",
    "#be able to draw a tree sequence just via editing the tables\n",
    "import tskit\n",
    "tables = tskit.TableCollection(sequence_length=1.0)\n",
    "n = tables.nodes\n",
    "#nodes\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0.4)\n",
    "n.add_row(flags=1, population=-1, time=0.7)\n",
    "n.add_row(flags=1, population=-1, time=1)\n",
    "\n",
    "# to make a straightforward tree: \n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=4, child=2) \n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=4, child=3)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=5, child=0)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=5, child=1)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=6, child=4)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=6, child=5)\n",
    "\n",
    "#adding in something where the edges actually matter: 1/2 of the genome has one genealogy, 1/2 has another. It doesn't make a lot of sense re: biology but the concept of different chunks having different ancestry I figured out here \n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=4, child=2)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=4, child=3)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=5, child=0)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=5, child=1)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=6, child=4)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=6, child=5)\n",
    "\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=4, child=0) \n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=4, child=1)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=5, child=2)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=5, child=3)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=6, child=4)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=6, child=5)\n",
    "\n",
    "#sort step, to ensure that your tables are ordered correctly\n",
    "tskit.TableCollection.sort(edge_start=0, self=tables, site_start=0, mutation_start=0) #in tskit docs, the self=tables portion is just *, so here I've assigned self the the variable name for all tables\n",
    "\n",
    "#tree sequence that prints in terminal and you don't have to deal w/ vscode and svgs\n",
    "ts = tables.tree_sequence()\n",
    "for t in ts.trees():\n",
    "  print(t.draw(format='unicode'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of samples for the runs 10\n",
      "expected Tmrca: 1.8\n",
      "mean tmrca, in generations/(ploidy*N): 1.7671071925460928 mean tmrca, in raw gens: 1767.1071925460928\n",
      "total expected branch lengths 5435.714285714284 mean summed branch lengths: 5582.498052890885\n",
      "mean expectation for the number of polymorphisms on the tree 5.435714285714284\n",
      "mean allele frequencies, from all sims binned {1: 1.79, 2: 0.99, 3: 0.69, 4: 0.56, 5: 0.21, 6: 0.36, 7: 0.2, 8: 0.32, 9: 0.18, 10: 0.0, 11: 0.0, 12: 0.0, 13: 0.0, 14: 0.0, 15: 0.0, 16: 0.0, 17: 0.0, 18: 0.0}\n",
      "expected mean site frequency spectrum [2.0, 1.0, 0.6666666666666666, 0.5, 0.4, 0.3333333333333333, 0.2857142857142857, 0.25, 0.2222222222222222] mean allele frequencies from sims, averaged {1: 1.79, 2: 0.99, 3: 0.69, 4: 0.56, 5: 0.21, 6: 0.36, 7: 0.2, 8: 0.32, 9: 0.18, 10: 0.0, 11: 0.0, 12: 0.0, 13: 0.0, 14: 0.0, 15: 0.0, 16: 0.0, 17: 0.0, 18: 0.0}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 18 artists>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAGdCAYAAAAxCSikAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAn/UlEQVR4nO3df1DUd37H8deCYVEnrBqUhYSI5oeenoAx5xabTM7LRqCOJ/1h1ElPpWpaq514e0mUToSk3hTzo1ZzRyWXSNCm549MEjOtFmO4oDVBrSKTmCaOWvwVWfzRwAqekMK3f2TcdMMCLvLjw/J8zHwn7uf7/n6+n08+7vLyy3d3bZZlWQIAADBYRG8PAAAAoCMEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8Qb09gC6QktLiy5cuKDbb79dNputt4cDAABugmVZunr1qhISEhQR0f41lLAILBcuXFBiYmJvDwMAAHTCuXPndNddd7VbExaB5fbbb5f07YRjYmJ6eTQAAOBm+Hw+JSYm+n+OtycsAsuNXwPFxMQQWAAA6GNu5nYObroFAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMN6A3h5AX5C0cmeX9nd6zfQu7Q8AgHDHFRYAAGA8AgsAADAegQUAABgv5MCyb98+zZgxQwkJCbLZbNqxY0e79QsWLJDNZmu1jR8/3l/z/PPPt9o/duzYkCcDAADCU8iBpaGhQSkpKSooKLip+vXr16u6utq/nTt3TsOGDdOsWbMC6saPHx9Qt3///lCHBgAAwlTI7xLKzMxUZmbmTdc7HA45HA7/4x07dujrr79WdnZ24EAGDJDT6Qx1OAAAoB/o8XtYNm7cKLfbrZEjRwa0nzhxQgkJCRo9erSeeOIJnT17ts0+Ghsb5fP5AjYAABC+ejSwXLhwQf/+7/+uRYsWBbS7XC4VFxerpKREGzZsUFVVlR5++GFdvXo1aD/5+fn+KzcOh0OJiYk9MXwAANBLejSwbNq0SUOGDFFWVlZAe2ZmpmbNmqXk5GSlp6dr165dqq2t1fbt24P2k5OTo7q6Ov927ty5Hhg9AADoLT32SbeWZamoqEg/+9nPFBUV1W7tkCFDdP/99+vkyZNB99vtdtnt9u4YJgAAMFCPXWHZu3evTp48qYULF3ZYW19fr1OnTik+Pr4HRgYAAEwXcmCpr69XZWWlKisrJUlVVVWqrKz03ySbk5OjefPmtTpu48aNcrlc+uEPf9hq39NPP629e/fq9OnT+uSTT/THf/zHioyM1Ny5c0MdHgAACEMh/0ro8OHDmjp1qv+xx+ORJM2fP1/FxcWqrq5u9Q6furo6vfPOO1q/fn3QPs+fP6+5c+fqypUrGj58uB566CEdOHBAw4cPD3V4AAAgDNksy7J6exC3yufzyeFwqK6uTjExMV3eP9/WDABA1wvl5zffJQQAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgvJADy759+zRjxgwlJCTIZrNpx44d7daXlZXJZrO12rxeb0BdQUGBkpKSFB0dLZfLpUOHDoU6NAAAEKZCDiwNDQ1KSUlRQUFBSMcdP35c1dXV/m3EiBH+fdu2bZPH41FeXp4qKiqUkpKi9PR0Xbx4MdThAQCAMDQg1AMyMzOVmZkZ8olGjBihIUOGBN23du1aLV68WNnZ2ZKkwsJC7dy5U0VFRVq5cmXI5wIAAOGlx+5hSU1NVXx8vB577DF9/PHH/vampiYdOXJEbrf7u0FFRMjtdqu8vDxoX42NjfL5fAEbAAAIX90eWOLj41VYWKh33nlH77zzjhITE/XjH/9YFRUVkqTLly+rublZcXFxAcfFxcW1us/lhvz8fDkcDv+WmJjY3dMAAAC9KORfCYVqzJgxGjNmjP/xlClTdOrUKf3jP/6j/vmf/7lTfebk5Mjj8fgf+3w+QgsAAGGs2wNLMJMnT9b+/fslSbGxsYqMjFRNTU1ATU1NjZxOZ9Dj7Xa77HZ7t48TAACYoVc+h6WyslLx8fGSpKioKE2aNEmlpaX+/S0tLSotLVVaWlpvDA8AABgm5Css9fX1OnnypP9xVVWVKisrNWzYMN19993KycnRV199pc2bN0uS1q1bp1GjRmn8+PG6fv263njjDf3ud7/TBx984O/D4/Fo/vz5evDBBzV58mStW7dODQ0N/ncNAQCA/i3kwHL48GFNnTrV//jGvSTz589XcXGxqqurdfbsWf/+pqYm/eIXv9BXX32lQYMGKTk5WR9++GFAH7Nnz9alS5eUm5srr9er1NRUlZSUtLoRFwAA9E82y7Ks3h7ErfL5fHI4HKqrq1NMTEyX95+0cmeX9nd6zfQu7Q8AgL4olJ/ffJcQAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPFCDiz79u3TjBkzlJCQIJvNph07drRb/+677+qxxx7T8OHDFRMTo7S0NO3evTug5vnnn5fNZgvYxo4dG+rQAABAmAo5sDQ0NCglJUUFBQU3Vb9v3z499thj2rVrl44cOaKpU6dqxowZOnr0aEDd+PHjVV1d7d/2798f6tAAAECYGhDqAZmZmcrMzLzp+nXr1gU8/vu//3u9//77+td//VdNnDjxu4EMGCCn0xnqcAAAQD/Q4/ewtLS06OrVqxo2bFhA+4kTJ5SQkKDRo0friSee0NmzZ9vso7GxUT6fL2ADAADhq8cDyyuvvKL6+no9/vjj/jaXy6Xi4mKVlJRow4YNqqqq0sMPP6yrV68G7SM/P18Oh8O/JSYm9tTwAQBAL+jRwPLb3/5WL7zwgrZv364RI0b42zMzMzVr1iwlJycrPT1du3btUm1trbZv3x60n5ycHNXV1fm3c+fO9dQUAABALwj5HpbO2rp1qxYtWqS3335bbre73dohQ4bo/vvv18mTJ4Put9vtstvt3TFMAABgoB65wrJlyxZlZ2dry5Ytmj59eof19fX1OnXqlOLj43tgdAAAwHQhX2Gpr68PuPJRVVWlyspKDRs2THfffbdycnL01VdfafPmzZK+/TXQ/PnztX79erlcLnm9XknSwIED5XA4JElPP/20ZsyYoZEjR+rChQvKy8tTZGSk5s6d2xVzBAAAfVzIV1gOHz6siRMn+t+S7PF4NHHiROXm5kqSqqurA97h85vf/Eb/+7//q6VLlyo+Pt6/PfXUU/6a8+fPa+7cuRozZowef/xx3XHHHTpw4ICGDx9+q/MDAABhwGZZltXbg7hVPp9PDodDdXV1iomJ6fL+k1bu7NL+Tq/p+NdiAACEu1B+fvNdQgAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGCzmw7Nu3TzNmzFBCQoJsNpt27NjR4TFlZWV64IEHZLfbde+996q4uLhVTUFBgZKSkhQdHS2Xy6VDhw6FOjQAABCmQg4sDQ0NSklJUUFBwU3VV1VVafr06Zo6daoqKyu1fPlyLVq0SLt37/bXbNu2TR6PR3l5eaqoqFBKSorS09N18eLFUIcHAADCkM2yLKvTB9tseu+995SVldVmzYoVK7Rz504dO3bM3zZnzhzV1taqpKREkuRyufSjH/1Iv/71ryVJLS0tSkxM1N/8zd9o5cqVHY7D5/PJ4XCorq5OMTExnZ1Om5JW7uzS/k6vmd6l/QEA0BeF8vO72+9hKS8vl9vtDmhLT09XeXm5JKmpqUlHjhwJqImIiJDb7fbXAACA/m1Ad5/A6/UqLi4uoC0uLk4+n0+///3v9fXXX6u5uTlozZdffhm0z8bGRjU2Nvof+3y+rh84AAAwRp98l1B+fr4cDod/S0xM7O0hAQCAbtTtgcXpdKqmpiagraamRjExMRo4cKBiY2MVGRkZtMbpdAbtMycnR3V1df7t3Llz3TZ+AADQ+7r9V0JpaWnatWtXQNuePXuUlpYmSYqKitKkSZNUWlrqv3m3paVFpaWlWrZsWdA+7Xa77HZ7t467p3X1jb0SN/cCAMJHyFdY6uvrVVlZqcrKSknfvm25srJSZ8+elfTt1Y958+b56//qr/5K//3f/61nn31WX375pf7pn/5J27dv189//nN/jcfj0euvv65Nmzbpiy++0JIlS9TQ0KDs7OxbnB4AAAgHIV9hOXz4sKZOnep/7PF4JEnz589XcXGxqqur/eFFkkaNGqWdO3fq5z//udavX6+77rpLb7zxhtLT0/01s2fP1qVLl5Sbmyuv16vU1FSVlJS0uhEXAAD0T7f0OSymCIfPYeFXQgCA/saoz2EBAAC4VQQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAON1KrAUFBQoKSlJ0dHRcrlcOnToUJu1P/7xj2Wz2Vpt06dP99csWLCg1f6MjIzODA0AAIShAaEesG3bNnk8HhUWFsrlcmndunVKT0/X8ePHNWLEiFb17777rpqamvyPr1y5opSUFM2aNSugLiMjQ2+++ab/sd1uD3VoAAAgTIV8hWXt2rVavHixsrOzNW7cOBUWFmrQoEEqKioKWj9s2DA5nU7/tmfPHg0aNKhVYLHb7QF1Q4cO7dyMAABA2AkpsDQ1NenIkSNyu93fdRARIbfbrfLy8pvqY+PGjZozZ44GDx4c0F5WVqYRI0ZozJgxWrJkia5cudJmH42NjfL5fAEbAAAIXyEFlsuXL6u5uVlxcXEB7XFxcfJ6vR0ef+jQIR07dkyLFi0KaM/IyNDmzZtVWlqqF198UXv37lVmZqaam5uD9pOfny+Hw+HfEhMTQ5kGAADoY0K+h+VWbNy4URMmTNDkyZMD2ufMmeP/84QJE5ScnKx77rlHZWVlevTRR1v1k5OTI4/H43/s8/kILQAAhLGQrrDExsYqMjJSNTU1Ae01NTVyOp3tHtvQ0KCtW7dq4cKFHZ5n9OjRio2N1cmTJ4Put9vtiomJCdgAAED4CimwREVFadKkSSotLfW3tbS0qLS0VGlpae0e+/bbb6uxsVF//ud/3uF5zp8/rytXrig+Pj6U4QEAgDAV8ruEPB6PXn/9dW3atElffPGFlixZooaGBmVnZ0uS5s2bp5ycnFbHbdy4UVlZWbrjjjsC2uvr6/XMM8/owIEDOn36tEpLSzVz5kzde++9Sk9P7+S0AABAOAn5HpbZs2fr0qVLys3NldfrVWpqqkpKSvw34p49e1YREYE56Pjx49q/f78++OCDVv1FRkbq008/1aZNm1RbW6uEhARNmzZNq1ev5rNYAACApE7edLts2TItW7Ys6L6ysrJWbWPGjJFlWUHrBw4cqN27d3dmGAAAoJ/gu4QAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABivRz+aH70vaeXOLu3v9JrpXdofAADBcIUFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAw3oDeHgDCT9LKnV3a3+k107u0PwBA38MVFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8ToVWAoKCpSUlKTo6Gi5XC4dOnSozdri4mLZbLaALTo6OqDGsizl5uYqPj5eAwcOlNvt1okTJzozNAAAEIZCDizbtm2Tx+NRXl6eKioqlJKSovT0dF28eLHNY2JiYlRdXe3fzpw5E7D/pZde0quvvqrCwkIdPHhQgwcPVnp6uq5fvx76jAAAQNgJObCsXbtWixcvVnZ2tsaNG6fCwkINGjRIRUVFbR5js9nkdDr9W1xcnH+fZVlat26dnnvuOc2cOVPJycnavHmzLly4oB07dnRqUgAAILyEFFiampp05MgRud3u7zqIiJDb7VZ5eXmbx9XX12vkyJFKTEzUzJkz9fnnn/v3VVVVyev1BvTpcDjkcrna7LOxsVE+ny9gAwAA4SukwHL58mU1NzcHXCGRpLi4OHm93qDHjBkzRkVFRXr//ff11ltvqaWlRVOmTNH58+clyX9cKH3m5+fL4XD4t8TExFCmAQAA+phuf5dQWlqa5s2bp9TUVD3yyCN69913NXz4cL322mud7jMnJ0d1dXX+7dy5c104YgAAYJqQAktsbKwiIyNVU1MT0F5TUyOn03lTfdx2222aOHGiTp48KUn+40Lp0263KyYmJmADAADhK6TAEhUVpUmTJqm0tNTf1tLSotLSUqWlpd1UH83Nzfrss88UHx8vSRo1apScTmdAnz6fTwcPHrzpPgEAQHgbEOoBHo9H8+fP14MPPqjJkydr3bp1amhoUHZ2tiRp3rx5uvPOO5Wfny9J+ru/+zv9wR/8ge69917V1tbq5Zdf1pkzZ7Ro0SJJ376DaPny5frlL3+p++67T6NGjdKqVauUkJCgrKysrpspAADos0IOLLNnz9alS5eUm5srr9er1NRUlZSU+G+aPXv2rCIivrtw8/XXX2vx4sXyer0aOnSoJk2apE8++UTjxo3z1zz77LNqaGjQk08+qdraWj300EMqKSlp9QFzAACgf7JZlmX19iBulc/nk8PhUF1dXbfcz5K0cmeX9nd6zfRuP0dPnae3zgEA6PtC+fnNdwkBAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgvAG9PQDAZEkrd3Zpf6fXTO/S/gCgv+AKCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeHwOC9APdPXnyUh8pgyAnsUVFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgvE4FloKCAiUlJSk6Oloul0uHDh1qs/b111/Xww8/rKFDh2ro0KFyu92t6hcsWCCbzRawZWRkdGZoAAAgDIUcWLZt2yaPx6O8vDxVVFQoJSVF6enpunjxYtD6srIyzZ07Vx999JHKy8uVmJioadOm6auvvgqoy8jIUHV1tX/bsmVL52YEAADCTsiBZe3atVq8eLGys7M1btw4FRYWatCgQSoqKgpa/y//8i/667/+a6Wmpmrs2LF644031NLSotLS0oA6u90up9Pp34YOHdq5GQEAgLATUmBpamrSkSNH5Ha7v+sgIkJut1vl5eU31ce1a9f0zTffaNiwYQHtZWVlGjFihMaMGaMlS5boypUrbfbR2Ngon88XsAEAgPAVUmC5fPmympubFRcXF9AeFxcnr9d7U32sWLFCCQkJAaEnIyNDmzdvVmlpqV588UXt3btXmZmZam5uDtpHfn6+HA6Hf0tMTAxlGgAAoI8Z0JMnW7NmjbZu3aqysjJFR0f72+fMmeP/84QJE5ScnKx77rlHZWVlevTRR1v1k5OTI4/H43/s8/kILQAAhLGQrrDExsYqMjJSNTU1Ae01NTVyOp3tHvvKK69ozZo1+uCDD5ScnNxu7ejRoxUbG6uTJ08G3W+32xUTExOwAQCA8BVSYImKitKkSZMCbpi9cQNtWlpam8e99NJLWr16tUpKSvTggw92eJ7z58/rypUrio+PD2V4AAAgTIX8LiGPx6PXX39dmzZt0hdffKElS5aooaFB2dnZkqR58+YpJyfHX//iiy9q1apVKioqUlJSkrxer7xer+rr6yVJ9fX1euaZZ3TgwAGdPn1apaWlmjlzpu69916lp6d30TQBAEBfFvI9LLNnz9alS5eUm5srr9er1NRUlZSU+G/EPXv2rCIivstBGzZsUFNTk/7sz/4soJ+8vDw9//zzioyM1KeffqpNmzaptrZWCQkJmjZtmlavXi273X6L0wMAAOGgUzfdLlu2TMuWLQu6r6ysLODx6dOn2+1r4MCB2r17d2eGAQAA+gm+SwgAABivR9/WDHSVpJU7u7zP02umd3mfAICuwRUWAABgPK6wAL2sq68WcaUIQDjiCgsAADAegQUAABiPwAIAAIzHPSwAugz34wDoLlxhAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwXqcCS0FBgZKSkhQdHS2Xy6VDhw61W//2229r7Nixio6O1oQJE7Rr166A/ZZlKTc3V/Hx8Ro4cKDcbrdOnDjRmaEBAIAwFHJg2bZtmzwej/Ly8lRRUaGUlBSlp6fr4sWLQes/+eQTzZ07VwsXLtTRo0eVlZWlrKwsHTt2zF/z0ksv6dVXX1VhYaEOHjyowYMHKz09XdevX+/8zAAAQNgIObCsXbtWixcvVnZ2tsaNG6fCwkINGjRIRUVFQevXr1+vjIwMPfPMM/rBD36g1atX64EHHtCvf/1rSd9eXVm3bp2ee+45zZw5U8nJydq8ebMuXLigHTt23NLkAABAeBgQSnFTU5OOHDminJwcf1tERITcbrfKy8uDHlNeXi6PxxPQlp6e7g8jVVVV8nq9crvd/v0Oh0Mul0vl5eWaM2dOqz4bGxvV2Njof1xXVydJ8vl8oUznprU0XuvS/oKNs6vP0VPnCZdz9NR5wuUcPXWe7npOAzDDjee4ZVkd1oYUWC5fvqzm5mbFxcUFtMfFxenLL78MeozX6w1a7/V6/ftvtLVV8335+fl64YUXWrUnJibe3ER6mWNd+JwnXM7RU+cJl3P01Hl6ai4AetfVq1flcDjarQkpsJgiJycn4KpNS0uL/ud//kd33HGHbDZbr4zJ5/MpMTFR586dU0xMTK+Mobf057lL/Xv+zL1/zl3q3/Nn7l03d8uydPXqVSUkJHRYG1JgiY2NVWRkpGpqagLaa2pq5HQ6gx7jdDrbrb/x35qaGsXHxwfUpKamBu3TbrfLbrcHtA0ZMiSUqXSbmJiYfvcX+Ib+PHepf8+fuffPuUv9e/7MvWvm3tGVlRtCuuk2KipKkyZNUmlpqb+tpaVFpaWlSktLC3pMWlpaQL0k7dmzx18/atQoOZ3OgBqfz6eDBw+22ScAAOhfQv6VkMfj0fz58/Xggw9q8uTJWrdunRoaGpSdnS1Jmjdvnu68807l5+dLkp566ik98sgj+od/+AdNnz5dW7du1eHDh/Wb3/xGkmSz2bR8+XL98pe/1H333adRo0Zp1apVSkhIUFZWVtfNFAAA9FkhB5bZs2fr0qVLys3NldfrVWpqqkpKSvw3zZ49e1YREd9duJkyZYp++9vf6rnnntPf/u3f6r777tOOHTv0wx/+0F/z7LPPqqGhQU8++aRqa2v10EMPqaSkRNHR0V0wxZ5ht9uVl5fX6ldV/UF/nrvUv+fP3Pvn3KX+PX/m3jtzt1k3814iAACAXsR3CQEAAOMRWAAAgPEILAAAwHgEFgAAYDwCSwgKCgqUlJSk6OhouVwuHTp0qN36t99+W2PHjlV0dLQmTJigXbt29dBIu1Z+fr5+9KMf6fbbb9eIESOUlZWl48ePt3tMcXGxbDZbwNaX3vV1w/PPP99qHmPHjm33mHBZ96SkpFZzt9lsWrp0adD6vr7m+/bt04wZM5SQkCCbzdbqy1cty1Jubq7i4+M1cOBAud1unThxosN+Q33d6A3tzf2bb77RihUrNGHCBA0ePFgJCQmaN2+eLly40G6fnXnu9IaO1n3BggWt5pGRkdFhv31h3aWO5x/sNcBms+nll19us8/uWnsCy03atm2bPB6P8vLyVFFRoZSUFKWnp+vixYtB6z/55BPNnTtXCxcu1NGjR5WVlaWsrCwdO3ash0d+6/bu3aulS5fqwIED2rNnj7755htNmzZNDQ0N7R4XExOj6upq/3bmzJkeGnHXGj9+fMA89u/f32ZtOK37f/7nfwbMe8+ePZKkWbNmtXlMX17zhoYGpaSkqKCgIOj+l156Sa+++qoKCwt18OBBDR48WOnp6bp+/XqbfYb6utFb2pv7tWvXVFFRoVWrVqmiokLvvvuujh8/rp/+9Kcd9hvKc6e3dLTukpSRkREwjy1btrTbZ19Zd6nj+f//eVdXV6uoqEg2m01/+qd/2m6/3bL2Fm7K5MmTraVLl/ofNzc3WwkJCVZ+fn7Q+scff9yaPn16QJvL5bL+8i//slvH2RMuXrxoSbL27t3bZs2bb75pORyOnhtUN8nLy7NSUlJuuj6c1/2pp56y7rnnHqulpSXo/nBZc8uyLEnWe++953/c0tJiOZ1O6+WXX/a31dbWWna73dqyZUub/YT6umGC7889mEOHDlmSrDNnzrRZE+pzxwTB5j5//nxr5syZIfXTF9fdsm5u7WfOnGn95Cc/abemu9aeKyw3oampSUeOHJHb7fa3RUREyO12q7y8POgx5eXlAfWSlJ6e3mZ9X1JXVydJGjZsWLt19fX1GjlypBITEzVz5kx9/vnnPTG8LnfixAklJCRo9OjReuKJJ3T27Nk2a8N13ZuamvTWW2/pL/7iL9r9gtFwWfPvq6qqktfrDVhbh8Mhl8vV5tp25nWjr6irq5PNZuvwO9xCee6YrKysTCNGjNCYMWO0ZMkSXblypc3acF73mpoa7dy5UwsXLuywtjvWnsByEy5fvqzm5mb/p/neEBcXJ6/XG/QYr9cbUn1f0dLSouXLl+sP//APAz6t+PvGjBmjoqIivf/++3rrrbfU0tKiKVOm6Pz58z042lvncrlUXFyskpISbdiwQVVVVXr44Yd19erVoPXhuu47duxQbW2tFixY0GZNuKx5MDfWL5S17czrRl9w/fp1rVixQnPnzm33y+9Cfe6YKiMjQ5s3b1ZpaalefPFF7d27V5mZmWpubg5aH67rLkmbNm3S7bffrj/5kz9pt6671j7kj+ZH/7Z06VIdO3asw99HpqWlBXx55ZQpU/SDH/xAr732mlavXt3dw+wymZmZ/j8nJyfL5XJp5MiR2r59+039KyNcbNy4UZmZme1+BXy4rDna9s033+jxxx+XZVnasGFDu7Xh8tyZM2eO/88TJkxQcnKy7rnnHpWVlenRRx/txZH1vKKiIj3xxBMd3kzfXWvPFZabEBsbq8jISNXU1AS019TUyOl0Bj3G6XSGVN8XLFu2TP/2b/+mjz76SHfddVdIx952222aOHGiTp482U2j6xlDhgzR/fff3+Y8wnHdz5w5ow8//FCLFi0K6bhwWXNJ/vULZW0787phshth5cyZM9qzZ0+7V1eC6ei501eMHj1asbGxbc4j3Nb9hv/4j//Q8ePHQ34dkLpu7QksNyEqKkqTJk1SaWmpv62lpUWlpaUB/6L8/9LS0gLqJWnPnj1t1pvMsiwtW7ZM7733nn73u99p1KhRIffR3Nyszz77TPHx8d0wwp5TX1+vU6dOtTmPcFr3G958802NGDFC06dPD+m4cFlzSRo1apScTmfA2vp8Ph08eLDNte3M64apboSVEydO6MMPP9Qdd9wRch8dPXf6ivPnz+vKlSttziOc1v3/27hxoyZNmqSUlJSQj+2yte/y23jD1NatWy273W4VFxdb//Vf/2U9+eST1pAhQyyv12tZlmX97Gc/s1auXOmv//jjj60BAwZYr7zyivXFF19YeXl51m233WZ99tlnvTWFTluyZInlcDissrIyq7q62r9du3bNX/P9+b/wwgvW7t27rVOnTllHjhyx5syZY0VHR1uff/55b0yh037xi19YZWVlVlVVlfXxxx9bbrfbio2NtS5evGhZVnivu2V9++6Gu+++21qxYkWrfeG25levXrWOHj1qHT161JJkrV271jp69Kj/nTBr1qyxhgwZYr3//vvWp59+as2cOdMaNWqU9fvf/97fx09+8hPrV7/6lf9xR68bpmhv7k1NTdZPf/pT66677rIqKysDXgMaGxv9fXx/7h09d0zR3tyvXr1qPf3001Z5eblVVVVlffjhh9YDDzxg3Xfffdb169f9ffTVdbesjv/eW5Zl1dXVWYMGDbI2bNgQtI+eWnsCSwh+9atfWXfffbcVFRVlTZ482Tpw4IB/3yOPPGLNnz8/oH779u3W/fffb0VFRVnjx4+3du7c2cMj7hqSgm5vvvmmv+b781++fLn//1VcXJz1R3/0R1ZFRUXPD/4WzZ4924qPj7eioqKsO++805o9e7Z18uRJ//5wXnfLsqzdu3dbkqzjx4+32hdua/7RRx8F/Xt+Y44tLS3WqlWrrLi4OMtut1uPPvpoq/8vI0eOtPLy8gLa2nvdMEV7c6+qqmrzNeCjjz7y9/H9uXf03DFFe3O/du2aNW3aNGv48OHWbbfdZo0cOdJavHhxq+DRV9fdsjr+e29ZlvXaa69ZAwcOtGpra4P20VNrb7Msy7q1azQAAADdi3tYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADDe/wHUhm1lvJjPmAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#make a no-recombination sim to get the stats right \n",
    "import msprime\n",
    "import tskit \n",
    "import numpy as np\n",
    "from numpy.random import default_rng\n",
    "rng = default_rng()\n",
    "import random\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sequence_length = 100\n",
    "mu = (1e-3) #normally would have this set to 1e-7 *2, but putting deme size down and #individuals up to be able to visualize tree sequences for working on placing mutations issue.\n",
    "# sequence_length = 100\n",
    "ploidy = 2\n",
    "population_sequences = 1000 #total number of genomes in the population- the number of individuals is this/ploidy\n",
    "sampled_sequences = 10\n",
    "sim_runs = 100\n",
    "#initialize the bottom two variables, to add/append to later \n",
    "mean_tmrca = 0\n",
    "i_values = []\n",
    "mean_summed_branch_lengths = []\n",
    "for ts in msprime.sim_ancestry(ploidy= ploidy, samples=(sampled_sequences/ploidy), discrete_genome = True, population_size=(population_sequences)/ploidy, num_replicates = sim_runs, sequence_length = sequence_length): #the sim needs everything in terms of the number of individuals, but it's convenient for downstream/interpretation to have our original variables be in terms of the number of lineages or individual sequences, aka in haploid terms \n",
    "    assert ts.num_samples == samples\n",
    "    last_node = ts.num_nodes - 1\n",
    "    tmrca = ts.node(last_node).time\n",
    "    mean_tmrca += tmrca\n",
    "    # print(\"tmrca for this run:\", tmrca)\n",
    "    newtables = ts.dump_tables() #modifying a copy of the original tables in order to be able to overwrite \n",
    "    #make a new table collection to hold everything\n",
    "    for tree in ts.trees():\n",
    "        # print(tree.draw(format='unicode'))\n",
    "        interval = tree.interval #find the interval the tree occupies in the genome \n",
    "        # print(newtables.edges.child)\n",
    "        nodes_and_branches = {} #empty dict to hold the pair of node:branch length to its parent\n",
    "        for child_node in newtables.edges.child:\n",
    "            nodes_and_branches.update({child_node:(tree.branch_length(child_node))}) \n",
    "        # print(nodes_and_branches)\n",
    "        # print(nodes_and_branches.keys())\n",
    "        summed_branch_lengths = sum(nodes_and_branches.values()) #hudson 1990, above equation 1. I could have just used the tskit function but I was going to loop over the nodes to make a dict anwyay \n",
    "        mean_summed_branch_lengths.append(summed_branch_lengths)\n",
    "        proportional_branch_lengths = {} #empty dict to hold the pair of node:scaled branch length\n",
    "        # print(\"sum of the lengths of all branches:\", summed_branch_lengths, \"what the tskit function gives you\", sum(tree.branch_length(u) for u in tree.nodes()))\n",
    "        # print(\"list of nodes in the tree that can have muts placed above them:\", nodes_and_branches.keys()))\n",
    "        for child_node in newtables.edges.child: #I probably can work around this and not have another loop by accessing the dictionary keys, but it doesn't seem worth the effort\n",
    "            proportional_branch_lengths.update({child_node:(tree.branch_length(child_node)/summed_branch_lengths)}) #hudson 1990 equation 1\n",
    "        # proportional_branch_lengths = dict(sorted(proportional_branch_lengths.items())) #sort the dictionary so the keys(representing nodes) go up in order. I thought this would matter for the np array step, but this may not help once we have >1 tree, as the incies of the np array will no longer match the dict\n",
    "        # print(\"nodes and the branch lengths above them, proportional to the total length of all branches:\" , proportional_branch_lengths) \n",
    "        #hudson 1990, equation 1: compare to the expectation of s above        \n",
    "        lambda_total = mu*summed_branch_lengths #this is for what happens in our sim, but not the expectation\n",
    "        genealogy_mutations = rng.poisson(lam=lambda_total)\n",
    "        # print(\"number of mutations we sampled from poisson distribution:\", genealogy_mutations, \"mean number of mutations in the poisson distribution expectations for this tree\", lambda_total)\n",
    "        # print(\"proportional branch lengths:\", proportional_branch_lengths)\n",
    "        p = list(proportional_branch_lengths.values())\n",
    "        if genealogy_mutations != 0:\n",
    "            # print(\"genealogy contains more than 0 mutations\")\n",
    "            ###########sample nodes from your list of child nodes \n",
    "            nodes_sampled =  rng.choice(list(nodes_and_branches.keys()), p=p, size = genealogy_mutations) #this can sample the same node twice, which IS what you want. however need to figure out a way to determine which parent node was associated with a given branch length, so you can select a site that makes sense given the edges possible for a child:parent node relationship. int is a pretty crude \n",
    "            # print(\"nodes sampled from distribution, aka chosen to have mutations\", nodes_sampled)\n",
    "            ###########pick out the spot in the genome that the tree we have iterated over actually occupies (with edges) \n",
    "            # print(\"interval this tree occupies\", tree.interval) #access this to pick which sites are possible\n",
    "            for child_node in nodes_sampled: #since I do a pretty much identical thing above, there's gotta be a way to combine the two, but I'll leave the stunts for later \n",
    "                ###########draw a site from the genome that HASN'T ALREADY had a mutation assigned to it.\n",
    "                # print(\"node mutation is being added onto:\", child_node) \n",
    "                possible_bases = set(range((int(interval.left)), int(interval.right))) #set them to integers for now? need to figure out how to actually implement using sets \n",
    "                bases_already_mutated = set(int(i) for i in newtables.sites.position)\n",
    "                unmutated_sites =(possible_bases.difference(bases_already_mutated)) #get the \n",
    "                position = np.random.randint(low=(interval.left), high=(int(interval.right)), size=1)\n",
    "                position = position[0]\n",
    "                # print(\"position of new mut:\", position)\n",
    "                ############use these nodes to write mutations to the mutations table\n",
    "                # print(\"position\", position)\n",
    "                if position in unmutated_sites: #size should be 1, since we're only placing one mutation above each loop over a node\n",
    "                    site_id = newtables.sites.add_row(position=position, ancestral_state=\"0\")\n",
    "                    # print(\"id of the site we are adding\", site_id)\n",
    "                    #####get times for the mutations- randomly select them from the interval between parent and child node times \n",
    "                    #get child node time\n",
    "                    child_time = newtables.nodes.time[child_node]\n",
    "                    # print(\"node ID\", child_node,  ts.tables.edges.child[child_node])#https://stackoverflow.com/questions/176918/finding-the-index-of-an-item-in-a-\n",
    "                    #now, find the parent node that goes with the child node: \n",
    "                    index = np.where(newtables.edges.child==child_node)[0][0] #https://www.geeksforgeeks.org/how-to-find-the-index-of-value-in-numpy-array/\n",
    "                    # print(\"index of the table entry:\", index, \"that contains this child node:\", child_node)\n",
    "                    parent_node = newtables.edges.parent[index]\n",
    "                    # print(\"parent node\", parent_node)\n",
    "                    parent_time = newtables.nodes.time[parent_node]\n",
    "                    # print(\"time\", child_time, \"of the child node\", child_node, \"time of the parent node\", parent_time)\n",
    "                    time = np.random.uniform(low=child_time, high=parent_time)\n",
    "                    # print(\"time the mutation has been randomly selected to have\", time)\n",
    "                    newtables.mutations.add_row(site=site_id, node=child_node, time=time, derived_state = \"1\") #this doesn't work as with the above-- perhaps something with the strings. \n",
    "                    # print(tskit.unpack_strings(newtables.sites.ancestral_state, newtables.sites.ancestral_state_offset))\n",
    "                    # elif genealogy_mutations == 0:\n",
    "                    #   print(\"genealogy does not contain any mutations\")\n",
    "            # print(newtables.mutations)\n",
    "    newtables.sort(edge_start=0, mutation_start=0, site_start=0)\n",
    "    new_ts = newtables.tree_sequence()\n",
    "    ####do sfs stats\n",
    "    for tree in new_ts.trees():\n",
    "        for mutation in tree.mutations():\n",
    "            # print(\"looking at mutation on node:\", mutation.node) \n",
    "            count = tree.num_samples(mutation.node)\n",
    "            # print(\"mutation count:\", count)\n",
    "            i_values.append(count) #dumping all i values into \n",
    "    # print(\"values for i, the number of occurences of a mutation\", i_values)\n",
    "    # print(\"one loop of a sim::\", \"mutation numbers:\", genealogy_mutations, \"occurrences of these mutations\", i_values)\n",
    "    # return [genealogy_mutations, lambda_total, i_values]\n",
    "# print(newtables)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(\"number of samples for the runs\", ts.num_samples) #confusing that ts.num_samples is for haploid lineages, while samples in sim_ancestry is for individuals (which may not be haploid, and defaults to diploid). Note that this will be representative of the last loop over which tree sequence we're dealing with\n",
    "#theoretical expectation for time to MRCA in units of 2N generations:\n",
    "expected_tmrca = 2*(1-(1/samples))\n",
    "print(\"expected Tmrca:\", expected_tmrca)\n",
    "print(\"mean tmrca, in generations/(ploidy*N):\", mean_tmrca/sim_runs/population_sequences, \"mean tmrca, in raw gens:\", mean_tmrca/sim_runs) #the first value is generations/(ploidy*N),and the second is in generations.\n",
    "#expected time that i lineages should exist: T(i) =  4N/i(i-1) (hudson 2015, #2)\n",
    "#sum i*(T(i)) from i=2 to n to get the total expected branch lengths: (above equation 6, hudson 1990)\n",
    "expected_total_length = 0\n",
    "for i in range(2, samples):\n",
    "    expected_total_length += i*((4*(population_sequences/ploidy)/(i*(i-1))))\n",
    "\n",
    "print(\"total expected branch lengths\", expected_total_length, \"mean summed branch lengths:\", (sum(mean_summed_branch_lengths)/sim_runs))\n",
    "#now, we can use this to find the expected total number of polymorphisms (hudson 1990 eq1)\n",
    "expected_lambda = mu*expected_total_length\n",
    "print(\"mean expectation for the number of polymorphisms on the tree\", expected_lambda)\n",
    "\n",
    "#sfs stats\n",
    "theta = 4*(population_sequences/ploidy)*mu\n",
    "# print(\"mutation counts from all loops\", i_values)\n",
    "mean_counts = {}\n",
    "for i in range(1, ((samples*2)-1)):\n",
    "    mean = (i_values.count(i))/sim_runs\n",
    "    mean_counts.update({i:mean})\n",
    "# mean_allele_freqs = (i_values.count(i))/sim_runs\n",
    "print(\"mean allele frequencies, from all sims binned\", mean_counts)\n",
    "# # print(\"average number of mutations in sims\", sum(mutations_in_run)/sim_runs, \"expected mean number of mutations\", expected_mutations)\n",
    "print(\"expected mean site frequency spectrum\", [(theta/i) for i in range(1, samples)], \"mean allele frequencies from sims, averaged\", mean_counts)\n",
    "ind = np.arange(len(mean_counts))\n",
    "plt.bar(ind, list(mean_counts.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected mean site frequency spectrum [2.0, 1.0, 0.6666666666666666, 0.5, 0.4, 0.3333333333333333, 0.2857142857142857, 0.25, 0.2222222222222222] mean allele frequencies from sims, averaged {1: 1.966, 2: 1.028, 3: 0.608, 4: 0.418, 5: 0.444, 6: 0.3, 7: 0.294, 8: 0.188, 9: 0.224, 10: 0.0, 11: 0.0, 12: 0.0, 13: 0.0, 14: 0.0, 15: 0.0, 16: 0.0, 17: 0.0, 18: 0.0}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BarContainer object of 18 artists>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAGdCAYAAAAxCSikAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAqjUlEQVR4nO3de3BUZZ7G8acJpgMUaUBIOtFIgiIIQoI4ZMLoKmNLyFIM2dlFoNwBsoC7LFTJxBuxJNHV2iAqA85miRdCYBy5lQq1wgYxGlgkwHJJKa5SwIabpMNlJ2kSh4RNzv5h0W6bGyfk8qb5fqpOSb/nd95+X1/SeTh9TrfDsixLAAAABuvW2QMAAABoCYEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGC87p09gLZQX1+vc+fOqXfv3nI4HJ09HAAAcB0sy9Lly5cVHR2tbt2aP4cSFIHl3LlziomJ6exhAACAVjhz5oxuv/32ZmuCIrD07t1b0g8TDg8P7+TRAACA6+Hz+RQTE+P/Pd6coAgs194GCg8PJ7AAANDFXM/lHFx0CwAAjEdgAQAAxiOwAAAA4xFYAACA8WwFluzsbP3sZz9T7969FRERodTUVB09erTF4zZt2qShQ4cqLCxMI0aM0LZt2wL2W5alzMxMRUVFqUePHvJ4PDp27Ji9mQAAgKBlK7Ds3LlT8+fP1969e7Vjxw5dvXpV48ePV3V1dZPH7NmzR9OnT9fs2bN1+PBhpaamKjU1VUeOHPHXLF26VG+++aZyc3O1b98+9erVS8nJybpy5UrrZwYAAIKGw7Isq7UHX7hwQREREdq5c6f+4i/+otGaqVOnqrq6Wh9//LG/7ec//7kSEhKUm5sry7IUHR2tp556Sk8//bQkqbKyUpGRkcrPz9e0adNaHIfP55PL5VJlZSW3NQMA0EXY+f19Q9ewVFZWSpL69evXZE1xcbE8Hk9AW3JysoqLiyVJpaWl8nq9ATUul0uJiYn+mp+qqamRz+cL2AAAQPBqdWCpr6/XwoUL9Ytf/EL33ntvk3Ver1eRkZEBbZGRkfJ6vf7919qaqvmp7OxsuVwu/8bH8gMAENxaHVjmz5+vI0eOaP369W05nuuSkZGhyspK/3bmzJkOHwMAAOg4rfpo/gULFujjjz/Wrl27WvyyIrfbrfLy8oC28vJyud1u//5rbVFRUQE1CQkJjfbpdDrldDpbM3QAANAF2TrDYlmWFixYoI8++kifffaZ4uLiWjwmKSlJhYWFAW07duxQUlKSJCkuLk5utzugxufzad++ff4aAABwc7N1hmX+/Pl6//33tWXLFvXu3dt/jYnL5VKPHj0kSTNmzNBtt92m7OxsSdKTTz6phx56SG+88YYmTpyo9evX68CBA3r77bcl/fCFRwsXLtQrr7yiwYMHKy4uTosXL1Z0dLRSU1PbcKoAAKCrshVYVq5cKUl6+OGHA9pXr16tWbNmSZJOnz6tbt1+PHEzduxYvf/++3rhhRf0/PPPa/Dgwdq8eXPAhbrPPvusqqur9cQTT6iiokIPPPCACgoKFBYW1sppAQCAYHJDn8Niivb+HJbYRVvbtL+TSya2aX8AAHRFHfY5LAAAAB2BwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjGc7sOzatUuTJk1SdHS0HA6HNm/e3Gz9rFmz5HA4GmzDhw/317z44osN9g8dOtT2ZAAAQHCyHViqq6sVHx+vnJyc66pfsWKFysrK/NuZM2fUr18/TZkyJaBu+PDhAXW7d++2OzQAABCkuts9ICUlRSkpKddd73K55HK5/I83b96sP/3pT0pLSwscSPfucrvddocDAABuAh1+DcuqVavk8Xg0cODAgPZjx44pOjpagwYN0uOPP67Tp0832UdNTY18Pl/ABgAAgleHBpZz587p3//93zVnzpyA9sTEROXn56ugoEArV65UaWmpHnzwQV2+fLnRfrKzs/1nblwul2JiYjpi+AAAoJN0aGBZs2aN+vTpo9TU1ID2lJQUTZkyRSNHjlRycrK2bdumiooKbdy4sdF+MjIyVFlZ6d/OnDnTAaMHAACdxfY1LK1lWZby8vL0m9/8RqGhoc3W9unTR3fffbeOHz/e6H6n0ymn09kewwQAAAbqsDMsO3fu1PHjxzV79uwWa6uqqnTixAlFRUV1wMgAAIDpbAeWqqoqlZSUqKSkRJJUWlqqkpIS/0WyGRkZmjFjRoPjVq1apcTERN17770N9j399NPauXOnTp48qT179uiv/uqvFBISounTp9sdHgAACEK23xI6cOCAxo0b53+cnp4uSZo5c6by8/NVVlbW4A6fyspKffDBB1qxYkWjfZ49e1bTp0/XpUuXNGDAAD3wwAPau3evBgwYYHd4AAAgCDksy7I6exA3yufzyeVyqbKyUuHh4W3ef+yirW3a38klE9u0PwAAuiI7v7/5LiEAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMJ7twLJr1y5NmjRJ0dHRcjgc2rx5c7P1RUVFcjgcDTav1xtQl5OTo9jYWIWFhSkxMVH79++3OzQAABCkbAeW6upqxcfHKycnx9ZxR48eVVlZmX+LiIjw79uwYYPS09OVlZWlQ4cOKT4+XsnJyTp//rzd4QEAgCDU3e4BKSkpSklJsf1EERER6tOnT6P7li1bprlz5yotLU2SlJubq61btyovL0+LFi2y/VwAACC4dNg1LAkJCYqKitKjjz6qL774wt9eW1urgwcPyuPx/Diobt3k8XhUXFzcaF81NTXy+XwBGwAACF7tHliioqKUm5urDz74QB988IFiYmL08MMP69ChQ5Kkixcvqq6uTpGRkQHHRUZGNrjO5Zrs7Gy5XC7/FhMT097TAAAAncj2W0J2DRkyREOGDPE/Hjt2rE6cOKHf/e53+sMf/tCqPjMyMpSenu5/7PP5CC0AAASxdg8sjRkzZox2794tSerfv79CQkJUXl4eUFNeXi63293o8U6nU06ns93HCQAAzNApn8NSUlKiqKgoSVJoaKhGjx6twsJC//76+noVFhYqKSmpM4YHAAAMY/sMS1VVlY4fP+5/XFpaqpKSEvXr10933HGHMjIy9N1332nt2rWSpOXLlysuLk7Dhw/XlStX9O677+qzzz7TJ5984u8jPT1dM2fO1P33368xY8Zo+fLlqq6u9t81BAAAbm62A8uBAwc0btw4/+Nr15LMnDlT+fn5Kisr0+nTp/37a2tr9dRTT+m7775Tz549NXLkSH366acBfUydOlUXLlxQZmamvF6vEhISVFBQ0OBCXAAAcHNyWJZldfYgbpTP55PL5VJlZaXCw8PbvP/YRVvbtL+TSya2aX8AAHRFdn5/811CAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4tgPLrl27NGnSJEVHR8vhcGjz5s3N1n/44Yd69NFHNWDAAIWHhyspKUnbt28PqHnxxRflcDgCtqFDh9odGgAACFK2A0t1dbXi4+OVk5NzXfW7du3So48+qm3btungwYMaN26cJk2apMOHDwfUDR8+XGVlZf5t9+7ddocGAACCVHe7B6SkpCglJeW665cvXx7w+J//+Z+1ZcsW/du//ZtGjRr140C6d5fb7bY7HAAAcBPo8GtY6uvrdfnyZfXr1y+g/dixY4qOjtagQYP0+OOP6/Tp0032UVNTI5/PF7ABAIDg1eGB5fXXX1dVVZUee+wxf1tiYqLy8/NVUFCglStXqrS0VA8++KAuX77caB/Z2dlyuVz+LSYmpqOGDwAAOkGHBpb3339fL730kjZu3KiIiAh/e0pKiqZMmaKRI0cqOTlZ27ZtU0VFhTZu3NhoPxkZGaqsrPRvZ86c6agpAACATmD7GpbWWr9+vebMmaNNmzbJ4/E0W9unTx/dfffdOn78eKP7nU6nnE5newwTAAAYqEPOsKxbt05paWlat26dJk6c2GJ9VVWVTpw4oaioqA4YHQAAMJ3tMyxVVVUBZz5KS0tVUlKifv366Y477lBGRoa+++47rV27VtIPbwPNnDlTK1asUGJiorxerySpR48ecrlckqSnn35akyZN0sCBA3Xu3DllZWUpJCRE06dPb4s5AgCALs72GZYDBw5o1KhR/luS09PTNWrUKGVmZkqSysrKAu7wefvtt/W///u/mj9/vqKiovzbk08+6a85e/aspk+friFDhuixxx7Trbfeqr1792rAgAE3Oj8AABAEHJZlWZ09iBvl8/nkcrlUWVmp8PDwNu8/dtHWNu3v5JKW3xYDACDY2fn9zXcJAQAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjde/sAeAHsYu2tnmfJ5dMbPM+AQDoDLbPsOzatUuTJk1SdHS0HA6HNm/e3OIxRUVFuu++++R0OnXXXXcpPz+/QU1OTo5iY2MVFhamxMRE7d+/3+7QAABAkLIdWKqrqxUfH6+cnJzrqi8tLdXEiRM1btw4lZSUaOHChZozZ462b9/ur9mwYYPS09OVlZWlQ4cOKT4+XsnJyTp//rzd4QEAgCBk+y2hlJQUpaSkXHd9bm6u4uLi9MYbb0iS7rnnHu3evVu/+93vlJycLElatmyZ5s6dq7S0NP8xW7duVV5enhYtWmR3iAAAIMi0+0W3xcXF8ng8AW3JyckqLi6WJNXW1urgwYMBNd26dZPH4/HX/FRNTY18Pl/ABgAAgle7Bxav16vIyMiAtsjISPl8Pv35z3/WxYsXVVdX12iN1+tttM/s7Gy5XC7/FhMT027jBwAAna9L3tackZGhyspK/3bmzJnOHhIAAGhH7X5bs9vtVnl5eUBbeXm5wsPD1aNHD4WEhCgkJKTRGrfb3WifTqdTTqez3cYMAADM0u5nWJKSklRYWBjQtmPHDiUlJUmSQkNDNXr06ICa+vp6FRYW+msAAMDNzXZgqaqqUklJiUpKSiT9cNtySUmJTp8+LemHt2tmzJjhr/+Hf/gH/fd//7eeffZZffvtt/rXf/1Xbdy4Ub/97W/9Nenp6XrnnXe0Zs0affPNN5o3b56qq6v9dw0BAICbm+23hA4cOKBx48b5H6enp0uSZs6cqfz8fJWVlfnDiyTFxcVp69at+u1vf6sVK1bo9ttv17vvvuu/pVmSpk6dqgsXLigzM1Ner1cJCQkqKChocCEuAAC4OTksy7I6exA3yufzyeVyqbKyUuHh4W3ef1t/bH5jH5nPR/MDAG42dn5/d8m7hAAAwM2FwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjNeqwJKTk6PY2FiFhYUpMTFR+/fvb7L24YcflsPhaLBNnDjRXzNr1qwG+ydMmNCaoQEAgCDU3e4BGzZsUHp6unJzc5WYmKjly5crOTlZR48eVURERIP6Dz/8ULW1tf7Hly5dUnx8vKZMmRJQN2HCBK1evdr/2Ol02h0aAAAIUrbPsCxbtkxz585VWlqahg0bptzcXPXs2VN5eXmN1vfr109ut9u/7dixQz179mwQWJxOZ0Bd3759WzcjAAAQdGwFltraWh08eFAej+fHDrp1k8fjUXFx8XX1sWrVKk2bNk29evUKaC8qKlJERISGDBmiefPm6dKlS032UVNTI5/PF7ABAIDgZSuwXLx4UXV1dYqMjAxoj4yMlNfrbfH4/fv368iRI5ozZ05A+4QJE7R27VoVFhbq1Vdf1c6dO5WSkqK6urpG+8nOzpbL5fJvMTExdqYBAAC6GNvXsNyIVatWacSIERozZkxA+7Rp0/x/HjFihEaOHKk777xTRUVFeuSRRxr0k5GRofT0dP9jn89HaAEAIIjZOsPSv39/hYSEqLy8PKC9vLxcbre72WOrq6u1fv16zZ49u8XnGTRokPr376/jx483ut/pdCo8PDxgAwAAwctWYAkNDdXo0aNVWFjob6uvr1dhYaGSkpKaPXbTpk2qqanR3/7t37b4PGfPntWlS5cUFRVlZ3gAACBI2b5LKD09Xe+8847WrFmjb775RvPmzVN1dbXS0tIkSTNmzFBGRkaD41atWqXU1FTdeuutAe1VVVV65plntHfvXp08eVKFhYWaPHmy7rrrLiUnJ7dyWgAAIJjYvoZl6tSpunDhgjIzM+X1epWQkKCCggL/hbinT59Wt26BOejo0aPavXu3Pvnkkwb9hYSE6Msvv9SaNWtUUVGh6OhojR8/Xi+//DKfxQIAACS18qLbBQsWaMGCBY3uKyoqatA2ZMgQWZbVaH2PHj20ffv21gwDAADcJPguIQAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADBe984eADpW7KKtbdrfySUT27Q/AAAawxkWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPFaFVhycnIUGxursLAwJSYmav/+/U3W5ufny+FwBGxhYWEBNZZlKTMzU1FRUerRo4c8Ho+OHTvWmqEBAIAgZDuwbNiwQenp6crKytKhQ4cUHx+v5ORknT9/vsljwsPDVVZW5t9OnToVsH/p0qV68803lZubq3379qlXr15KTk7WlStX7M8IAAAEHduBZdmyZZo7d67S0tI0bNgw5ebmqmfPnsrLy2vyGIfDIbfb7d8iIyP9+yzL0vLly/XCCy9o8uTJGjlypNauXatz585p8+bNrZoUAAAILrYCS21trQ4ePCiPx/NjB926yePxqLi4uMnjqqqqNHDgQMXExGjy5Mn6+uuv/ftKS0vl9XoD+nS5XEpMTGyyz5qaGvl8voANAAAEL1uB5eLFi6qrqws4QyJJkZGR8nq9jR4zZMgQ5eXlacuWLXrvvfdUX1+vsWPH6uzZs5LkP85On9nZ2XK5XP4tJibGzjQAAEAX0+53CSUlJWnGjBlKSEjQQw89pA8//FADBgzQW2+91eo+MzIyVFlZ6d/OnDnThiMGAACmsRVY+vfvr5CQEJWXlwe0l5eXy+12X1cft9xyi0aNGqXjx49Lkv84O306nU6Fh4cHbAAAIHjZCiyhoaEaPXq0CgsL/W319fUqLCxUUlLSdfVRV1enr776SlFRUZKkuLg4ud3ugD59Pp/27dt33X0CAIDg1t3uAenp6Zo5c6buv/9+jRkzRsuXL1d1dbXS0tIkSTNmzNBtt92m7OxsSdI//dM/6ec//7nuuusuVVRU6LXXXtOpU6c0Z84cST/cQbRw4UK98sorGjx4sOLi4rR48WJFR0crNTW17WYKAAC6LNuBZerUqbpw4YIyMzPl9XqVkJCggoIC/0Wzp0+fVrduP564+dOf/qS5c+fK6/Wqb9++Gj16tPbs2aNhw4b5a5599llVV1friSeeUEVFhR544AEVFBQ0+IA5AABwc3JYlmV19iBulM/nk8vlUmVlZbtczxK7aGub9ndyycR2f46Oep7GngMAgOth5/c33yUEAACMR2ABAADGI7AAAADj2b7oFjBBR13zAwAwA2dYAACA8QgsAADAeAQWAABgPK5hQZvjs14AAG2NMywAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACM172zBwCYLHbR1jbt7+SSiW3aHwDcLAgsQCcjFAFAy3hLCAAAGI/AAgAAjNeqwJKTk6PY2FiFhYUpMTFR+/fvb7L2nXfe0YMPPqi+ffuqb9++8ng8DepnzZolh8MRsE2YMKE1QwMAAEHIdmDZsGGD0tPTlZWVpUOHDik+Pl7Jyck6f/58o/VFRUWaPn26Pv/8cxUXFysmJkbjx4/Xd999F1A3YcIElZWV+bd169a1bkYAACDo2A4sy5Yt09y5c5WWlqZhw4YpNzdXPXv2VF5eXqP1f/zjH/WP//iPSkhI0NChQ/Xuu++qvr5ehYWFAXVOp1Nut9u/9e3bt3UzAgAAQcdWYKmtrdXBgwfl8Xh+7KBbN3k8HhUXF19XH99//72uXr2qfv36BbQXFRUpIiJCQ4YM0bx583Tp0iU7QwMAAEHM1m3NFy9eVF1dnSIjIwPaIyMj9e23315XH88995yio6MDQs+ECRP061//WnFxcTpx4oSef/55paSkqLi4WCEhIQ36qKmpUU1Njf+xz+ezMw0AANDFdOjnsCxZskTr169XUVGRwsLC/O3Tpk3z/3nEiBEaOXKk7rzzThUVFemRRx5p0E92drZeeumlDhkzAADofLbeEurfv79CQkJUXl4e0F5eXi63293ssa+//rqWLFmiTz75RCNHjmy2dtCgQerfv7+OHz/e6P6MjAxVVlb6tzNnztiZBgAA6GJsBZbQ0FCNHj064ILZaxfQJiUlNXnc0qVL9fLLL6ugoED3339/i89z9uxZXbp0SVFRUY3udzqdCg8PD9gAAEDwsn2XUHp6ut555x2tWbNG33zzjebNm6fq6mqlpaVJkmbMmKGMjAx//auvvqrFixcrLy9PsbGx8nq98nq9qqqqkiRVVVXpmWee0d69e3Xy5EkVFhZq8uTJuuuuu5ScnNxG0wQAAF2Z7WtYpk6dqgsXLigzM1Ner1cJCQkqKCjwX4h7+vRpdev2Yw5auXKlamtr9Td/8zcB/WRlZenFF19USEiIvvzyS61Zs0YVFRWKjo7W+PHj9fLLL8vpdN7g9AB0JL4XCUB7adVFtwsWLNCCBQsa3VdUVBTw+OTJk8321aNHD23fvr01wwAAADcJvksIAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxOvSj+QF0jra+3VjilmMAHYszLAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGK9VgSUnJ0exsbEKCwtTYmKi9u/f32z9pk2bNHToUIWFhWnEiBHatm1bwH7LspSZmamoqCj16NFDHo9Hx44da83QAABAELIdWDZs2KD09HRlZWXp0KFDio+PV3Jyss6fP99o/Z49ezR9+nTNnj1bhw8fVmpqqlJTU3XkyBF/zdKlS/Xmm28qNzdX+/btU69evZScnKwrV660fmYAACBo2A4sy5Yt09y5c5WWlqZhw4YpNzdXPXv2VF5eXqP1K1as0IQJE/TMM8/onnvu0csvv6z77rtP//Iv/yLph7Mry5cv1wsvvKDJkydr5MiRWrt2rc6dO6fNmzff0OQAAEBw6G6nuLa2VgcPHlRGRoa/rVu3bvJ4PCouLm70mOLiYqWnpwe0JScn+8NIaWmpvF6vPB6Pf7/L5VJiYqKKi4s1bdq0Bn3W1NSopqbG/7iyslKS5PP57EznutXXfN+m/TU2zrZ+jo56nmB5jo56nmB5jo56nvb6mQZghms/45ZltVhrK7BcvHhRdXV1ioyMDGiPjIzUt99+2+gxXq+30Xqv1+vff62tqZqfys7O1ksvvdSgPSYm5vom0slcy4PneYLlOTrqeYLlOTrqeTpqLgA61+XLl+VyuZqtsRVYTJGRkRFw1qa+vl7/8z//o1tvvVUOh6NTxuTz+RQTE6MzZ84oPDy8U8bQWW7muUs39/yZ+805d+nmnj9zb7u5W5aly5cvKzo6usVaW4Glf//+CgkJUXl5eUB7eXm53G53o8e43e5m66/9t7y8XFFRUQE1CQkJjfbpdDrldDoD2vr06WNnKu0mPDz8pvsLfM3NPHfp5p4/c7855y7d3PNn7m0z95bOrFxj66Lb0NBQjR49WoWFhf62+vp6FRYWKikpqdFjkpKSAuolaceOHf76uLg4ud3ugBqfz6d9+/Y12ScAALi52H5LKD09XTNnztT999+vMWPGaPny5aqurlZaWpokacaMGbrtttuUnZ0tSXryySf10EMP6Y033tDEiRO1fv16HThwQG+//bYkyeFwaOHChXrllVc0ePBgxcXFafHixYqOjlZqamrbzRQAAHRZtgPL1KlTdeHCBWVmZsrr9SohIUEFBQX+i2ZPnz6tbt1+PHEzduxYvf/++3rhhRf0/PPPa/Dgwdq8ebPuvfdef82zzz6r6upqPfHEE6qoqNADDzyggoIChYWFtcEUO4bT6VRWVlaDt6puBjfz3KWbe/7M/eacu3Rzz5+5d87cHdb13EsEAADQifguIQAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgsSEnJ0exsbEKCwtTYmKi9u/f32z9pk2bNHToUIWFhWnEiBHatm1bB420bWVnZ+tnP/uZevfurYiICKWmpuro0aPNHpOfny+HwxGwdaW7vq558cUXG8xj6NChzR4TLOseGxvbYO4Oh0Pz589vtL6rr/muXbs0adIkRUdHy+FwNPjyVcuylJmZqaioKPXo0UMej0fHjh1rsV+7rxudobm5X716Vc8995xGjBihXr16KTo6WjNmzNC5c+ea7bM1PzudoaV1nzVrVoN5TJgwocV+u8K6Sy3Pv7HXAIfDoddee63JPttr7Qks12nDhg1KT09XVlaWDh06pPj4eCUnJ+v8+fON1u/Zs0fTp0/X7NmzdfjwYaWmpio1NVVHjhzp4JHfuJ07d2r+/Pnau3evduzYoatXr2r8+PGqrq5u9rjw8HCVlZX5t1OnTnXQiNvW8OHDA+axe/fuJmuDad3/8z//M2DeO3bskCRNmTKlyWO68ppXV1crPj5eOTk5je5funSp3nzzTeXm5mrfvn3q1auXkpOTdeXKlSb7tPu60Vmam/v333+vQ4cOafHixTp06JA+/PBDHT16VL/61a9a7NfOz05naWndJWnChAkB81i3bl2zfXaVdZdanv//n3dZWZny8vLkcDj013/918322y5rb+G6jBkzxpo/f77/cV1dnRUdHW1lZ2c3Wv/YY49ZEydODGhLTEy0/v7v/75dx9kRzp8/b0mydu7c2WTN6tWrLZfL1XGDaidZWVlWfHz8ddcH87o/+eST1p133mnV19c3uj9Y1tyyLEuS9dFHH/kf19fXW26323rttdf8bRUVFZbT6bTWrVvXZD92XzdM8NO5N2b//v2WJOvUqVNN1tj92TFBY3OfOXOmNXnyZFv9dMV1t6zrW/vJkydbv/zlL5utaa+15wzLdaitrdXBgwfl8Xj8bd26dZPH41FxcXGjxxQXFwfUS1JycnKT9V1JZWWlJKlfv37N1lVVVWngwIGKiYnR5MmT9fXXX3fE8NrcsWPHFB0drUGDBunxxx/X6dOnm6wN1nWvra3Ve++9p7/7u79r9gtGg2XNf6q0tFRerzdgbV0ulxITE5tc29a8bnQVlZWVcjgcLX6Hm52fHZMVFRUpIiJCQ4YM0bx583Tp0qUma4N53cvLy7V161bNnj27xdr2WHsCy3W4ePGi6urq/J/me01kZKS8Xm+jx3i9Xlv1XUV9fb0WLlyoX/ziFwGfVvxTQ4YMUV5enrZs2aL33ntP9fX1Gjt2rM6ePduBo71xiYmJys/PV0FBgVauXKnS0lI9+OCDunz5cqP1wbrumzdvVkVFhWbNmtVkTbCseWOurZ+dtW3N60ZXcOXKFT333HOaPn16s19+Z/dnx1QTJkzQ2rVrVVhYqFdffVU7d+5USkqK6urqGq0P1nWXpDVr1qh379769a9/3Wxde6297Y/mx81t/vz5OnLkSIvvRyYlJQV8eeXYsWN1zz336K233tLLL7/c3sNsMykpKf4/jxw5UomJiRo4cKA2btx4Xf/KCBarVq1SSkpKs18BHyxrjqZdvXpVjz32mCzL0sqVK5utDZafnWnTpvn/PGLECI0cOVJ33nmnioqK9Mgjj3TiyDpeXl6eHn/88RYvpm+vtecMy3Xo37+/QkJCVF5eHtBeXl4ut9vd6DFut9tWfVewYMECffzxx/r88891++232zr2lltu0ahRo3T8+PF2Gl3H6NOnj+6+++4m5xGM637q1Cl9+umnmjNnjq3jgmXNJfnXz87atuZ1w2TXwsqpU6e0Y8eOZs+uNKaln52uYtCgQerfv3+T8wi2db/mP/7jP3T06FHbrwNS2609geU6hIaGavTo0SosLPS31dfXq7CwMOBflP9fUlJSQL0k7dixo8l6k1mWpQULFuijjz7SZ599pri4ONt91NXV6auvvlJUVFQ7jLDjVFVV6cSJE03OI5jW/ZrVq1crIiJCEydOtHVcsKy5JMXFxcntdgesrc/n0759+5pc29a8bpjqWlg5duyYPv30U9166622+2jpZ6erOHv2rC5dutTkPIJp3f+/VatWafTo0YqPj7d9bJutfZtfxhuk1q9fbzmdTis/P9/6r//6L+uJJ56w+vTpY3m9XsuyLOs3v/mNtWjRIn/9F198YXXv3t16/fXXrW+++cbKysqybrnlFuurr77qrCm02rx58yyXy2UVFRVZZWVl/u3777/31/x0/i+99JK1fft268SJE9bBgwetadOmWWFhYdbXX3/dGVNotaeeesoqKiqySktLrS+++MLyeDxW//79rfPnz1uWFdzrblk/3N1wxx13WM8991yDfcG25pcvX7YOHz5sHT582JJkLVu2zDp8+LD/TpglS5ZYffr0sbZs2WJ9+eWX1uTJk624uDjrz3/+s7+PX/7yl9bvf/97/+OWXjdM0dzca2trrV/96lfW7bffbpWUlAS8BtTU1Pj7+OncW/rZMUVzc798+bL19NNPW8XFxVZpaan16aefWvfdd581ePBg68qVK/4+uuq6W1bLf+8ty7IqKyutnj17WitXrmy0j45aewKLDb///e+tO+64wwoNDbXGjBlj7d2717/voYcesmbOnBlQv3HjRuvuu++2QkNDreHDh1tbt27t4BG3DUmNbqtXr/bX/HT+Cxcu9P+/ioyMtP7yL//SOnToUMcP/gZNnTrVioqKskJDQ63bbrvNmjp1qnX8+HH//mBed8uyrO3bt1uSrKNHjzbYF2xr/vnnnzf69/zaHOvr663FixdbkZGRltPptB555JEG/18GDhxoZWVlBbQ197phiubmXlpa2uRrwOeff+7v46dzb+lnxxTNzf3777+3xo8fbw0YMMC65ZZbrIEDB1pz585tEDy66rpbVst/7y3Lst566y2rR48eVkVFRaN9dNTaOyzLsm7sHA0AAED74hoWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIz3f15VdVLxKQnBAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#do the above with multiple trees\n",
    "import msprime\n",
    "import tskit \n",
    "import numpy as np\n",
    "from numpy.random import default_rng\n",
    "rng = default_rng()\n",
    "import random\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sequence_length = 100\n",
    "mu = (1e-3) #normally would have this set to 1e-7 *2, but putting deme size down and #individuals up to be able to visualize tree sequences for working on placing mutations issue.\n",
    "# sequence_length = 100\n",
    "ploidy = 2\n",
    "population_sequences = 1000 #total number of genomes in the population- the number of individuals is this/ploidy\n",
    "sampled_sequences = 10\n",
    "sim_runs = 500\n",
    "rho = (1e-6)*5\n",
    "#initialize the bottom two variables, to add/append to later \n",
    "mean_tmrca = 0\n",
    "i_values = []\n",
    "mean_summed_branch_lengths = []\n",
    "\n",
    "for ts in msprime.sim_ancestry(ploidy= ploidy, samples=(sampled_sequences/ploidy), discrete_genome = True, population_size=(population_sequences)/ploidy, num_replicates = sim_runs, sequence_length = sequence_length, recombination_rate=rho): #the sim needs everything in terms of the number of individuals, but it's convenient for downstream/interpretation to have our original variables be in terms of the number of lineages or individual sequences, aka in haploid terms \n",
    "    newtables = ts.dump_tables() #modifying a copy of the original tables in order to be able to overwrite \n",
    "    assert ts.num_samples == samples\n",
    "    last_node = ts.num_nodes - 1\n",
    "    tmrca = ts.node(last_node).time\n",
    "    mean_tmrca += tmrca\n",
    "    # print(\"tmrca for this run:\", tmrca)\n",
    "    # print(ts.draw_text())\n",
    "    # print(newtables.edges)\n",
    "    genealogy_sizes = []\n",
    "    weighted_genealogy_sizes = []\n",
    "    sequence_weights = []\n",
    "    for tree in ts.trees():\n",
    "        # print(tree.draw(format='unicode'))\n",
    "        current_tree_interval = tree.interval #find the interval the tree occupies in the genome \n",
    "        span = tree.interval.right-tree.interval.left #find the fraction of the total genome/number of bases, for multiple trees \n",
    "        # print(newtables.edges)\n",
    "        # print(\"interval this tree exists on\", current_tree_interval)\n",
    "        nodes_and_branches = {} #empty dict to hold the pair of node:branch length to its parent. each tree gets its own list of these \n",
    "        for node in tree.nodes(): #needed in the recombination version of the sim due to not all nodes being present in all trees \n",
    "            for child_node in newtables.edges.child:\n",
    "                if child_node ==node: \n",
    "                    nodes_and_branches.update({child_node:(tree.branch_length(child_node))}) \n",
    "                    \n",
    "        # x = (list(nodes_and_branches.keys()))\n",
    "        # x.sort()\n",
    "        # print(\"list of non-root nodes in the tree (ones that can have muts placed on them):\", x)\n",
    "        summed_branch_lengths = sum(nodes_and_branches.values()) #hudson 1990, above equation 1. I could have just used the tskit function but I was going to loop over the nodes to make a dict anwyay \n",
    "        mean_summed_branch_lengths.append(summed_branch_lengths)\n",
    "        proportional_branch_lengths = {} #empty dict to hold the pair of node:scaled branch length\n",
    "        # print(\"sum of the lengths of all branches:\", summed_branch_lengths, \"what the tskit function gives you\", sum(tree.branch_length(u) for u in tree.nodes()))\n",
    "        # print(\"list of nodes in the tree that can have muts placed above them:\", child_nodes_list)\n",
    "        for node in tree.nodes():\n",
    "            for child_node in newtables.edges.child: #I probably can work around this and not have another loop by accessing the dictionary keys, but it doesn't seem worth the effort\n",
    "                if child_node == node: \n",
    "                    proportional_branch_lengths.update({child_node:(tree.branch_length(child_node)/summed_branch_lengths)}) #hudson 1990 equation 1\n",
    "        # print(proportional_branch_lengths)\n",
    "        # proportional_branch_lengths = dict(sorted(proportional_branch_lengths.items())) #sort the dictionary so the keys(representing nodes) go up in order. I thought this would matter for the np array step, but this may not help once we have >1 tree, as the incies of the np array will no longer match the dict\n",
    "        # print(\"nodes and the branch lengths above them, proportional to the total length of all branches:\" , proportional_branch_lengths) \n",
    "        #hudson 1990, equation 1: compare to the expectation of s above        \n",
    "        lambda_tree = mu*summed_branch_lengths*(span/sequence_length) #this is for what happens in our sim, but not the expectation\n",
    "        tree_mutations = rng.poisson(lam=lambda_tree)\n",
    "        # print(\"number of mutations we sampled from poisson distribution:\", genealogy_mutations, \"mean number of mutations in the poisson distribution expectations for this tree\", lambda_total)\n",
    "        # print(\"proportional branch lengths:\", proportional_branch_lengths)\n",
    "        p = list(proportional_branch_lengths.values())\n",
    "        # print(\"child nodes\", len(child_nodes_list))\n",
    "        # print(\"proportional branch lengths\", len(p))\n",
    " \n",
    "        if tree_mutations != 0:\n",
    "            # print(\"genealogy contains more than 0 mutations\")\n",
    "            ###########sample nodes from your list of child nodes \n",
    "            nodes_sampled =  rng.choice(list(nodes_and_branches.keys()), p=p, size=tree_mutations) #this can sample the same node twice, which IS what you want. however need to figure out a way to determine which parent node was associated with a given branch length, so you can select a site that makes sense given the edges possible for a child:parent node relationship. int is a pretty crude \n",
    "            # print(\"nodes sampled from distribution, aka chosen to have mutations\", nodes_sampled)\n",
    "            ###########pick out the spot in the genome that the tree we have iterated over actually occupies (with edges) \n",
    "            # print(\"interval this tree occupies\", current_tree_interval) #access this to pick which sites are possible\n",
    "            for child_node in nodes_sampled: #since I do a pretty much identical thing above, there's gotta be a way to combine the two, but I'll leave the stunts for later \n",
    "                ###########draw a site from the genome that HASN'T ALREADY had a mutation assigned to it.\n",
    "                # print(\"node mutation is being added onto:\", child_node) \n",
    "                possible_bases = set(range((int(current_tree_interval.left)), int(current_tree_interval.right))) #set them to integers for now? need to figure out how to actually implement using sets \n",
    "                bases_already_mutated = set(int(i) for i in newtables.sites.position)\n",
    "                unmutated_sites =(possible_bases.difference(bases_already_mutated)) #get the \n",
    "                position = np.random.randint(low=(current_tree_interval.left), high=(int(current_tree_interval.right)), size=1) #looks like this is working fine for \n",
    "                position = position[0]\n",
    "                # print(\"position of new mut:\", position)\n",
    "                ############use these nodes to write mutations to the mutations table\n",
    "                if position in unmutated_sites: #size should be 1, since we're only placing one mutation above each loop over a node\n",
    "                    site_id = newtables.sites.add_row(position=position, ancestral_state=\"0\")\n",
    "                    # print(\"id of the site we are adding\", site_id)\n",
    "                    #####get times for the mutations- randomly select them from the interval between parent and child node times \n",
    "                    # get child node time\n",
    "                    child_time = newtables.nodes.time[child_node]\n",
    "                    # print(\"node ID\", child_node,  ts.tables.edges.child[child_node])#https://stackoverflow.com/questions/176918/finding-the-index-of-an-item-in-a-\n",
    "                    #now, find the parent node that goes with the child node: \n",
    "                    index = np.where(newtables.edges.child==child_node)[0] #https://www.geeksforgeeks.org/how-to-find-the-index-of-value-in-numpy-array/\n",
    "                    # print(\"child node\", child_node, \"row in edge tables\", index)\n",
    "                    entries = newtables.edges[index]\n",
    "                    # print(entries)\n",
    "                    for row_id in index:\n",
    "                        entry = newtables.edges[row_id]\n",
    "                        # print(\"indices of the edge table entry\", row_id, \"that contains this child node:\", child_node, \"full entry:\", entry)\n",
    "                        if (entry.left) == current_tree_interval.left:   \n",
    "                            if entry.right == current_tree_interval.right: #if the interval matches the interval of the tree we're looping over exactly\n",
    "                                # print(\"entry that is actually in current tree interval\", entry)\n",
    "                                parent_node = entry.parent\n",
    "                                parent_time = newtables.nodes.time[parent_node]\n",
    "                            elif entry.right >= current_tree_interval.right: \n",
    "                                parent_node = entry.parent\n",
    "                                parent_time = newtables.nodes.time[parent_node]  \n",
    "                        elif entry.left <= current_tree_interval.left:\n",
    "                            if entry.right >= current_tree_interval.right: #if the tree sequence is contained entirely within the entry interval\n",
    "                                parent_node = entry.parent\n",
    "                                parent_time = newtables.nodes.time[parent_node]\n",
    "                        elif (entry.left, entry.right) == (0, sequence_length): #this happens when the node is not associated with a recombination event \n",
    "                            # print(\"this node has the same parent regardless of tree\") #do I need to do this above??)\n",
    "                            parent_node = entry.parent\n",
    "                            parent_time = newtables.nodes.time[parent_node]\n",
    "                        # print(\"parent node\", parent_node)\n",
    "                    # print(newtables.nodes)\n",
    "                    # print(\"time\", child_time, \"of the child node\", child_node, \"time\", parent_time, \"of the parent node\", parent_node)\n",
    "                    time = np.random.uniform(low=child_time, high=parent_time)\n",
    "                    # print(\"time the mutation has been randomly selected to have\", time)\n",
    "                    newtables.mutations.add_row(site=site_id, node=child_node, time=time, derived_state = \"1\") #this doesn't work as with the above-- perhaps something with the strings. \n",
    "                    # print(tskit.unpack_strings(newtables.sites.ancestral_state, newtables.sites.ancestral_state_offset))\n",
    "        # elif tree_mutations == 0:\n",
    "            # print(\"tree does not contain any mutations\")\\\n",
    "\n",
    "        #expected number of mutations:hudson p. 17, \n",
    "        #average all the sizes of the genealogies of each tree together: \n",
    "        \n",
    "\n",
    "        #     sequence_weights.append(span/sequence_length)\n",
    "        #     genealogy_sizes.append(summed_branch_lengths)\n",
    "        #     weighted_genealogy_sizes.append(summed_branch_lengths*(span/sequence_length)) #multiply the size of the genealogy * the proportion of the genome it occupies \n",
    "\n",
    "        # #find the size of the tree's genealogy, weighted by its length:\n",
    "        # print(\"list of tree sizes of the trees, weighted by their span:\", weighted_genealogy_sizes)\n",
    "        # T_value = sum(weighted_genealogy_sizes)/sum(sequence_weights)\n",
    "        # print(T_value)\n",
    "        #making sure the method above works\n",
    "        # print(\"make sure the above numbers are correct:\", sequence_weights, genealogy_sizes)\n",
    "        # for i, value in enumerate(genealogy_sizes):\n",
    "        #     print(\"Calculation\", genealogy_sizes[i]*sequence_weights[i])\n",
    "        # average_tree_size = sum(weighted_genealogy_sizes)\n",
    "\n",
    "        #calculating a weighted average of genealogies \n",
    "    #     sequence_weights.append(span)\n",
    "    #     genealogy_sizes.append(summed_branch_lengths)\n",
    "    #     weighted_genealogy_sizes.append(summed_branch_lengths*(span)) #multiply the size of the genealogy * the proportion of the genome it occupies \n",
    "\n",
    "    # #find the size of the tree's genealogy, weighted by its length:\n",
    "    # # print(\"list of tree sizes of the trees, weighted by their span:\", weighted_genealogy_sizes)\n",
    "    # T = sum(weighted_genealogy_sizes)/sum(sequence_weights) #t value per sim run\n",
    "    # print(\"t value\", T)\n",
    "    # print(1e-3)\n",
    "    \n",
    "    # # #making sure the method above works\n",
    "    # # print(\"make sure the above numbers are correct:\", sequence_weights, genealogy_sizes)\n",
    "    # # for i, value in enumerate(genealogy_sizes):\n",
    "    # #     print(\"Calculation\", genealogy_sizes[i]*sequence_weights[i])\n",
    "    # # average_tree_size = sum(weighted_genealogy_sizes)\n",
    "\n",
    "    # expected_mean_number_mutations = (T * (4*(population_sequences/ploidy) * (mu))) / 2\n",
    "    # print(\"mutation mean expectation:\", expected_mean_number_mutations)\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    newtables.sort(edge_start=0, mutation_start=0, site_start=0)\n",
    "    new_ts = newtables.tree_sequence()\n",
    "    ####do sfs stats\n",
    "    for tree in new_ts.trees():\n",
    "        for mutation in tree.mutations():\n",
    "            # print(\"looking at mutation on node:\", mutation.node) \n",
    "            count = tree.num_samples(mutation.node)\n",
    "            # print(\"mutation count:\", count)\n",
    "            i_values.append(count) #dumping all i values from multiple sim runs\n",
    "    # print(\"values for i, the number of occurences of a mutation\", i_values)\n",
    "    # print(\"one loop of a sim::\", \"mutation numbers:\", genealogy_mutations, \"occurrences of these mutations\", i_values)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# print(newtables)\n",
    "# print(newtables.mutations)\n",
    "# print(\"number of samples for the runs\", ts.num_samples) #confusing that ts.num_samples is for haploid lineages, while samples in sim_ancestry is for individuals (which may not be haploid, and defaults to diploid). Note that this will be representative of the last loop over which tree sequence we're dealing with\n",
    "#theoretical expectation for time to MRCA in units of 2N generations:\n",
    "expected_tmrca = 2*(1-(1/samples))\n",
    "# print(\"expected Tmrca:\", expected_tmrca)\n",
    "# print(\"mean tmrca, in generations/(ploidy*N):\", mean_tmrca/sim_runs/population_sequences, \"mean tmrca, in raw gens:\", mean_tmrca/sim_runs) #the first value is generations/(ploidy*N),and the second is in generations.\n",
    "#expected time that i lineages should exist: T(i) =  4N/i(i-1) (hudson 2015, #2)\n",
    "#sum i*(T(i)) from i=2 to n to get the total expected branch lengths: (above equation 6, hudson 1990)\n",
    "expected_total_length = 0\n",
    "for i in range(2, samples):\n",
    "    expected_total_length += i*((4*(population_sequences/ploidy)/(i*(i-1))))\n",
    "\n",
    "# print(\"total expected branch lengths\", expected_total_length, \"mean summed branch lengths:\", (sum(mean_summed_branch_lengths)/sim_runs))\n",
    "#now, we can use this to find the expected total number of polymorphisms (hudson 1990 eq1)\n",
    "expected_lambda = mu*expected_total_length\n",
    "# print(\"mean expectation for the number of polymorphisms on the tree\", expected_lambda)\n",
    "\n",
    "\n",
    "#sfs stats\n",
    "theta = 4*(population_sequences/ploidy)*mu\n",
    "# print(\"mutation counts from all loops\", i_values)\n",
    "mean_counts = {}\n",
    "for i in range(1, ((samples*2)-1)):\n",
    "    mean = (i_values.count(i))/sim_runs\n",
    "    mean_counts.update({i:mean})\n",
    "# mean_allele_freqs = (i_values.count(i))/sim_runs\n",
    "# print(\"mean allele frequencies, from all sims binned\", mean_counts)\n",
    "# # print(\"average number of mutations in sims\", sum(mutations_in_run)/sim_runs, \"expected mean number of mutations\", expected_mutations)\n",
    "print(\"expected mean site frequency spectrum\", [(theta/i) for i in range(1, samples)], \"mean allele frequencies from sims, averaged\", mean_counts)\n",
    "ind = np.arange(len(mean_counts))\n",
    "plt.bar(ind, list(mean_counts.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.7912818296162087 1791.2818296162088\n",
      "1.822813472845434 1822.813472845434\n",
      "1.8013986092376217 1801.3986092376217\n",
      "1.797466850685748 1797.466850685748\n",
      "4888.190532188829\n"
     ]
    }
   ],
   "source": [
    "#kevin's script for how to deal with units of generations in msprime\n",
    "import msprime\n",
    "\n",
    "nreps = 10000\n",
    "\n",
    "for ploidy in [1, 2, 5, 10]:\n",
    "    mean = 0\n",
    "    # Here is where it (the interface) gets confusing.\n",
    "    # We simulate 10 lineages.\n",
    "    # But there's a GIANT gotcha:\n",
    "    # population_size is number of INDIVIDUALS and not\n",
    "    # the number of LINEAGES.\n",
    "    # Therefore, we have to divied that by PLOIDY\n",
    "    # to keep everything comparable.\n",
    "    for ts in msprime.sim_ancestry(samples =(10/ploidy), population_size=1000/ploidy,\n",
    "                                   ploidy=ploidy, model=\"hudson\",\n",
    "                                   num_replicates=nreps):\n",
    "        assert ts.num_samples == 10 ##I forget this every time, but I think the point is ts.num_samples will return the number of sampled SEQUENCES, while \n",
    "        last_node = ts.num_nodes - 1\n",
    "        tmrca = ts.node(last_node).time\n",
    "        mean += tmrca\n",
    "\n",
    "    # Now, the first value is generations/(ploidy*N),\n",
    "    # And the second is in generations.\n",
    "    print(mean/nreps/1000, mean/nreps)\n",
    "\n",
    "# Divide out the ploidy\n",
    "ts = msprime.sim_ancestry(samples = 5, population_size=1000/2,\n",
    "                          ploidy=ploidy, model=\"hudson\")\n",
    "last_node = ts.num_nodes - 1\n",
    "tmrca = ts.node(last_node).time\n",
    "\n",
    "# needs tskit 0.5.4!!!!!!!!!!!!!!!!!!!!!!!!!\n",
    "ts.draw_svg(path=\"test.svg\", y_axis=True, size=(\n",
    "    600, 600), canvas_size=(800, 800))\n",
    "print(tmrca) #in units of generations- we don't need to divide by the reps above because it's only one rep! Also shows how wide the variance is on that mean- not crazy to get 11,800 gens for the time when the expectation is 1,800 gens, or 1.8 in units gens/ploidy*N \n",
    "\n",
    "#these should all come up with the theoretical expectation: for ploidy and n\n",
    "# print(2*(1-(1/10)))\n",
    "# print(2*(1-1/10)) #me, seeing how 'smart' python is with orders of operation\n",
    "# print(2*(1-1/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nodes in the current tree [0. 1. 2. 3. 4. 5. 6. 0.]\n",
      "mean number of mutations expected on this tree/poisson lambda 0.4709072046998257 number of mutations drawn from poisson distribution 0\n",
      "fraction of the genome this tree occupies 0.545\n",
      "summed branch lengths: 8.64049916880414\n",
      "interval this tree occupies Interval(left=0.0, right=545.0)\n",
      "LOOP BREAK between trees \n",
      "\n",
      "Expected site frequency spectrum [40.0, 20.0, 13.333333333333334, 10.0, 8.0, 6.666666666666667, 5.714285714285714, 5.0, 4.444444444444445]\n",
      "nodes in the current tree [0. 1. 2. 3. 4. 5. 0. 7.]\n",
      "mean number of mutations expected on this tree/poisson lambda 0.9707031794118754 number of mutations drawn from poisson distribution 2\n",
      "fraction of the genome this tree occupies 0.455\n",
      "summed branch lengths: 21.334135811250007\n",
      "interval this tree occupies Interval(left=545.0, right=1000.0)\n",
      "893\n",
      "912\n",
      "LOOP BREAK between trees \n",
      "\n",
      "Expected site frequency spectrum [40.0, 20.0, 13.333333333333334, 10.0, 8.0, 6.666666666666667, 5.714285714285714, 5.0, 4.444444444444445]\n",
      "loop break between trees\n",
      "TableCollection\n",
      "\n",
      "Sequence Length: 1000.0\n",
      "Time units: generations\n",
      "Metadata: b''\n",
      "\n",
      "Individuals\n",
      "╔══╤═════╤════════╤═══════╤════════╗\n",
      "║id│flags│location│parents│metadata║\n",
      "╠══╪═════╪════════╪═══════╪════════╣\n",
      "║0 │    0│        │       │        ║\n",
      "║1 │    0│        │       │        ║\n",
      "╚══╧═════╧════════╧═══════╧════════╝\n",
      "\n",
      "Nodes\n",
      "╔══╤═════╤══════════╤══════════╤═══════════╤════════╗\n",
      "║id│flags│population│individual│time       │metadata║\n",
      "╠══╪═════╪══════════╪══════════╪═══════════╪════════╣\n",
      "║0 │    1│         0│         0│ 0.00000000│        ║\n",
      "║1 │    1│         0│         0│ 0.00000000│        ║\n",
      "║2 │    1│         0│         1│ 0.00000000│        ║\n",
      "║3 │    1│         0│         1│ 0.00000000│        ║\n",
      "║4 │    0│         0│        -1│ 0.22949513│        ║\n",
      "║5 │    0│         0│        -1│ 0.86085404│        ║\n",
      "║6 │    0│         0│        -1│ 4.86542417│        ║\n",
      "║7 │    0│         0│        -1│11.21224249│        ║\n",
      "╚══╧═════╧══════════╧══════════╧═══════════╧════════╝\n",
      "\n",
      "Edges\n",
      "╔══╤════╤═════╤══════╤═════╤════════╗\n",
      "║id│left│right│parent│child│metadata║\n",
      "╠══╪════╪═════╪══════╪═════╪════════╣\n",
      "║0 │   0│ 1000│     4│    0│        ║\n",
      "║1 │   0│ 1000│     4│    1│        ║\n",
      "║2 │   0│ 1000│     5│    2│        ║\n",
      "║3 │   0│ 1000│     5│    3│        ║\n",
      "║4 │   0│  545│     6│    4│        ║\n",
      "║5 │   0│  545│     6│    5│        ║\n",
      "║6 │ 545│ 1000│     7│    4│        ║\n",
      "║7 │ 545│ 1000│     7│    5│        ║\n",
      "╚══╧════╧═════╧══════╧═════╧════════╝\n",
      "\n",
      "Sites\n",
      "╔══╤════════╤═══════════════╤════════╗\n",
      "║id│position│ancestral_state│metadata║\n",
      "╠══╪════════╪═══════════════╪════════╣\n",
      "║0 │     893│              0│        ║\n",
      "║1 │     912│              0│        ║\n",
      "╚══╧════════╧═══════════════╧════════╝\n",
      "\n",
      "Mutations\n",
      "╔══╤════╤════╤═══════════╤═════════════╤══════╤════════╗\n",
      "║id│site│node│time       │derived_state│parent│metadata║\n",
      "╠══╪════╪════╪═══════════╪═════════════╪══════╪════════╣\n",
      "║0 │ 893│   5│11.21224249│            1│    -1│        ║\n",
      "║1 │ 912│   5│ 0.86085404│            1│    -1│        ║\n",
      "╚══╧════╧════╧═══════════╧═════════════╧══════╧════════╝\n",
      "\n",
      "Migrations\n",
      "╔══╤════╤═════╤════╤══════╤════╤════╤════════╗\n",
      "║id│left│right│node│source│dest│time│metadata║\n",
      "╠══╪════╪═════╪════╪══════╪════╪════╪════════╣\n",
      "╚══╧════╧═════╧════╧══════╧════╧════╧════════╝\n",
      "\n",
      "Populations\n",
      "╔══╤════════════════════════════════════╗\n",
      "║id│metadata                            ║\n",
      "╠══╪════════════════════════════════════╣\n",
      "║0 │{'description': '', 'name': 'pop_0'}║\n",
      "╚══╧════════════════════════════════════╝\n",
      "\n",
      "Provenances\n",
      "╔══╤══════════════════════════╤════════════════════════════════════════════════════════════╗\n",
      "║id│timestamp                 │record                                                      ║\n",
      "╠══╪══════════════════════════╪════════════════════════════════════════════════════════════╣\n",
      "║0 │2023-02-27T16:52:50.550967│{\"schema_version\": \"1.0.0\", \"software\": {\"name\": \"msprime...║\n",
      "╚══╧══════════════════════════╧════════════════════════════════════════════════════════════╝\n",
      "\n",
      "   6   \n",
      " ┏━┻━┓ \n",
      " ┃   5 \n",
      " ┃  ┏┻┓\n",
      " 4  ┃ ┃\n",
      "┏┻┓ ┃ ┃\n",
      "0 1 2 3\n",
      "\n",
      "   7   \n",
      " ┏━┻━┓ \n",
      " ┃   5 \n",
      " ┃  ┏┻┓\n",
      " 4  ┃ ┃\n",
      "┏┻┓ ┃ ┃\n",
      "0 1 2 3\n",
      "\n"
     ]
    },
    {
     "ename": "LibraryError",
     "evalue": "Site out of bounds. (TSK_ERR_SITE_OUT_OF_BOUNDS)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLibraryError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 136\u001b[0m\n\u001b[1;32m    132\u001b[0m \u001b[39mfor\u001b[39;00m t \u001b[39min\u001b[39;00m ts\u001b[39m.\u001b[39mtrees():\n\u001b[1;32m    133\u001b[0m   \u001b[39mprint\u001b[39m(t\u001b[39m.\u001b[39mdraw(\u001b[39mformat\u001b[39m\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39municode\u001b[39m\u001b[39m'\u001b[39m))\n\u001b[0;32m--> 136\u001b[0m newtables\u001b[39m.\u001b[39;49msort()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/tskit/tables.py:3524\u001b[0m, in \u001b[0;36mTableCollection.sort\u001b[0;34m(self, edge_start, site_start, mutation_start)\u001b[0m\n\u001b[1;32m   3468\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39msort\u001b[39m(\u001b[39mself\u001b[39m, edge_start\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m, \u001b[39m*\u001b[39m, site_start\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m, mutation_start\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m):\n\u001b[1;32m   3469\u001b[0m     \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m   3470\u001b[0m \u001b[39m    Sorts the tables in place. This ensures that all tree sequence ordering\u001b[39;00m\n\u001b[1;32m   3471\u001b[0m \u001b[39m    requirements listed in the\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3522\u001b[0m \u001b[39m        (default=0; must be one of [0, len(mutations)]).\u001b[39;00m\n\u001b[1;32m   3523\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 3524\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_ll_tables\u001b[39m.\u001b[39;49msort(edge_start, site_start, mutation_start)\n",
      "\u001b[0;31mLibraryError\u001b[0m: Site out of bounds. (TSK_ERR_SITE_OUT_OF_BOUNDS)"
     ]
    }
   ],
   "source": [
    "#make a standalone tree sequence to learn how to do the mutation placement, make sim predictions, etc\n",
    "import msprime\n",
    "import tskit\n",
    "import numpy as np\n",
    "from numpy.random import default_rng\n",
    "rng = default_rng()\n",
    "import random\n",
    "#simulate one population fo now\n",
    "mu = (1e-1) #normally would have this set to 1e-7 *2, but putting deme size down and #individuals up to be able to visualize tree sequences for working on placing mutations issue.\n",
    "rho = (1e-4)*2\n",
    "bases = 1000\n",
    "deme_size = 100\n",
    "ts = msprime.sim_ancestry(recombination_rate=rho, sequence_length = bases, samples=2, random_seed=1234, discrete_genome = True)\n",
    "newtables = ts.dump_tables() #modifying a copy of the original tables in order to be able to overwrite \n",
    "#make a new table collection to hold everything\n",
    "for tree in ts.trees():\n",
    "    nodes_list = np.zeros(tree.tree_sequence.num_nodes) #make a np array for each node to track the branch lengths above it (needed for calculations around placing mutations)\n",
    "    branch_lengths = np.zeros(tree.tree_sequence.num_nodes) #need two np arrays initiated with the same number- one for branch lengths, and one to keep track of the nodes\n",
    "    summed_branch_lengths = 0\n",
    "    interval = tree.interval #find the interval the tree occupies in the genome \n",
    "    span = tree.interval.right-tree.interval.left #find the fraction of the total genome/number of bases\n",
    "    \n",
    "    for node in tree.nodes(): #type error: tree object is not iterable-- this happens w/ you frequently b/c you're forgetting to include the parenthesis to call the method properly, so the code tries to iterate over the method and not the data structure it's referencing \n",
    "        branch_lengths[node] =+ tree.branch_length(node)  #https://stackoverflow.com/questions/43505921/python-trying-to-create-a-dictionary-through-a-for-loop\n",
    "        nodes_list[node] =+ node\n",
    "        child_time = newtables.nodes.time[node]\n",
    "\n",
    "        for index, child_node in enumerate(newtables.edges.child):\n",
    "            if child_node == node:\n",
    "                # print(\"index of edgetable entry\", index, \"child node\", child_node)\n",
    "                # print(\"index of edgetable entry\", index, \"child node\", child_node)\n",
    "                entry = newtables.edges[index]\n",
    "                # print(\"entry that matches a node in this tree\", entry)\n",
    "                # print(\"line of edgetable accessed\", index, \"child node for this loop\", child_node, \"edgetable entry\", entry)\n",
    "                current_tree_interval = (interval.left, interval.right)\n",
    "                if (entry.left, entry.right) == current_tree_interval: \n",
    "                    parent_node = entry.parent\n",
    "                    parent_time = newtables.nodes.time[parent_node]\n",
    "                    branch_length = (parent_time - child_time)    \n",
    "                    summed_branch_lengths = summed_branch_lengths + branch_length #hudson 1990, above equation 1\n",
    "                    ####place mutations on nodes proportional to branch lengths, and proportional to how long of a span of the genome a given tree occupies\n",
    "                    lambda_lineage = mu *branch_length *(span/bases) #my idea to populate the sim with the expected number of mutations given the parameters I'm using: get mu * t in here, after the hudson 2015 paper having lineage mu * branch length be the number of mutations occuring along that lineage. Summing them all up to get expectations for each tree. I'm multiplying by the fraction of the genome, since I believe mu* branch length as the expectation for the # of mutations assumes the whole genome\n",
    "\n",
    "    print(\"nodes in the current tree\", nodes_list)\n",
    "    lambda_total = mu*summed_branch_lengths*(span/bases)\n",
    "    genealogy_mutations = rng.poisson(lam=lambda_total)\n",
    "    print(\"mean number of mutations expected on this tree/poisson lambda\", lambda_total, \"number of mutations drawn from poisson distribution\", genealogy_mutations)\n",
    "    print(\"fraction of the genome this tree occupies\", span/bases)\n",
    "    print(\"summed branch lengths:\", summed_branch_lengths)\n",
    "    proportional_branch_lengths = branch_lengths/summed_branch_lengths \n",
    "    p = np.asarray(proportional_branch_lengths).astype('float64')\n",
    "    p = p/np.sum(p) #not sure if doing this as a workaround to the proportional branch lengths array not summing to 0 is the best idea, but to get it working for now\n",
    "    # print(\"probabilities\",p, \"sums\", sum(p))\n",
    "\n",
    "    #######temporarily hardcoding in some mutations to test the below, changed size = int(number_mutations) to size = 3\n",
    "    nodes_sampled =  rng.choice(nodes_list, p=p, size = genealogy_mutations) #this can sample the same node twice, which IS what you want. however need to figure out a way to determine which parent node was associated with a given branch length, so you can select a site that makes sense given the edges possible for a child:parent node relationship. int is a pretty crude \n",
    "    # nodes_sampled =  np.random.choice(nodes_list, p=p, size = np.random.choice(expected mutations))\n",
    "\n",
    "    nodes_sampled = nodes_sampled.astype(int) ##these need to be integers to be able to use them to access indices/values in the edge table. See if you can make it such when the  variable is created \n",
    "    # print(\"nodes sampled from distribution\", nodes_sampled)\n",
    "    #pick out the spot in the genome that the tree we have iterated over actually occupies (with edges) \n",
    "    print(\"interval this tree occupies\", tree.interval) #access this to pick which sites are possible\n",
    "    # print(\"left side of interval\", interval.left)\n",
    "    for node in nodes_sampled: #since I do a pretty much identical thing above, there's gotta be a way to combine the two, but I'll leave the stunts for later \n",
    "        ######draw a site from the genome that HASN'T ALREADY had a mutation assigned to it. \n",
    "        #use these nodes to write mutations to the mutations table\n",
    "        possible_bases = set(range((int(interval.left)), int(interval.right))) #set them to integers for now? need to figure out how to actually implement using sets \n",
    "        bases_already_mutated = set(int(i) for i in newtables.sites.position)\n",
    "        unmutated_sites =(possible_bases.difference(bases_already_mutated)) #get the difference between two sets- it returns only those bases NOT in bases_already_mutated\n",
    "        # print(\"checking if unmutated sites are constrained to those within the tree\", unmutated_sites)\n",
    "        # print(\"set difference test\", x)\n",
    "        # position = np.random.choice(tuple(unmutated_sites))#suggestion from one stackoverflow entry for this(https://stackoverflow.com/questions/15837729/random-choice-from-set-python) links to another to try to make this process constant time???: https://stackoverflow.com/questions/15993447/python-data-structure-for-efficient-add-remove-and-random-choice\n",
    "        # print(position)\n",
    "        ######add binary mutations in- 0/1, whether it's been mutated or not. Or could use just two nucleotides, for instance\n",
    "        # print(\"SITES THAT DO NOT YET HAVE A MUTATION\", unmutated_sites)\n",
    "        ###suggestion for how to make it faster\n",
    "        position = np.random.randint(low=(interval.left), high=(int(interval.right)), size=1)\n",
    "        position = position[0]\n",
    "        print(position)\n",
    "        if position in unmutated_sites: #size should be 1, since we're only placing one mutation above each loop over a node\n",
    "            newtables.sites.add_row(position=position, ancestral_state=\"0\")\n",
    "\n",
    "        #####get times for the mutations- randomly select them from the interval between parent and child node times \n",
    "        #get child node time\n",
    "        child_time = newtables.nodes.time[node]\n",
    "        # print(\"node ID\", node,  ts.tables.edges.child[node])#https://stackoverflow.com/questions/176918/finding-the-index-of-an-item-in-a-\n",
    "        \n",
    "        for index, child_node in enumerate(newtables.edges.child): #I think the child node is the appropriate one to pick, since we chose nodes BELOW branches. Mutations are associated with the node below them (according to tskit data model page)\n",
    "            # print(index, j) #this print statement coupled with the one below lets you check if it's behaving as intended\n",
    "            if child_node == node:\n",
    "                entry = newtables.edges[index]\n",
    "                # print(\"node\", node, \"edges associated with the node selected\", entry)\n",
    "                current_tree_interval = (interval.left, interval.right)\n",
    "                if (entry.left, entry.right) == current_tree_interval: \n",
    "                    parent_node = entry.parent\n",
    "                    # print(\"tuple matched, parent node\", parent_node)\n",
    "                elif (entry.left, entry.right) == (0, bases): #this happens when the node is not associated with a recombination event \n",
    "                    # print(\"this node has the same parent regardless of tree\") (#do I need to do this above??)\n",
    "                    parent_node=entry.parent\n",
    "\n",
    "        #now, find actual parent time:\n",
    "        parent_time = newtables.nodes.time[parent_node]\n",
    "        # print(\"parent time\", parent_time)\n",
    "        time_span = (child_time, parent_time)\n",
    "        # print(\"time interval between a node and its parent\", time_span)\n",
    "        time=random.choice(time_span) #you get a suspiciously high amount of '0' times. \n",
    "\n",
    "        #####assign all variables to the mutations table \n",
    "        newtables.mutations.add_row(site=position, node=node, time=time, derived_state = \"1\") #this doesn't work as with the above-- perhaps something with the strings. \n",
    "        # print(tskit.unpack_strings(newtables.sites.ancestral_state, newtables.sites.ancestral_state_offset))\n",
    "\n",
    "        # print(np.setdiff1d(nucleotides, newtables.sites.ancestral_state))\n",
    "    print(\"LOOP BREAK between trees\", \"\\n\")\n",
    "\n",
    "\n",
    "    expected_sfs = []\n",
    "    theta = 4*(deme_size)*(mu)\n",
    "    for i in range(1,10):\n",
    "        expected_sfs.append(theta/i)\n",
    "    print(\"Expected site frequency spectrum\", expected_sfs)\n",
    "\n",
    "# mutations expected for whole genome: \n",
    "\n",
    "\n",
    "\n",
    "print(\"loop break between trees\")\n",
    "print(newtables)\n",
    "# simulated_sfs = []\n",
    "# a = ts.allele_frequency_spectrum(sample_sets=[ts.samples(population=1), ts.samples(population=2)], windows=None, mode='site', span_normalise=True, polarised=False)\n",
    "# print(\"allele frequency spectrum function\", a)\n",
    "\n",
    "for t in ts.trees():\n",
    "  print(t.draw(format='unicode'))\n",
    "\n",
    "\n",
    "newtables.sort()\n",
    "\n",
    "# new_ts = newtables.tree_sequence()\n",
    "# for t in new_ts.trees():\n",
    "#   print(t.draw(format='unicode'))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
