{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   6   \n",
      " ┏━┻━┓ \n",
      " 5   ┃ \n",
      "┏┻┓  ┃ \n",
      "┃ ┃  4 \n",
      "┃ ┃ ┏┻┓\n",
      "0 1 2 3\n",
      "\n",
      "   6   \n",
      " ┏━┻━┓ \n",
      " ┃   5 \n",
      " ┃  ┏┻┓\n",
      " 4  ┃ ┃\n",
      "┏┻┓ ┃ ┃\n",
      "0 1 2 3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#conda install ipykernel -y put this in the terminal to fix the module missing error \n",
    "\n",
    "#be able to draw a tree sequence just via editing the tables\n",
    "import tskit\n",
    "tables = tskit.TableCollection(sequence_length=1.0)\n",
    "n = tables.nodes\n",
    "#nodes\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0)\n",
    "n.add_row(flags=1, population=-1, time=0.4)\n",
    "n.add_row(flags=1, population=-1, time=0.7)\n",
    "n.add_row(flags=1, population=-1, time=1)\n",
    "\n",
    "# to make a straightforward tree: \n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=4, child=2) \n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=4, child=3)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=5, child=0)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=5, child=1)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=6, child=4)\n",
    "# tables.edges.add_row(left=0.0, right=1.0, parent=6, child=5)\n",
    "\n",
    "#adding in something where the edges actually matter: 1/2 of the genome has one genealogy, 1/2 has another. It doesn't make a lot of sense re: biology but the concept of different chunks having different ancestry I figured out here \n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=4, child=2)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=4, child=3)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=5, child=0)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=5, child=1)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=6, child=4)\n",
    "tables.edges.add_row(left=0.0, right=0.5, parent=6, child=5)\n",
    "\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=4, child=0) \n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=4, child=1)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=5, child=2)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=5, child=3)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=6, child=4)\n",
    "tables.edges.add_row(left=0.5, right=1.0, parent=6, child=5)\n",
    "\n",
    "#sort step, to ensure that your tables are ordered correctly\n",
    "tskit.TableCollection.sort(edge_start=0, self=tables, site_start=0, mutation_start=0) #in tskit docs, the self=tables portion is just *, so here I've assigned self the the variable name for all tables\n",
    "\n",
    "#tree sequence that prints in terminal and you don't have to deal w/ vscode and svgs\n",
    "ts = tables.tree_sequence()\n",
    "for t in ts.trees():\n",
    "  print(t.draw(format='unicode'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of samples for the runs 10\n",
      "expected Tmrca: 1.8\n",
      "mean tmrca, in generations/(ploidy*N): 1.6747558418648067 mean tmrca, in raw gens: 1674.7558418648068\n",
      "total expected branch lengths 5435.714285714284 mean summed branch lengths: 5389.228475067943\n",
      "mean expectation for the number of polymorphisms on the tree 5.435714285714284 mean lambda value, from lambdas used for sampling 5.3892284750679424 number of average polymorphisms actually on the trees 5.42\n",
      "mean allele frequencies, from all sims binned {1: 2.29, 2: 0.74, 3: 0.59, 4: 0.46, 5: 0.23, 6: 0.34, 7: 0.26, 8: 0.21, 9: 0.3}\n",
      "expected mean site frequency spectrum [2.0, 1.0, 0.6666666666666666, 0.5, 0.4, 0.3333333333333333, 0.2857142857142857, 0.25, 0.2222222222222222] mean allele frequencies from sims, averaged {1: 2.29, 2: 0.74, 3: 0.59, 4: 0.46, 5: 0.23, 6: 0.34, 7: 0.26, 8: 0.21, 9: 0.3}\n",
      "       18          \n",
      "  ┏━━━━━┻━━━━━┓    \n",
      "  ┃          17    \n",
      "  ┃         ┏━┻━━┓ \n",
      " 16         ┃    ┃ \n",
      "┏━┻┓        ┃    ┃ \n",
      "┃  ┃       15    ┃ \n",
      "┃  ┃      ┏━┻━┓  ┃ \n",
      "┃  ┃     14   ┃  ┃ \n",
      "┃  ┃    ┏━┻━┓ ┃  ┃ \n",
      "┃  ┃    ┃   ┃ ┃ 13 \n",
      "┃  ┃    ┃   ┃ ┃ ┏┻┓\n",
      "┃  ┃   12   ┃ ┃ ┃ ┃\n",
      "┃  ┃   ┏┻━┓ ┃ ┃ ┃ ┃\n",
      "┃  ┃  11  ┃ ┃ ┃ ┃ ┃\n",
      "┃  ┃  ┏┻┓ ┃ ┃ ┃ ┃ ┃\n",
      "┃ 10  ┃ ┃ ┃ ┃ ┃ ┃ ┃\n",
      "┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃ ┃ ┃\n",
      "0 1 9 2 5 4 8 3 6 7\n",
      "\n",
      "allele frequency spectrum [0. 1. 2. 1. 0. 0. 0. 0. 0. 0. 0.] total branch lengths above each bin [     0.         186027.85432088  72326.17447268  36819.99367468\n",
      "   4027.59265659  18246.98619346      0.          18216.31450388\n",
      "      0.              0.              0.        ]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiMAAAGdCAYAAADAAnMpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAYL0lEQVR4nO3df6zVdf3A8dcF5V5s3OuvuIBehLKFKPJT8MImukhi5GJrZs4GI3VrgwJvs4GVbJleraE0RZHMXBkDzcRSs+g6IOQ6BLxNKnXOHxByL7jsXrnVxXHP94++3boJwsGLLy738dg+f9zP/XzOeX12YPe59/2ce0oKhUIhAACS9MoeAADo2cQIAJBKjAAAqcQIAJBKjAAAqcQIAJBKjAAAqcQIAJDqhOwBDkd7e3u8+eab0a9fvygpKckeBwA4DIVCId55550YNGhQ9Op18PWPbhEjb775ZlRVVWWPAQAcgR07dsSZZ5550O93ixjp169fRPzrYsrLy5OnAQAOR0tLS1RVVXX8HD+YbhEj//7VTHl5uRgBgG7mULdYuIEVAEglRgCAVGIEAEglRgCAVGIEAEglRgCAVGIEAEglRgCAVGIEAEglRgCAVGIEAEglRgCAVGIEAEglRgCAVCdkD5BtyIInskc4pNdvnZ49AgAcNVZGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASCVGAIBUYgQASFVUjNTW1sYFF1wQ/fr1i/79+8eMGTPipZdeOuR5Dz/8cAwbNizKyspixIgR8eSTTx7xwADA8aWoGFm3bl3MmTMnnn322VizZk28++67cemll0Zra+tBz9m4cWNceeWVcfXVV8fzzz8fM2bMiBkzZsS2bds+8PAAQPdXUigUCkd68p49e6J///6xbt26uOiiiw54zBVXXBGtra3x+OOPd+y78MILY9SoUbFs2bLDep6WlpaoqKiI5ubmKC8vP9JxD2jIgie69PGOhtdvnZ49AgAU7XB/fn+ge0aam5sjIuLUU0896DH19fUxZcqUTvumTp0a9fX1Bz2nra0tWlpaOm0AwPHpiGOkvb095s+fH5MmTYrzzjvvoMc1NjZGZWVlp32VlZXR2Nh40HNqa2ujoqKiY6uqqjrSMQGAY9wRx8icOXNi27ZtsXLlyq6cJyIiFi5cGM3NzR3bjh07uvw5AIBjwwlHctLcuXPj8ccfj/Xr18eZZ575vscOGDAgmpqaOu1ramqKAQMGHPSc0tLSKC0tPZLRAIBupqiVkUKhEHPnzo1HH300nn766Rg6dOghz6muro66urpO+9asWRPV1dXFTQoAHJeKWhmZM2dOrFixIh577LHo169fx30fFRUV0bdv34iImDlzZpxxxhlRW1sbERHz5s2LyZMnx+LFi2P69OmxcuXK2Lx5cyxfvryLLwUA6I6KWhm55557orm5OS6++OIYOHBgx7Zq1aqOY7Zv3x67du3q+HrixImxYsWKWL58eYwcOTJ+/vOfx+rVq9/3plcAoOcoamXkcP4kydq1a9+z7/LLL4/LL7+8mKcCAHoIn00DAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAKjECAKQSIwBAqqJjZP369XHZZZfFoEGDoqSkJFavXv2+x69duzZKSkreszU2Nh7pzADAcaToGGltbY2RI0fG0qVLizrvpZdeil27dnVs/fv3L/apAYDj0AnFnjBt2rSYNm1a0U/Uv3//OPnkk4s+DwA4vn1o94yMGjUqBg4cGJ/+9KfjmWeeed9j29raoqWlpdMGAByfjnqMDBw4MJYtWxaPPPJIPPLII1FVVRUXX3xxbN269aDn1NbWRkVFRcdWVVV1tMcEAJKUFAqFwhGfXFISjz76aMyYMaOo8yZPnhyDBw+On/70pwf8fltbW7S1tXV83dLSElVVVdHc3Bzl5eVHOu4BDVnwRJc+3tHw+q3Ts0cAgKK1tLRERUXFIX9+F33PSFcYP358bNiw4aDfLy0tjdLS0g9xIgAgS8rfGWloaIiBAwdmPDUAcIwpemVk79698corr3R8/dprr0VDQ0OceuqpMXjw4Fi4cGHs3LkzfvKTn0RExJIlS2Lo0KFx7rnnxj//+c+477774umnn47f/va3XXcVAEC3VXSMbN68OS655JKOr2tqaiIiYtasWfHAAw/Erl27Yvv27R3f37dvX3z961+PnTt3xkknnRTnn39+/O53v+v0GABAz/WBbmD9sBzuDTBHwg2sAHB0HO7Pb59NAwCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkKjpG1q9fH5dddlkMGjQoSkpKYvXq1Yc8Z+3atTFmzJgoLS2Ns88+Ox544IEjGBUAOB4VHSOtra0xcuTIWLp06WEd/9prr8X06dPjkksuiYaGhpg/f35cc8018Zvf/KboYQGA488JxZ4wbdq0mDZt2mEfv2zZshg6dGgsXrw4IiLOOeec2LBhQ9xxxx0xderUYp8eADjOHPV7Rurr62PKlCmd9k2dOjXq6+uP9lMDAN1A0SsjxWpsbIzKyspO+yorK6OlpSX+8Y9/RN++fd9zTltbW7S1tXV83dLScrTHBACSHJPvpqmtrY2KioqOraqqKnskAOAoOeoxMmDAgGhqauq0r6mpKcrLyw+4KhIRsXDhwmhubu7YduzYcbTHBACSHPVf01RXV8eTTz7Zad+aNWuiurr6oOeUlpZGaWnp0R4NADgGFL0ysnfv3mhoaIiGhoaI+NdbdxsaGmL79u0R8a9VjZkzZ3Yc/5WvfCVeffXV+MY3vhEvvvhi3H333fHQQw/Fdddd1zVXAAB0a0XHyObNm2P06NExevToiIioqamJ0aNHx4033hgREbt27eoIk4iIoUOHxhNPPBFr1qyJkSNHxuLFi+O+++7ztl4AICIiSgqFQiF7iENpaWmJioqKaG5ujvLy8i597CELnujSxzsaXr91evYIAFC0w/35fUy+mwYA6DnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQ6oTsAehaQxY8kT3C+3r91unZIwBwjLEyAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQCoxAgCkEiMAQKoTsgeAgxmy4InsEQ7p9VunZ48A0O1ZGQEAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUokRACDVEcXI0qVLY8iQIVFWVhYTJkyITZs2HfTYBx54IEpKSjptZWVlRzwwAHB8KTpGVq1aFTU1NbFo0aLYunVrjBw5MqZOnRq7d+8+6Dnl5eWxa9euju2NN974QEMDAMePomPk9ttvj2uvvTZmz54dw4cPj2XLlsVJJ50U999//0HPKSkpiQEDBnRslZWVH2hoAOD4UVSM7Nu3L7Zs2RJTpkz5zwP06hVTpkyJ+vr6g563d+/eOOuss6Kqqio+97nPxR//+Mf3fZ62trZoaWnptAEAx6eiYuStt96K/fv3v2dlo7KyMhobGw94zic/+cm4//7747HHHosHH3ww2tvbY+LEifGXv/zloM9TW1sbFRUVHVtVVVUxYwIA3chRfzdNdXV1zJw5M0aNGhWTJ0+OX/ziF/HRj3407r333oOes3Dhwmhubu7YduzYcbTHBACSnFDMwaeffnr07t07mpqaOu1vamqKAQMGHNZjnHjiiTF69Oh45ZVXDnpMaWlplJaWFjMaANBNFbUy0qdPnxg7dmzU1dV17Gtvb4+6urqorq4+rMfYv39/vPDCCzFw4MDiJgUAjktFrYxERNTU1MSsWbNi3LhxMX78+FiyZEm0trbG7NmzIyJi5syZccYZZ0RtbW1ERHznO9+JCy+8MM4+++z429/+Ft///vfjjTfeiGuuuaZrrwQA6JaKjpErrrgi9uzZEzfeeGM0NjbGqFGj4qmnnuq4qXX79u3Rq9d/FlzefvvtuPbaa6OxsTFOOeWUGDt2bGzcuDGGDx/edVcBAHRbRcdIRMTcuXNj7ty5B/ze2rVrO319xx13xB133HEkTwMA9ABHFCNAcYYseCJ7hEN6/dbp2SMAPZQPygMAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUokRACCVGAEAUp2QPQBAliELnsge4ZBev3V69ghw1FkZAQBSiREAIJUYAQBSiREAIJUYAQBSeTcNABxF3rV1aFZGAIBUYgQASCVGAIBUYgQASCVGAIBU3k0DcJzwrg26KysjAEAqMQIApBIjAEAqMQIApHIDK1AUN0kCXc3KCACQSowAAKn8mgaAY45fB/YsVkYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRiBABIJUYAgFRHFCNLly6NIUOGRFlZWUyYMCE2bdr0vsc//PDDMWzYsCgrK4sRI0bEk08+eUTDAgDHn6JjZNWqVVFTUxOLFi2KrVu3xsiRI2Pq1Kmxe/fuAx6/cePGuPLKK+Pqq6+O559/PmbMmBEzZsyIbdu2feDhAYDur+gYuf322+Paa6+N2bNnx/Dhw2PZsmVx0kknxf3333/A43/wgx/EZz7zmbj++uvjnHPOiZtuuinGjBkTd9111wceHgDo/k4o5uB9+/bFli1bYuHChR37evXqFVOmTIn6+voDnlNfXx81NTWd9k2dOjVWr1590Odpa2uLtra2jq+bm5sjIqKlpaWYcQ9Le9vfu/wxu1ox132sX8/xdC0Rh389ruXD1ROvJeL4uh7X8uE6Gj9f//txC4XC+x9YKMLOnTsLEVHYuHFjp/3XX399Yfz48Qc858QTTyysWLGi076lS5cW+vfvf9DnWbRoUSEibDabzWazHQfbjh073rcviloZ+bAsXLiw02pKe3t7/PWvf43TTjstSkpKEic7tJaWlqiqqoodO3ZEeXl59jj8P6/Lsctrc2zyuhy7utNrUygU4p133olBgwa973FFxcjpp58evXv3jqampk77m5qaYsCAAQc8Z8CAAUUdHxFRWloapaWlnfadfPLJxYyarry8/Jj/R9ITeV2OXV6bY5PX5djVXV6bioqKQx5T1A2sffr0ibFjx0ZdXV3Hvvb29qirq4vq6uoDnlNdXd3p+IiINWvWHPR4AKBnKfrXNDU1NTFr1qwYN25cjB8/PpYsWRKtra0xe/bsiIiYOXNmnHHGGVFbWxsREfPmzYvJkyfH4sWLY/r06bFy5crYvHlzLF++vGuvBADoloqOkSuuuCL27NkTN954YzQ2NsaoUaPiqaeeisrKyoiI2L59e/Tq9Z8Fl4kTJ8aKFSviW9/6Vtxwww3xiU98IlavXh3nnXde113FMaS0tDQWLVr0nl8zkcvrcuzy2hybvC7HruPxtSkpFA71fhsAgKPHZ9MAAKnECACQSowAAKnECACQSox0oaVLl8aQIUOirKwsJkyYEJs2bcoeqcerra2NCy64IPr16xf9+/ePGTNmxEsvvZQ9Fv/j1ltvjZKSkpg/f372KETEzp0740tf+lKcdtpp0bdv3xgxYkRs3rw5e6webf/+/fHtb387hg4dGn379o2Pf/zjcdNNNx36M1+6CTHSRVatWhU1NTWxaNGi2Lp1a4wcOTKmTp0au3fvzh6tR1u3bl3MmTMnnn322VizZk28++67cemll0Zra2v2aPy/5557Lu699944//zzs0chIt5+++2YNGlSnHjiifHrX/86/vSnP8XixYvjlFNOyR6tR7vtttvinnvuibvuuiv+/Oc/x2233Rbf+9734s4778werUt4a28XmTBhQlxwwQVx1113RcS//jJtVVVVfPWrX40FCxYkT8e/7dmzJ/r37x/r1q2Liy66KHucHm/v3r0xZsyYuPvuu+O73/1ujBo1KpYsWZI9Vo+2YMGCeOaZZ+L3v/999ij8l89+9rNRWVkZP/rRjzr2ff7zn4++ffvGgw8+mDhZ17Ay0gX27dsXW7ZsiSlTpnTs69WrV0yZMiXq6+sTJ+N/NTc3R0TEqaeemjwJERFz5syJ6dOnd/q/Q65f/vKXMW7cuLj88sujf//+MXr06PjhD3+YPVaPN3HixKirq4uXX345IiL+8Ic/xIYNG2LatGnJk3WNY/JTe7ubt956K/bv39/xV2j/rbKyMl588cWkqfhf7e3tMX/+/Jg0adJx+xeAu5OVK1fG1q1b47nnnssehf/y6quvxj333BM1NTVxww03xHPPPRdf+9rXok+fPjFr1qzs8XqsBQsWREtLSwwbNix69+4d+/fvj5tvvjmuuuqq7NG6hBihx5gzZ05s27YtNmzYkD1Kj7djx46YN29erFmzJsrKyrLH4b+0t7fHuHHj4pZbbomIiNGjR8e2bdti2bJlYiTRQw89FD/72c9ixYoVce6550ZDQ0PMnz8/Bg0adFy8LmKkC5x++unRu3fvaGpq6rS/qakpBgwYkDQV/23u3Lnx+OOPx/r16+PMM8/MHqfH27JlS+zevTvGjBnTsW///v2xfv36uOuuu6KtrS169+6dOGHPNXDgwBg+fHinfeecc0488sgjSRMREXH99dfHggUL4otf/GJERIwYMSLeeOONqK2tPS5ixD0jXaBPnz4xduzYqKur69jX3t4edXV1UV1dnTgZhUIh5s6dG48++mg8/fTTMXTo0OyRiIhPfepT8cILL0RDQ0PHNm7cuLjqqquioaFBiCSaNGnSe97+/vLLL8dZZ52VNBEREX//+987fQhtRETv3r2jvb09aaKuZWWki9TU1MSsWbNi3LhxMX78+FiyZEm0trbG7Nmzs0fr0ebMmRMrVqyIxx57LPr16xeNjY0REVFRURF9+/ZNnq7n6tev33vu2/nIRz4Sp512mvt5kl133XUxceLEuOWWW+ILX/hCbNq0KZYvXx7Lly/PHq1Hu+yyy+Lmm2+OwYMHx7nnnhvPP/983H777fHlL385e7SuUaDL3HnnnYXBgwcX+vTpUxg/fnzh2WefzR6px4uIA24//vGPs0fjf0yePLkwb9687DEoFAq/+tWvCuedd16htLS0MGzYsMLy5cuzR+rxWlpaCvPmzSsMHjy4UFZWVvjYxz5W+OY3v1loa2vLHq1L+DsjAEAq94wAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQSowAAKnECACQ6v8A/DxtchhQIjMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#make a no-recombination sim to get the stats right \n",
    "import msprime\n",
    "import tskit \n",
    "import numpy as np\n",
    "from numpy.random import default_rng\n",
    "rng = default_rng()\n",
    "import random\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sequence_length = 100\n",
    "mu = (1e-3) #normally would have this set to 1e-7 *2, but putting deme size down and #individuals up to be able to visualize tree sequences for working on placing mutations issue.\n",
    "# sequence_length = 100\n",
    "ploidy = 2\n",
    "population_sequences = 1000 #total number of genomes in the population- the number of individuals is this/ploidy\n",
    "sampled_sequences = 10\n",
    "sim_runs = 100\n",
    "#initialize the bottom two variables, to add/append to later \n",
    "mean_tmrca = 0\n",
    "i_values = []\n",
    "mean_summed_branch_lengths = []\n",
    "total_mut_count = []\n",
    "mut_expectations_given_branchlengths = []\n",
    "for ts in msprime.sim_ancestry(ploidy= ploidy, samples=(sampled_sequences/ploidy), discrete_genome = True, population_size=(population_sequences)/ploidy, num_replicates = sim_runs, sequence_length = sequence_length): #the sim needs everything in terms of the number of individuals, but it's convenient for downstream/interpretation to have our original variables be in terms of the number of lineages or individual sequences, aka in haploid terms \n",
    "    assert ts.num_samples == sampled_sequences\n",
    "    last_node = ts.num_nodes - 1\n",
    "    tmrca = ts.node(last_node).time\n",
    "    mean_tmrca += tmrca\n",
    "    # print(\"tmrca for this run:\", tmrca)\n",
    "    newtables = ts.dump_tables() #modifying a copy of the original tables in order to be able to overwrite \n",
    "    #make a new table collection to hold everything\n",
    "    average_genealogy_size = []\n",
    "    for tree in ts.trees():\n",
    "        # print(tree.draw(format='unicode'))\n",
    "        interval = tree.interval #find the interval the tree occupies in the genome \n",
    "        # print(newtables.edges.child)\n",
    "        nodes_and_branches = {} #empty dict to hold the pair of node:branch length to its parent\n",
    "        for child_node in newtables.edges.child:\n",
    "            nodes_and_branches.update({child_node:(tree.branch_length(child_node))}) \n",
    "        # print(nodes_and_branches)\n",
    "        # print(nodes_and_branches.keys())\n",
    "        summed_branch_lengths = sum(nodes_and_branches.values()) #hudson 1990, above equation 1. I could have just used the tskit function but I was going to loop over the nodes to make a dict anwyay \n",
    "        mean_summed_branch_lengths.append(summed_branch_lengths)\n",
    "        proportional_branch_lengths = {} #empty dict to hold the pair of node:scaled branch length\n",
    "        # print(\"sum of the lengths of all branches:\", summed_branch_lengths, \"what the tskit function gives you\", sum(tree.branch_length(u) for u in tree.nodes()))\n",
    "        # print(\"list of nodes in the tree that can have muts placed above them:\", nodes_and_branches.keys()))\n",
    "        for child_node in newtables.edges.child: #I probably can work around this and not have another loop by accessing the dictionary keys, but it doesn't seem worth the effort\n",
    "            proportional_branch_lengths.update({child_node:(tree.branch_length(child_node)/summed_branch_lengths)}) #hudson 1990 equation 1\n",
    "        # proportional_branch_lengths = dict(sorted(proportional_branch_lengths.items())) #sort the dictionary so the keys(representing nodes) go up in order. I thought this would matter for the np array step, but this may not help once we have >1 tree, as the incies of the np array will no longer match the dict\n",
    "        # print(\"nodes and the branch lengths above them, proportional to the total length of all branches:\" , proportional_branch_lengths) \n",
    "        #hudson 1990, equation 1: compare to the expectation of s above        \n",
    "        lambda_total = mu*summed_branch_lengths #this is for what happens in our sim, but not the expectation\n",
    "        genealogy_mutations = rng.poisson(lam=lambda_total)\n",
    "        # print(\"number of mutations we sampled from poisson distribution:\", genealogy_mutations, \"mean number of mutations in the poisson distribution expectations for this tree\", lambda_total)\n",
    "        # print(\"proportional branch lengths:\", proportional_branch_lengths)\n",
    "        p = list(proportional_branch_lengths.values())\n",
    "        if genealogy_mutations != 0:\n",
    "            # print(\"genealogy contains more than 0 mutations\")\n",
    "            ###########sample nodes from your list of child nodes \n",
    "            nodes_sampled =  rng.choice(list(nodes_and_branches.keys()), p=p, size = genealogy_mutations) #this can sample the same node twice, which IS what you want. however need to figure out a way to determine which parent node was associated with a given branch length, so you can select a site that makes sense given the edges possible for a child:parent node relationship. int is a pretty crude \n",
    "            # print(\"nodes sampled from distribution, aka chosen to have mutations\", nodes_sampled)\n",
    "            ###########pick out the spot in the genome that the tree we have iterated over actually occupies (with edges) \n",
    "            # print(\"interval this tree occupies\", tree.interval) #access this to pick which sites are possible\n",
    "            for child_node in nodes_sampled: #since I do a pretty much identical thing above, there's gotta be a way to combine the two, but I'll leave the stunts for later \n",
    "                ###########draw a site from the genome that HASN'T ALREADY had a mutation assigned to it.\n",
    "                # print(\"node mutation is being added onto:\", child_node) \n",
    "                possible_bases = set(range((int(interval.left)), int(interval.right))) #set them to integers for now? need to figure out how to actually implement using sets \n",
    "                bases_already_mutated = set(int(i) for i in newtables.sites.position)\n",
    "                unmutated_sites =(possible_bases.difference(bases_already_mutated)) #get the \n",
    "                position = np.random.randint(low=(interval.left), high=(int(interval.right)), size=1)\n",
    "                position = position[0]\n",
    "                # print(\"position of new mut:\", position)\n",
    "                ############use these nodes to write mutations to the mutations table\n",
    "                # print(\"position\", position)\n",
    "                if position in unmutated_sites: #size should be 1, since we're only placing one mutation above each loop over a node\n",
    "                    site_id = newtables.sites.add_row(position=position, ancestral_state=\"0\")\n",
    "                    # print(\"id of the site we are adding\", site_id)\n",
    "                    #####get times for the mutations- randomly select them from the interval between parent and child node times \n",
    "                    #get child node time\n",
    "                    child_time = newtables.nodes.time[child_node]\n",
    "                    # print(\"node ID\", child_node,  ts.tables.edges.child[child_node])#https://stackoverflow.com/questions/176918/finding-the-index-of-an-item-in-a-\n",
    "                    #now, find the parent node that goes with the child node: \n",
    "                    index = np.where(newtables.edges.child==child_node)[0][0] #https://www.geeksforgeeks.org/how-to-find-the-index-of-value-in-numpy-array/\n",
    "                    # print(\"index of the table entry:\", index, \"that contains this child node:\", child_node)\n",
    "                    parent_node = newtables.edges.parent[index]\n",
    "                    # print(\"parent node\", parent_node)\n",
    "                    parent_time = newtables.nodes.time[parent_node]\n",
    "                    # print(\"time\", child_time, \"of the child node\", child_node, \"time of the parent node\", parent_time)\n",
    "                    time = np.random.uniform(low=child_time, high=parent_time)\n",
    "                    # print(\"time the mutation has been randomly selected to have\", time)\n",
    "                    newtables.mutations.add_row(site=site_id, node=child_node, time=time, derived_state = \"1\") #this doesn't work as with the above-- perhaps something with the strings. \n",
    "                    # print(tskit.unpack_strings(newtables.sites.ancestral_state, newtables.sites.ancestral_state_offset))\n",
    "                    # elif genealogy_mutations == 0:\n",
    "                    #   print(\"genealogy does not contain any mutations\")\n",
    "            # print(newtables.mutations)\n",
    "    \n",
    "    newtables.sort(edge_start=0, mutation_start=0, site_start=0)\n",
    "    new_ts = newtables.tree_sequence()\n",
    "    ####do sfs stats\n",
    "    # print(newtables.mutations)\n",
    "    total_mut_count.append(len(newtables.mutations))\n",
    "    mut_expectations_given_branchlengths.append(lambda_total)\n",
    "    for tree in new_ts.trees():\n",
    "        average_genealogy_size.append(tree.branch_length)\n",
    "        for mutation in tree.mutations():\n",
    "            # print(\"looking at mutation on node:\", mutation.node) \n",
    "            count = tree.num_samples(mutation.node)\n",
    "            # print(\"mutation count:\", count)\n",
    "            i_values.append(count) #dumping all i values into \n",
    "            \n",
    "    afs = new_ts.allele_frequency_spectrum(polarised=True, span_normalise=False)\n",
    "    afs_branches = new_ts.allele_frequency_spectrum(mode=\"branch\", polarised=True, span_normalise=False)\n",
    "    # print(\"values for i, the number of occurences of a mutation\", i_values)\n",
    "    # print(\"one loop of a sim::\", \"mutation numbers:\", genealogy_mutations, \"occurrences of these mutations\", i_values)\n",
    "    # return [genealogy_mutations, lambda_total, i_values]\n",
    "# print(newtables)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(\"number of samples for the runs\", ts.num_samples) #confusing that ts.num_samples is for haploid lineages, while samples in sim_ancestry is for individuals (which may not be haploid, and defaults to diploid). Note that this will be representative of the last loop over which tree sequence we're dealing with\n",
    "#theoretical expectation for time to MRCA in units of 2N generations:\n",
    "expected_tmrca = 2*(1-(1/sampled_sequences))\n",
    "print(\"expected Tmrca:\", expected_tmrca)\n",
    "print(\"mean tmrca, in generations/(ploidy*N):\", mean_tmrca/sim_runs/population_sequences, \"mean tmrca, in raw gens:\", mean_tmrca/sim_runs) #the first value is generations/(ploidy*N),and the second is in generations.\n",
    "#expected time that i lineages should exist: T(i) =  4N/i(i-1) (hudson 2015, #2)\n",
    "#sum i*(T(i)) from i=2 to n to get the total expected branch lengths: (above equation 6, hudson 1990)\n",
    "expected_total_length = 0\n",
    "for i in range(2, sampled_sequences):\n",
    "    expected_total_length += i*((4*(population_sequences/ploidy)/(i*(i-1))))\n",
    "\n",
    "print(\"total expected branch lengths\", expected_total_length, \"mean summed branch lengths:\", (sum(mean_summed_branch_lengths)/sim_runs))\n",
    "#now, we can use this to find the expected total number of polymorphisms (hudson 1990 eq1)\n",
    "expected_lambda = mu*expected_total_length\n",
    "print(\"mean expectation for the number of polymorphisms on the tree\", expected_lambda, \"mean lambda value, from lambdas used for sampling\", sum(mut_expectations_given_branchlengths)/sim_runs, \"number of average polymorphisms actually on the trees\", (sum(total_mut_count))/sim_runs)\n",
    "#sfs stats\n",
    "theta = 4*(population_sequences/ploidy)*mu\n",
    "\n",
    "# print(\"mutation counts from all loops\", i_values)\n",
    "mean_counts = {}\n",
    "for i in range(1, ((sampled_sequences))): #in eq 6 the range is from 1 to n-1, but due to the vagaries of python ranges this will produce that even though it doesn't *look* like it.\n",
    "    mean = (i_values.count(i))/sim_runs\n",
    "    mean_counts.update({i:mean})\n",
    "# mean_allele_freqs = (i_values.count(i))/sim_runs\n",
    "print(\"mean allele frequencies, from all sims binned\", mean_counts)\n",
    "# # print(\"average number of mutations in sims\", sum(mutations_in_run)/sim_runs, \"expected mean number of mutations\", expected_mutations)\n",
    "print(\"expected mean site frequency spectrum\", [(theta/i) for i in range(1, sampled_sequences)], \"mean allele frequencies from sims, averaged\", mean_counts)\n",
    "ind = np.arange(len(mean_counts))\n",
    "plt.bar(ind, list(mean_counts.values()))\n",
    "for t in new_ts.trees():\n",
    "  print(t.draw(format='unicode'))\n",
    "\n",
    "\n",
    "print(\"STATS FOR LAST LOOPED REPLICATE, as a sample:\", \"allele frequency spectrum\", afs, \"total branch lengths above each bin\", afs_branches)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAMPLE PRINT STATEMENTS: only apply to last loop over the treesequences\n",
      "list of tree sizes, in units of JUST generations [2768.2064850295615, 3262.9498204229712, 3407.8998780588627] list of sizes, weighted by how much of the genome they occupy: [1024.2363994609377, 1859.8813976410934, 204.47399268353175] weighting values [0.37, 0.57, 0.06]\n",
      "value of T, weighted average genealogy length 3088.591789785563\n",
      "mean mutation expectation, over genealogies in this ts: 3.088591789785563\n",
      "allele frequency spectrum [0. 1. 0. 1. 0. 0. 0. 0. 0. 0. 0.] total branch lengths above each bin [     0.         132298.26708021  71739.82167146  47752.35609973\n",
      "  13221.65530164      0.          22085.86732682   4176.4057003\n",
      "      0.          17584.80579839      0.        ]\n",
      "STATEMENTS THAT APPLY TO ALL TS\n",
      "mean expectation for the number of polymorphisms generated by each ts 5.763541392891398 number of average polymorphisms actually on the trees 5.622\n",
      "expected mean site frequency spectrum [2.0, 1.0, 0.6666666666666666, 0.5, 0.4, 0.3333333333333333, 0.2857142857142857, 0.25, 0.2222222222222222] mean allele frequencies from sims, averaged {1: 1.941, 2: 0.973, 3: 0.678, 4: 0.556, 5: 0.355, 6: 0.396, 7: 0.305, 8: 0.232, 9: 0.186, 10: 0.0, 11: 0.0, 12: 0.0, 13: 0.0, 14: 0.0, 15: 0.0, 16: 0.0, 17: 0.0, 18: 0.0}\n",
      "             20    \n",
      "          ┏━━━┻━━━┓\n",
      "         18       ┃\n",
      "      ┏━━━┻━━━━┓  ┃\n",
      "     16        ┃  ┃\n",
      "  ┏━━━┻━━┓     ┃  ┃\n",
      "  ┃     15     ┃  ┃\n",
      "  ┃    ┏━┻━┓   ┃  ┃\n",
      " 14    ┃   ┃   ┃  ┃\n",
      " ┏┻━┓  ┃   ┃   ┃  ┃\n",
      " ┃  ┃  ┃  13   ┃  ┃\n",
      " ┃  ┃  ┃  ┏┻┓  ┃  ┃\n",
      " ┃  ┃ 12  ┃ ┃  ┃  ┃\n",
      " ┃  ┃ ┏┻┓ ┃ ┃  ┃  ┃\n",
      "11  ┃ ┃ ┃ ┃ ┃  ┃  ┃\n",
      "┏┻┓ ┃ ┃ ┃ ┃ ┃  ┃  ┃\n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ 10  ┃\n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ ┏┻┓ ┃\n",
      "0 4 3 1 8 2 9 5 7 6\n",
      "\n",
      "           20      \n",
      "       ┏━━━━┻━━━━━┓\n",
      "      19          ┃\n",
      "  ┏━━━━┻━━━━┓     ┃\n",
      "  ┃        18     ┃\n",
      "  ┃      ┏━━┻━━┓  ┃\n",
      "  ┃     15     ┃  ┃\n",
      "  ┃    ┏━┻━┓   ┃  ┃\n",
      " 14    ┃   ┃   ┃  ┃\n",
      " ┏┻━┓  ┃   ┃   ┃  ┃\n",
      " ┃  ┃  ┃  13   ┃  ┃\n",
      " ┃  ┃  ┃  ┏┻┓  ┃  ┃\n",
      " ┃  ┃ 12  ┃ ┃  ┃  ┃\n",
      " ┃  ┃ ┏┻┓ ┃ ┃  ┃  ┃\n",
      "11  ┃ ┃ ┃ ┃ ┃  ┃  ┃\n",
      "┏┻┓ ┃ ┃ ┃ ┃ ┃  ┃  ┃\n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ 10  ┃\n",
      "┃ ┃ ┃ ┃ ┃ ┃ ┃ ┏┻┓ ┃\n",
      "0 4 3 1 8 2 9 5 7 6\n",
      "\n",
      "          20       \n",
      "      ┏━━━━┻━━━━┓  \n",
      "     19         ┃  \n",
      "  ┏━━━┻━━┓      ┃  \n",
      "  ┃     18      ┃  \n",
      "  ┃    ┏━┻━┓    ┃  \n",
      "  ┃    ┃   ┃   17  \n",
      "  ┃    ┃   ┃   ┏┻━┓\n",
      " 14    ┃   ┃   ┃  ┃\n",
      " ┏┻━┓  ┃   ┃   ┃  ┃\n",
      " ┃  ┃  ┃   ┃  13  ┃\n",
      " ┃  ┃  ┃   ┃  ┏┻┓ ┃\n",
      " ┃  ┃ 12   ┃  ┃ ┃ ┃\n",
      " ┃  ┃ ┏┻┓  ┃  ┃ ┃ ┃\n",
      "11  ┃ ┃ ┃  ┃  ┃ ┃ ┃\n",
      "┏┻┓ ┃ ┃ ┃  ┃  ┃ ┃ ┃\n",
      "┃ ┃ ┃ ┃ ┃ 10  ┃ ┃ ┃\n",
      "┃ ┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃ ┃\n",
      "0 4 3 1 8 5 7 2 9 6\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAGdCAYAAAAxCSikAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAqqElEQVR4nO3de3BUZZ7G8acJpgMUaUAgnWgkQTEMCAni0BPUVcaWJEMxZC8KlDtAFnCXhS2ZeIFMCZHV3QA6LDiTJYqEwLpysUSoFTeIkcAiAZZLSnGVAjYQkHS47CRN4pCwydk/LNptc6NDLm+a76fqlPR7fuft9/UlnYfTp0/bLMuyBAAAYLBunT0AAACAlhBYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADG697ZA2gL9fX1unDhgnr37i2bzdbZwwEAADfBsixdvXpVUVFR6tat+XMoQRFYLly4oOjo6M4eBgAAaIVz587p7rvvbrYmKAJL7969JX0/4fDw8E4eDQAAuBler1fR0dG+3+PNCSiwZGVlaevWrfrmm2/Uo0cPjR07VsuWLVNcXFyzx73//vtatGiRzpw5oyFDhmjZsmX6xS9+4dtvWZYyMzO1Zs0aVVRU6OGHH9bq1as1ZMiQmxrXjbeBwsPDCSwAAHQxN3M5R0AX3e7Zs0dz587VgQMHtGvXLl2/fl3jx49XdXV1k8fs379fU6dO1cyZM3Xs2DGlpqYqNTVVx48f99UsX75cb775pnJycnTw4EH16tVLSUlJunbtWiDDAwAAQcp2K9/WfOnSJQ0cOFB79uzRn/zJnzRaM3nyZFVXV+ujjz7ytf3sZz9TQkKCcnJyZFmWoqKi9Pzzz+uFF16QJFVWVioiIkJ5eXmaMmVKi+Pwer1yOByqrKzkDAsAAF1EIL+/b+ljzZWVlZKkfv36NVlTVFQkt9vt15aUlKSioiJJUklJiTwej1+Nw+GQy+Xy1QAAgNtbqy+6ra+v1/z58/Xwww/rgQceaLLO4/EoIiLCry0iIkIej8e3/0ZbUzU/VlNTo5qaGt9jr9fbqjkAAICuodVnWObOnavjx49r06ZNbTmem5KVlSWHw+Hb+EgzAADBrVWBZd68efroo4+0e/fuFj837XQ6VV5e7tdWXl4up9Pp23+jramaH8vIyFBlZaVvO3fuXGumAQAAuoiAAotlWZo3b54+/PBDffbZZ4qNjW3xmMTERBUUFPi17dq1S4mJiZKk2NhYOZ1Ovxqv16uDBw/6an7Mbrf7PsLMR5kBAAh+AV3DMnfuXL333nvavn27evfu7bvGxOFwqEePHpKkadOm6a677lJWVpYk6bnnntNjjz2m3/72t5owYYI2bdqkw4cP6+2335b0/Wev58+fr9dee01DhgxRbGysFi1apKioKKWmprbhVAEAQFcVUGBZvXq1JOnxxx/3a1+3bp1mzJghSSotLfX7PoCxY8fqvffe08svv6zf/OY3GjJkiLZt2+Z3oe5LL72k6upqPfvss6qoqNAjjzyi/Px8hYWFtXJaAAAgmNzSfVhMwX1YAADoejrsPiwAAAAdgcACAACMR2ABAADGI7AAAADjtfrW/LeTmIU72rS/M0sntGl/AAAEO86wAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYL+DAsnfvXk2cOFFRUVGy2Wzatm1bs/UzZsyQzWZrsA0fPtxX88orrzTYP3To0IAnAwAAglPAgaW6ulrx8fHKzs6+qfpVq1aprKzMt507d079+vXTU0895Vc3fPhwv7p9+/YFOjQAABCkugd6QEpKilJSUm663uFwyOFw+B5v27ZNf/jDH5SWluY/kO7d5XQ6Ax0OAAC4DXT4NSxr166V2+3WoEGD/NpPnjypqKgoDR48WM8884xKS0ub7KOmpkZer9dvAwAAwatDA8uFCxf07//+75o1a5Zfu8vlUl5envLz87V69WqVlJTo0Ucf1dWrVxvtJysry3fmxuFwKDo6uiOGDwAAOkmHBpb169erT58+Sk1N9WtPSUnRU089pZEjRyopKUkff/yxKioqtGXLlkb7ycjIUGVlpW87d+5cB4weAAB0loCvYWkty7KUm5urX/3qVwoNDW22tk+fPrr//vt16tSpRvfb7XbZ7fb2GCYAADBQh51h2bNnj06dOqWZM2e2WFtVVaXTp08rMjKyA0YGAABMF3BgqaqqUnFxsYqLiyVJJSUlKi4u9l0km5GRoWnTpjU4bu3atXK5XHrggQca7HvhhRe0Z88enTlzRvv379ef/umfKiQkRFOnTg10eAAAIAgF/JbQ4cOHNW7cON/j9PR0SdL06dOVl5ensrKyBp/wqays1AcffKBVq1Y12uf58+c1depUXblyRQMGDNAjjzyiAwcOaMCAAYEODwAABCGbZVlWZw/iVnm9XjkcDlVWVio8PLzN+49ZuKNN+zuzdEKb9gcAQFcUyO9vvksIAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIwXcGDZu3evJk6cqKioKNlsNm3btq3Z+sLCQtlstgabx+Pxq8vOzlZMTIzCwsLkcrl06NChQIcGAACCVMCBpbq6WvHx8crOzg7ouBMnTqisrMy3DRw40Ldv8+bNSk9PV2Zmpo4ePar4+HglJSXp4sWLgQ4PAAAEoe6BHpCSkqKUlJSAn2jgwIHq06dPo/tWrFih2bNnKy0tTZKUk5OjHTt2KDc3VwsXLgz4uQAAQHDpsGtYEhISFBkZqSeffFKff/65r722tlZHjhyR2+3+YVDdusntdquoqKjRvmpqauT1ev02AAAQvNo9sERGRionJ0cffPCBPvjgA0VHR+vxxx/X0aNHJUmXL19WXV2dIiIi/I6LiIhocJ3LDVlZWXI4HL4tOjq6vacBAAA6UcBvCQUqLi5OcXFxvsdjx47V6dOn9U//9E/6l3/5l1b1mZGRofT0dN9jr9dLaAEAIIi1e2BpzJgxY7Rv3z5JUv/+/RUSEqLy8nK/mvLycjmdzkaPt9vtstvt7T5OAABghk65D0txcbEiIyMlSaGhoRo9erQKCgp8++vr61VQUKDExMTOGB4AADBMwGdYqqqqdOrUKd/jkpISFRcXq1+/frrnnnuUkZGhb7/9Vhs2bJAkrVy5UrGxsRo+fLiuXbumd955R5999pk++eQTXx/p6emaPn26HnroIY0ZM0YrV65UdXW171NDAADg9hZwYDl8+LDGjRvne3zjWpLp06crLy9PZWVlKi0t9e2vra3V888/r2+//VY9e/bUyJEj9emnn/r1MXnyZF26dEmLFy+Wx+NRQkKC8vPzG1yICwAAbk82y7Kszh7ErfJ6vXI4HKqsrFR4eHib9x+zcEeb9ndm6YQ27Q8AgK4okN/ffJcQAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABgv4MCyd+9eTZw4UVFRUbLZbNq2bVuz9Vu3btWTTz6pAQMGKDw8XImJidq5c6dfzSuvvCKbzea3DR06NNChAQCAIBVwYKmurlZ8fLyys7Nvqn7v3r168skn9fHHH+vIkSMaN26cJk6cqGPHjvnVDR8+XGVlZb5t3759gQ4NAAAEqe6BHpCSkqKUlJSbrl+5cqXf43/8x3/U9u3b9W//9m8aNWrUDwPp3l1OpzPQ4QAAgNtAh1/DUl9fr6tXr6pfv35+7SdPnlRUVJQGDx6sZ555RqWlpR09NAAAYKiAz7DcqjfeeENVVVV6+umnfW0ul0t5eXmKi4tTWVmZlixZokcffVTHjx9X7969G/RRU1Ojmpoa32Ov19shYwcAAJ2jQwPLe++9pyVLlmj79u0aOHCgr/3/v8U0cuRIuVwuDRo0SFu2bNHMmTMb9JOVlaUlS5Z0yJgBAEDn67C3hDZt2qRZs2Zpy5Ytcrvdzdb26dNH999/v06dOtXo/oyMDFVWVvq2c+fOtceQAQCAIToksGzcuFFpaWnauHGjJkyY0GJ9VVWVTp8+rcjIyEb32+12hYeH+20AACB4BfyWUFVVld+Zj5KSEhUXF6tfv3665557lJGRoW+//VYbNmyQ9P3bQNOnT9eqVavkcrnk8XgkST169JDD4ZAkvfDCC5o4caIGDRqkCxcuKDMzUyEhIZo6dWpbzBEAAHRxAZ9hOXz4sEaNGuX7SHJ6erpGjRqlxYsXS5LKysr8PuHz9ttv63//9381d+5cRUZG+rbnnnvOV3P+/HlNnTpVcXFxevrpp3XnnXfqwIEDGjBgwK3ODwAABAGbZVlWZw/iVnm9XjkcDlVWVrbL20MxC3e0aX9nlrb8thgAAMEukN/ffJcQAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIwXcGDZu3evJk6cqKioKNlsNm3btq3FYwoLC/Xggw/KbrfrvvvuU15eXoOa7OxsxcTEKCwsTC6XS4cOHQp0aAAAIEgFHFiqq6sVHx+v7Ozsm6ovKSnRhAkTNG7cOBUXF2v+/PmaNWuWdu7c6avZvHmz0tPTlZmZqaNHjyo+Pl5JSUm6ePFioMMDAABByGZZltXqg202ffjhh0pNTW2yZsGCBdqxY4eOHz/ua5syZYoqKiqUn58vSXK5XPrpT3+q3//+95Kk+vp6RUdH6+/+7u+0cOHCFsfh9XrlcDhUWVmp8PDw1k6nSTELd7Rpf2eWTmjT/gAA6IoC+f3d7tewFBUVye12+7UlJSWpqKhIklRbW6sjR4741XTr1k1ut9tX82M1NTXyer1+GwAACF7tHlg8Ho8iIiL82iIiIuT1evXHP/5Rly9fVl1dXaM1Ho+n0T6zsrLkcDh8W3R0dLuNHwAAdL4u+SmhjIwMVVZW+rZz58519pAAAEA76t7eT+B0OlVeXu7XVl5ervDwcPXo0UMhISEKCQlptMbpdDbap91ul91ub7cxAwAAs7T7GZbExEQVFBT4te3atUuJiYmSpNDQUI0ePdqvpr6+XgUFBb4aAABwews4sFRVVam4uFjFxcWSvv/YcnFxsUpLSyV9/3bNtGnTfPV/8zd/o//+7//WSy+9pG+++Ub//M//rC1btujXv/61ryY9PV1r1qzR+vXr9fXXX2vOnDmqrq5WWlraLU4PAAAEg4DfEjp8+LDGjRvne5yeni5Jmj59uvLy8lRWVuYLL5IUGxurHTt26Ne//rVWrVqlu+++W++8846SkpJ8NZMnT9alS5e0ePFieTweJSQkKD8/v8GFuMGsrT86LfHxaQBA8Lil+7CYIhjuw0JgAQDcboy6DwsAAMCtIrAAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOO1KrBkZ2crJiZGYWFhcrlcOnToUJO1jz/+uGw2W4NtwoQJvpoZM2Y02J+cnNyaoQEAgCDUPdADNm/erPT0dOXk5MjlcmnlypVKSkrSiRMnNHDgwAb1W7duVW1tre/xlStXFB8fr6eeesqvLjk5WevWrfM9ttvtgQ4NAAAEqYDPsKxYsUKzZ89WWlqahg0bppycHPXs2VO5ubmN1vfr109Op9O37dq1Sz179mwQWOx2u19d3759WzcjAAAQdAIKLLW1tTpy5IjcbvcPHXTrJrfbraKiopvqY+3atZoyZYp69erl115YWKiBAwcqLi5Oc+bM0ZUrV5rso6amRl6v128DAADBK6DAcvnyZdXV1SkiIsKvPSIiQh6Pp8XjDx06pOPHj2vWrFl+7cnJydqwYYMKCgq0bNky7dmzRykpKaqrq2u0n6ysLDkcDt8WHR0dyDQAAEAXE/A1LLdi7dq1GjFihMaMGePXPmXKFN+fR4wYoZEjR+ree+9VYWGhnnjiiQb9ZGRkKD093ffY6/USWgAACGIBnWHp37+/QkJCVF5e7tdeXl4up9PZ7LHV1dXatGmTZs6c2eLzDB48WP3799epU6ca3W+32xUeHu63AQCA4BVQYAkNDdXo0aNVUFDga6uvr1dBQYESExObPfb9999XTU2N/vIv/7LF5zl//ryuXLmiyMjIQIYHAACCVMCfEkpPT9eaNWu0fv16ff3115ozZ46qq6uVlpYmSZo2bZoyMjIaHLd27Vqlpqbqzjvv9GuvqqrSiy++qAMHDujMmTMqKCjQpEmTdN999ykpKamV0wIAAMEk4GtYJk+erEuXLmnx4sXyeDxKSEhQfn6+70Lc0tJSdevmn4NOnDihffv26ZNPPmnQX0hIiL744gutX79eFRUVioqK0vjx4/Xqq69yLxYAACCplRfdzps3T/PmzWt0X2FhYYO2uLg4WZbVaH2PHj20c+fO1gwDAADcJvguIQAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA43Xodwmh88Us3NGm/Z1ZOqFN+wMAoDGcYQEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGC87p09AASfmIU72rS/M0sntGl/AICuhzMsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACM16rAkp2drZiYGIWFhcnlcunQoUNN1ubl5clms/ltYWFhfjWWZWnx4sWKjIxUjx495Ha7dfLkydYMDQAABKGAA8vmzZuVnp6uzMxMHT16VPHx8UpKStLFixebPCY8PFxlZWW+7ezZs377ly9frjfffFM5OTk6ePCgevXqpaSkJF27di3wGQEAgKATcGBZsWKFZs+erbS0NA0bNkw5OTnq2bOncnNzmzzGZrPJ6XT6toiICN8+y7K0cuVKvfzyy5o0aZJGjhypDRs26MKFC9q2bVurJgUAAIJLQIGltrZWR44ckdvt/qGDbt3kdrtVVFTU5HFVVVUaNGiQoqOjNWnSJH311Ve+fSUlJfJ4PH59OhwOuVyuZvsEAAC3j4ACy+XLl1VXV+d3hkSSIiIi5PF4Gj0mLi5Oubm52r59u959913V19dr7NixOn/+vCT5jgukz5qaGnm9Xr8NAAAEr3b/lFBiYqKmTZumhIQEPfbYY9q6dasGDBigt956q9V9ZmVlyeFw+Lbo6Og2HDEAADBNQIGlf//+CgkJUXl5uV97eXm5nE7nTfVxxx13aNSoUTp16pQk+Y4LpM+MjAxVVlb6tnPnzgUyDQAA0MUEFFhCQ0M1evRoFRQU+Nrq6+tVUFCgxMTEm+qjrq5OX375pSIjIyVJsbGxcjqdfn16vV4dPHiwyT7tdrvCw8P9NgAAELwC/rbm9PR0TZ8+XQ899JDGjBmjlStXqrq6WmlpaZKkadOm6a677lJWVpYk6e///u/1s5/9TPfdd58qKir0+uuv6+zZs5o1a5ak7z9BNH/+fL322msaMmSIYmNjtWjRIkVFRSk1NbXtZgoAALqsgAPL5MmTdenSJS1evFgej0cJCQnKz8/3XTRbWlqqbt1+OHHzhz/8QbNnz5bH41Hfvn01evRo7d+/X8OGDfPVvPTSS6qurtazzz6riooKPfLII8rPz29wgzkAAHB7slmWZXX2IG6V1+uVw+FQZWVlu7w9FLNwR5v2d2bphHZ/jo56ns56DgBA1xfI72++SwgAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4wX8bc3A7YQvcgQAM3CGBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPO7Dgi6pre+PInGPFAAwGWdYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjceM4oJO19U3wuAEegGDEGRYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMZrVWDJzs5WTEyMwsLC5HK5dOjQoSZr16xZo0cffVR9+/ZV37595Xa7G9TPmDFDNpvNb0tOTm7N0AAAQBAKOLBs3rxZ6enpyszM1NGjRxUfH6+kpCRdvHix0frCwkJNnTpVu3fvVlFRkaKjozV+/Hh9++23fnXJyckqKyvzbRs3bmzdjAAAQNAJOLCsWLFCs2fPVlpamoYNG6acnBz17NlTubm5jdb/67/+q/72b/9WCQkJGjp0qN555x3V19eroKDAr85ut8vpdPq2vn37tm5GAAAg6AQUWGpra3XkyBG53e4fOujWTW63W0VFRTfVx3fffafr16+rX79+fu2FhYUaOHCg4uLiNGfOHF25cqXJPmpqauT1ev02AAAQvAIKLJcvX1ZdXZ0iIiL82iMiIuTxeG6qjwULFigqKsov9CQnJ2vDhg0qKCjQsmXLtGfPHqWkpKiurq7RPrKysuRwOHxbdHR0INMAAABdTIfemn/p0qXatGmTCgsLFRYW5mufMmWK788jRozQyJEjde+996qwsFBPPPFEg34yMjKUnp7ue+z1egktAAAEsYDOsPTv318hISEqLy/3ay8vL5fT6Wz22DfeeENLly7VJ598opEjRzZbO3jwYPXv31+nTp1qdL/dbld4eLjfBgAAgldAZ1hCQ0M1evRoFRQUKDU1VZJ8F9DOmzevyeOWL1+uf/iHf9DOnTv10EMPtfg858+f15UrVxQZGRnI8AA0oa2/YFHiSxYBdKyAPyWUnp6uNWvWaP369fr66681Z84cVVdXKy0tTZI0bdo0ZWRk+OqXLVumRYsWKTc3VzExMfJ4PPJ4PKqqqpIkVVVV6cUXX9SBAwd05swZFRQUaNKkSbrvvvuUlJTURtMEAABdWcDXsEyePFmXLl3S4sWL5fF4lJCQoPz8fN+FuKWlperW7YcctHr1atXW1uov/uIv/PrJzMzUK6+8opCQEH3xxRdav369KioqFBUVpfHjx+vVV1+V3W6/xekBAIBg0KqLbufNm9fkW0CFhYV+j8+cOdNsXz169NDOnTtbMwwAAHCb4LuEAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYr0NvzQ8guLX1Deq4OR2AGzjDAgAAjEdgAQAAxiOwAAAA4xFYAACA8QgsAADAeAQWAABgPAILAAAwHoEFAAAYj8ACAACMR2ABAADGI7AAAADjEVgAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMB6BBQAAGI/AAgAAjEdgAQAAxiOwAAAA4xFYAACA8VoVWLKzsxUTE6OwsDC5XC4dOnSo2fr3339fQ4cOVVhYmEaMGKGPP/7Yb79lWVq8eLEiIyPVo0cPud1unTx5sjVDAwAAQSjgwLJ582alp6crMzNTR48eVXx8vJKSknTx4sVG6/fv36+pU6dq5syZOnbsmFJTU5Wamqrjx4/7apYvX64333xTOTk5OnjwoHr16qWkpCRdu3at9TMDAABBI+DAsmLFCs2ePVtpaWkaNmyYcnJy1LNnT+Xm5jZav2rVKiUnJ+vFF1/UT37yE7366qt68MEH9fvf/17S92dXVq5cqZdfflmTJk3SyJEjtWHDBl24cEHbtm27pckBAIDg0D2Q4traWh05ckQZGRm+tm7dusntdquoqKjRY4qKipSenu7XlpSU5AsjJSUl8ng8crvdvv0Oh0Mul0tFRUWaMmVKgz5rampUU1Pje1xZWSlJ8nq9gUznptXXfNem/TU2zrZ+jo56nmB5jo56nmB5jo56nvb6mQZghhs/45ZltVgbUGC5fPmy6urqFBER4dceERGhb775ptFjPB5Po/Uej8e3/0ZbUzU/lpWVpSVLljRoj46OvrmJdDLHyuB5nmB5jo56nmB5jo56no6aC4DOdfXqVTkcjmZrAgospsjIyPA7a1NfX6//+Z//0Z133imbzdYpY/J6vYqOjta5c+cUHh7eKWPoLLfz3KXbe/7M/facu3R7z5+5t93cLcvS1atXFRUV1WJtQIGlf//+CgkJUXl5uV97eXm5nE5no8c4nc5m62/8t7y8XJGRkX41CQkJjfZpt9tlt9v92vr06RPIVNpNeHj4bfcX+Ibbee7S7T1/5n57zl26vefP3Ntm7i2dWbkhoItuQ0NDNXr0aBUUFPja6uvrVVBQoMTExEaPSUxM9KuXpF27dvnqY2Nj5XQ6/Wq8Xq8OHjzYZJ8AAOD2EvBbQunp6Zo+fboeeughjRkzRitXrlR1dbXS0tIkSdOmTdNdd92lrKwsSdJzzz2nxx57TL/97W81YcIEbdq0SYcPH9bbb78tSbLZbJo/f75ee+01DRkyRLGxsVq0aJGioqKUmpradjMFAABdVsCBZfLkybp06ZIWL14sj8ejhIQE5efn+y6aLS0tVbduP5y4GTt2rN577z29/PLL+s1vfqMhQ4Zo27ZteuCBB3w1L730kqqrq/Xss8+qoqJCjzzyiPLz8xUWFtYGU+wYdrtdmZmZDd6quh3cznOXbu/5M/fbc+7S7T1/5t45c7dZN/NZIgAAgE7EdwkBAADjEVgAAIDxCCwAAMB4BBYAAGA8AksAsrOzFRMTo7CwMLlcLh06dKjZ+vfff19Dhw5VWFiYRowYoY8//riDRtq2srKy9NOf/lS9e/fWwIEDlZqaqhMnTjR7TF5enmw2m9/WlT71dcMrr7zSYB5Dhw5t9phgWfeYmJgGc7fZbJo7d26j9V19zffu3auJEycqKipKNputwZevWpalxYsXKzIyUj169JDb7dbJkydb7DfQ143O0Nzcr1+/rgULFmjEiBHq1auXoqKiNG3aNF24cKHZPlvzs9MZWlr3GTNmNJhHcnJyi/12hXWXWp5/Y68BNptNr7/+epN9ttfaE1hu0ubNm5Wenq7MzEwdPXpU8fHxSkpK0sWLFxut379/v6ZOnaqZM2fq2LFjSk1NVWpqqo4fP97BI791e/bs0dy5c3XgwAHt2rVL169f1/jx41VdXd3sceHh4SorK/NtZ8+e7aARt63hw4f7zWPfvn1N1gbTuv/nf/6n37x37dolSXrqqaeaPKYrr3l1dbXi4+OVnZ3d6P7ly5frzTffVE5Ojg4ePKhevXopKSlJ165da7LPQF83Oktzc//uu+909OhRLVq0SEePHtXWrVt14sQJ/fKXv2yx30B+djpLS+suScnJyX7z2LhxY7N9dpV1l1qe//+fd1lZmXJzc2Wz2fTnf/7nzfbbLmtv4aaMGTPGmjt3ru9xXV2dFRUVZWVlZTVa//TTT1sTJkzwa3O5XNZf//Vft+s4O8LFixctSdaePXuarFm3bp3lcDg6blDtJDMz04qPj7/p+mBe9+eee8669957rfr6+kb3B8uaW5ZlSbI+/PBD3+P6+nrL6XRar7/+uq+toqLCstvt1saNG5vsJ9DXDRP8eO6NOXTokCXJOnv2bJM1gf7smKCxuU+fPt2aNGlSQP10xXW3rJtb+0mTJlk///nPm61pr7XnDMtNqK2t1ZEjR+R2u31t3bp1k9vtVlFRUaPHFBUV+dVLUlJSUpP1XUllZaUkqV+/fs3WVVVVadCgQYqOjtakSZP01VdfdcTw2tzJkycVFRWlwYMH65lnnlFpaWmTtcG67rW1tXr33Xf1V3/1V81+wWiwrPmPlZSUyOPx+K2tw+GQy+Vqcm1b87rRVVRWVspms7X4HW6B/OyYrLCwUAMHDlRcXJzmzJmjK1euNFkbzOteXl6uHTt2aObMmS3WtsfaE1huwuXLl1VXV+e7m+8NERER8ng8jR7j8XgCqu8q6uvrNX/+fD388MN+dyv+sbi4OOXm5mr79u169913VV9fr7Fjx+r8+fMdONpb53K5lJeXp/z8fK1evVolJSV69NFHdfXq1Ubrg3Xdt23bpoqKCs2YMaPJmmBZ88bcWL9A1rY1rxtdwbVr17RgwQJNnTq12S+/C/Rnx1TJycnasGGDCgoKtGzZMu3Zs0cpKSmqq6trtD5Y112S1q9fr969e+vP/uzPmq1rr7UP+Nb8uL3NnTtXx48fb/H9yMTERL8vrxw7dqx+8pOf6K233tKrr77a3sNsMykpKb4/jxw5Ui6XS4MGDdKWLVtu6l8ZwWLt2rVKSUlp9ivgg2XN0bTr16/r6aeflmVZWr16dbO1wfKzM2XKFN+fR4wYoZEjR+ree+9VYWGhnnjiiU4cWcfLzc3VM8880+LF9O219pxhuQn9+/dXSEiIysvL/drLy8vldDobPcbpdAZU3xXMmzdPH330kXbv3q277747oGPvuOMOjRo1SqdOnWqn0XWMPn366P77729yHsG47mfPntWnn36qWbNmBXRcsKy5JN/6BbK2rXndMNmNsHL27Fnt2rWr2bMrjWnpZ6erGDx4sPr379/kPIJt3W/4j//4D504cSLg1wGp7daewHITQkNDNXr0aBUUFPja6uvrVVBQ4Pcvyv8vMTHRr16Sdu3a1WS9ySzL0rx58/Thhx/qs88+U2xsbMB91NXV6csvv1RkZGQ7jLDjVFVV6fTp003OI5jW/YZ169Zp4MCBmjBhQkDHBcuaS1JsbKycTqff2nq9Xh08eLDJtW3N64apboSVkydP6tNPP9Wdd94ZcB8t/ex0FefPn9eVK1eanEcwrfv/t3btWo0ePVrx8fEBH9tma9/ml/EGqU2bNll2u93Ky8uz/uu//st69tlnrT59+lgej8eyLMv61a9+ZS1cuNBX//nnn1vdu3e33njjDevrr7+2MjMzrTvuuMP68ssvO2sKrTZnzhzL4XBYhYWFVllZmW/77rvvfDU/nv+SJUusnTt3WqdPn7aOHDliTZkyxQoLC7O++uqrzphCqz3//PNWYWGhVVJSYn3++eeW2+22+vfvb128eNGyrOBed8v6/tMN99xzj7VgwYIG+4Jtza9evWodO3bMOnbsmCXJWrFihXXs2DHfJ2GWLl1q9enTx9q+fbv1xRdfWJMmTbJiY2OtP/7xj74+fv7zn1u/+93vfI9bet0wRXNzr62ttX75y19ad999t1VcXOz3GlBTU+Pr48dzb+lnxxTNzf3q1avWCy+8YBUVFVklJSXWp59+aj344IPWkCFDrGvXrvn66Krrblkt/723LMuqrKy0evbsaa1evbrRPjpq7QksAfjd735n3XPPPVZoaKg1ZswY68CBA759jz32mDV9+nS/+i1btlj333+/FRoaag0fPtzasWNHB4+4bUhqdFu3bp2v5sfznz9/vu//VUREhPWLX/zCOnr0aMcP/hZNnjzZioyMtEJDQ6277rrLmjx5snXq1Cnf/mBed8uyrJ07d1qSrBMnTjTYF2xrvnv37kb/nt+YY319vbVo0SIrIiLCstvt1hNPPNHg/8ugQYOszMxMv7bmXjdM0dzcS0pKmnwN2L17t6+PH8+9pZ8dUzQ39++++84aP368NWDAAOuOO+6wBg0aZM2ePbtB8Oiq625ZLf+9tyzLeuutt6wePXpYFRUVjfbRUWtvsyzLurVzNAAAAO2La1gAAIDxCCwAAMB4BBYAAGA8AgsAADAegQUAABiPwAIAAIxHYAEAAMYjsAAAAOMRWAAAgPEILAAAwHgEFgAAYDwCCwAAMN7/Ae2Yint2coL8AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#do the above with multiple trees\n",
    "import msprime\n",
    "import tskit \n",
    "import numpy as np\n",
    "from numpy.random import default_rng\n",
    "rng = default_rng()\n",
    "import random\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sequence_length = 100\n",
    "mu = (1e-3) #normally would have this set to 1e-7 *2, but putting deme size down and #individuals up to be able to visualize tree sequences for working on placing mutations issue.\n",
    "# sequence_length = 100\n",
    "ploidy = 2\n",
    "population_sequences = 1000 #total number of genomes in the population- the number of individuals is this/ploidy\n",
    "sampled_sequences = 10\n",
    "sim_runs = 1000\n",
    "rho = (1e-6)*5\n",
    "#initialize the bottom two variables, to add/append to later \n",
    "mean_tmrca = 0\n",
    "i_values = []\n",
    "mean_summed_branch_lengths = []\n",
    "total_mut_count = []\n",
    "mean_expected_lambda = []\n",
    "mean_utilised_lambda = []\n",
    "\n",
    "for ts in msprime.sim_ancestry(ploidy= ploidy, samples=(sampled_sequences/ploidy), discrete_genome = True, population_size=(population_sequences)/ploidy, num_replicates = sim_runs, sequence_length = sequence_length, recombination_rate=rho): #the sim needs everything in terms of the number of individuals, but it's convenient for downstream/interpretation to have our original variables be in terms of the number of lineages or individual sequences, aka in haploid terms \n",
    "    newtables = ts.dump_tables() #modifying a copy of the original tables in order to be able to overwrite \n",
    "    assert ts.num_samples == sampled_sequences\n",
    "    last_node = ts.num_nodes - 1\n",
    "    tmrca = ts.node(last_node).time\n",
    "    mean_tmrca += tmrca\n",
    "    # print(\"tmrca for this run:\", tmrca)\n",
    "    # print(ts.draw_text())\n",
    "    # print(newtables.edges)\n",
    "    genealogy_sizes = []\n",
    "    weighted_genealogy_sizes = []\n",
    "    genealogy_sizes = []\n",
    "    sequence_weights = []\n",
    "    for tree in ts.trees():\n",
    "        # print(tree.draw(format='unicode'))\n",
    "        current_tree_interval = tree.interval #find the interval the tree occupies in the genome \n",
    "        span = tree.interval.right-tree.interval.left #find the fraction of the total genome/number of bases, for multiple trees \n",
    "        # print(newtables.edges)\n",
    "        # print(\"interval this tree exists on\", current_tree_interval)\n",
    "        nodes_and_branches = {} #empty dict to hold the pair of node:branch length to its parent. each tree gets its own list of these \n",
    "        for node in tree.nodes(): #needed in the recombination version of the sim due to not all nodes being present in all trees \n",
    "            for child_node in newtables.edges.child:\n",
    "                if child_node ==node: \n",
    "                    nodes_and_branches.update({child_node:(tree.branch_length(child_node))}) \n",
    "                    \n",
    "        # x = (list(nodes_and_branches.keys()))\n",
    "        # x.sort()\n",
    "        # print(\"list of non-root nodes in the tree (ones that can have muts placed on them):\", x)\n",
    "        summed_branch_lengths = sum(nodes_and_branches.values()) #hudson 1990, above equation 1. I could have just used the tskit function but I was going to loop over the nodes to make a dict anwyay \n",
    "        mean_summed_branch_lengths.append(summed_branch_lengths)\n",
    "        proportional_branch_lengths = {} #empty dict to hold the pair of node:scaled branch length\n",
    "        # print(\"sum of the lengths of all branches:\", summed_branch_lengths, \"what the tskit function gives you\", sum(tree.branch_length(u) for u in tree.nodes()))\n",
    "        # print(\"list of nodes in the tree that can have muts placed above them:\", child_nodes_list)\n",
    "        for node in tree.nodes():\n",
    "            for child_node in newtables.edges.child: #I probably can work around this and not have another loop by accessing the dictionary keys, but it doesn't seem worth the effort\n",
    "                if child_node == node: \n",
    "                    proportional_branch_lengths.update({child_node:(tree.branch_length(child_node)/summed_branch_lengths)}) #hudson 1990 equation 1\n",
    "        # print(proportional_branch_lengths)\n",
    "        # proportional_branch_lengths = dict(sorted(proportional_branch_lengths.items())) #sort the dictionary so the keys(representing nodes) go up in order. I thought this would matter for the np array step, but this may not help once we have >1 tree, as the incies of the np array will no longer match the dict\n",
    "        # print(\"nodes and the branch lengths above them, proportional to the total length of all branches:\" , proportional_branch_lengths) \n",
    "        #hudson 1990, equation 1: compare to the expectation of s above        \n",
    "        lambda_tree = mu*summed_branch_lengths*(span/sequence_length) #hudson 1990 eq 1\n",
    "        tree_mutations = rng.poisson(lam=lambda_tree)\n",
    "        # print(\"number of mutations we sampled from poisson distribution:\", genealogy_mutations, \"mean number of mutations in the poisson distribution expectations for this tree\", lambda_total)\n",
    "        # print(\"proportional branch lengths:\", proportional_branch_lengths)\n",
    "        p = list(proportional_branch_lengths.values())\n",
    "        # print(\"child nodes\", len(child_nodes_list))\n",
    "        # print(\"proportional branch lengths\", p, \"sum\", sum(p))\n",
    " \n",
    "        if tree_mutations != 0:\n",
    "            # print(\"genealogy contains more than 0 mutations\")\n",
    "            ###########sample nodes from your list of child nodes \n",
    "            nodes_sampled =  rng.choice(list(nodes_and_branches.keys()), p=p, size=tree_mutations) #this can sample the same node twice, which IS what you want. however need to figure out a way to determine which parent node was associated with a given branch length, so you can select a site that makes sense given the edges possible for a child:parent node relationship. int is a pretty crude \n",
    "            # print(\"nodes sampled from distribution, aka chosen to have mutations\", nodes_sampled)\n",
    "            ###########pick out the spot in the genome that the tree we have iterated over actually occupies (with edges) \n",
    "            # print(\"interval this tree occupies\", current_tree_interval) #access this to pick which sites are possible\n",
    "            for child_node in nodes_sampled: #since I do a pretty much identical thing above, there's gotta be a way to combine the two, but I'll leave the stunts for later \n",
    "                ###########draw a site from the genome that HASN'T ALREADY had a mutation assigned to it.\n",
    "                # print(\"node mutation is being added onto:\", child_node) \n",
    "                possible_bases = set(range((int(current_tree_interval.left)), int(current_tree_interval.right))) #set them to integers for now? need to figure out how to actually implement using sets \n",
    "                bases_already_mutated = set(int(i) for i in newtables.sites.position)\n",
    "                unmutated_sites =(possible_bases.difference(bases_already_mutated)) #get the \n",
    "                position = np.random.randint(low=(current_tree_interval.left), high=(int(current_tree_interval.right)), size=1) #looks like this is working fine for \n",
    "                position = position[0]\n",
    "                # print(\"position of new mut:\", position)\n",
    "                ############use these nodes to write mutations to the mutations table\n",
    "                if position in unmutated_sites: #size should be 1, since we're only placing one mutation above each loop over a node\n",
    "                    site_id = newtables.sites.add_row(position=position, ancestral_state=\"0\")\n",
    "                    # print(\"id of the site we are adding\", site_id)\n",
    "                    #####get times for the mutations- randomly select them from the interval between parent and child node times \n",
    "                    # get child node time\n",
    "                    child_time = newtables.nodes.time[child_node]\n",
    "                    # print(\"node ID\", child_node,  ts.tables.edges.child[child_node])#https://stackoverflow.com/questions/176918/finding-the-index-of-an-item-in-a-\n",
    "                    #now, find the parent node that goes with the child node: \n",
    "                    index = np.where(newtables.edges.child==child_node)[0] #https://www.geeksforgeeks.org/how-to-find-the-index-of-value-in-numpy-array/\n",
    "                    # print(\"child node\", child_node, \"row in edge tables\", index)\n",
    "                    entries = newtables.edges[index]\n",
    "                    # print(entries)\n",
    "                    for row_id in index:\n",
    "                        entry = newtables.edges[row_id]\n",
    "                        # print(\"indices of the edge table entry\", row_id, \"that contains this child node:\", child_node, \"full entry:\", entry)\n",
    "                        if (entry.left) == current_tree_interval.left:   \n",
    "                            if entry.right == current_tree_interval.right: #if the interval matches the interval of the tree we're looping over exactly\n",
    "                                # print(\"entry that is actually in current tree interval\", entry)\n",
    "                                parent_node = entry.parent\n",
    "                                parent_time = newtables.nodes.time[parent_node]\n",
    "                            elif entry.right >= current_tree_interval.right: \n",
    "                                parent_node = entry.parent\n",
    "                                parent_time = newtables.nodes.time[parent_node]  \n",
    "                        elif entry.left <= current_tree_interval.left:\n",
    "                            if entry.right >= current_tree_interval.right: #if the tree sequence is contained entirely within the entry interval\n",
    "                                parent_node = entry.parent\n",
    "                                parent_time = newtables.nodes.time[parent_node]\n",
    "                        elif (entry.left, entry.right) == (0, sequence_length): #this happens when the node is not associated with a recombination event \n",
    "                            # print(\"this node has the same parent regardless of tree\") #do I need to do this above??)\n",
    "                            parent_node = entry.parent\n",
    "                            parent_time = newtables.nodes.time[parent_node]\n",
    "                        # print(\"parent node\", parent_node)\n",
    "                    # print(newtables.nodes)\n",
    "                    # print(\"time\", child_time, \"of the child node\", child_node, \"time\", parent_time, \"of the parent node\", parent_node)\n",
    "                    time = np.random.uniform(low=child_time, high=parent_time)\n",
    "                    # print(\"time the mutation has been randomly selected to have\", time)\n",
    "                    newtables.mutations.add_row(site=site_id, node=child_node, time=time, derived_state = \"1\") #this doesn't work as with the above-- perhaps something with the strings. \n",
    "                    # print(tskit.unpack_strings(newtables.sites.ancestral_state, newtables.sites.ancestral_state_offset))\n",
    "        # elif tree_mutations == 0:\n",
    "            # print(\"tree does not contain any mutations\")\\\n",
    "        ##########expected number of mutations:hudson p. 17\n",
    "        #average all the sizes of the genealogies of each tree together:\n",
    "        genealogy_sizes.append(summed_branch_lengths) \n",
    "        sequence_weights.append(span/sequence_length)\n",
    "        weighted_genealogy_sizes.append(summed_branch_lengths*(span/sequence_length)) #multiply the size of the genealogy * the proportion of the genome it occupies \n",
    "\n",
    "    ########find the size of the tree sequence's genealogies, weighted by their lengths:\n",
    "    # print(\"list of tree sizes\", genealogy_sizes, \"list of sizes, weighted by how much of the genome they occupy:\", weighted_genealogy_sizes, \"weighting values\", sequence_weights)\n",
    "    #find the weighted average genealogy length for this tree sequence\n",
    "    T_value = sum(weighted_genealogy_sizes)/sum(sequence_weights) \n",
    "    # print(\"value of T, weighted average genealogy length\", T_value)\n",
    "    #making sure the method above works\n",
    "    # print(\"make sure the above numbers are correct:\", sequence_weights, genealogy_sizes)\n",
    "    # for i, value in enumerate(genealogy_sizes):\n",
    "    #     print(\"Calculation\", genealogy_sizes[i]*sequence_weights[i])\n",
    "    # average_tree_size = sum(weighted_genealogy_sizes)\n",
    "    theta = 4*(population_sequences/ploidy)*mu\n",
    "    expected_mean_number_mutations = (((T_value * (4*(population_sequences/ploidy)*(mu*sequence_length))) / 2) ) /(population_sequences*sequence_length)#for recombination, p. 17 in hudson 1990. TO make it the same order of magnitude, I did the math backwards and just tried out dividing by N*ploidy*sequence length. \n",
    "    # print(\"mutation mean expectation:\", expected_mean_number_mutations)\n",
    "    mean_expected_lambda.append(expected_mean_number_mutations)\n",
    "    # print(\"mean expected lambda output\", mean_expected_lambda)\n",
    "    \n",
    "    \n",
    "    newtables.sort(edge_start=0, mutation_start=0, site_start=0)\n",
    "    new_ts = newtables.tree_sequence()\n",
    "    ####do sfs stats\n",
    "    total_mut_count.append(len(newtables.mutations))\n",
    "    # mut_expectations_given_branchlengths.append(lambda_total)\n",
    "    for tree in new_ts.trees():\n",
    "        for mutation in tree.mutations():\n",
    "            # print(\"looking at mutation on node:\", mutation.node) \n",
    "            count = tree.num_samples(mutation.node)\n",
    "            # print(\"mutation count:\", count)\n",
    "            i_values.append(count) #dumping all i values from multiple sim runs\n",
    "    # print(\"values for i, the number of occurences of a mutation\", i_values)\n",
    "    # print(\"one loop of a sim::\", \"mutation numbers:\", genealogy_mutations, \"occurrences of these mutations\", i_values)\n",
    "    # print(\"break between loops over a sim\")\n",
    "\n",
    "\n",
    "\n",
    "# print(newtables)\n",
    "# print(newtables.mutations)\n",
    "# print(\"number of samples for the runs\", ts.num_samples) #confusing that ts.num_samples is for haploid lineages, while samples in sim_ancestry is for individuals (which may not be haploid, and defaults to diploid). Note that this will be representative of the last loop over which tree sequence we're dealing with\n",
    "#theoretical expectation for time to MRCA in units of 2N generations:\n",
    "# expected_tmrca = 2*(1-(1/samples))\n",
    "# print(\"expected Tmrca:\", expected_tmrca)\n",
    "# print(\"mean tmrca, in generations/(ploidy*N):\", mean_tmrca/sim_runs/population_sequences, \"mean tmrca, in raw gens:\", mean_tmrca/sim_runs) #the first value is generations/(ploidy*N),and the second is in generations.\n",
    "#expected time that i lineages should exist: T(i) =  4N/i(i-1) (hudson 2015, #2)\n",
    "#sum i*(T(i)) from i=2 to n to get the total expected branch lengths: (above equation 6, hudson 1990)\n",
    "expected_total_length = 0\n",
    "for i in range(2, sampled_sequences):\n",
    "    expected_total_length += i*((4*(population_sequences/ploidy)/(i*(i-1))))\n",
    "\n",
    "# print(\"total expected branch lengths\", expected_total_length, \"mean summed branch lengths:\", (sum(mean_summed_branch_lengths)/sim_runs))\n",
    "#now, we can use this to find the expected total number of polymorphisms (hudson 1990 eq1)\n",
    "expected_lambda = mu*expected_total_length\n",
    "# print(\"mean expectation for the number of polymorphisms on the tree\", expected_lambda)\n",
    "\n",
    "\n",
    "#sfs stats\n",
    "theta = 4*(population_sequences/ploidy)*mu\n",
    "# print(\"mutation counts from all loops\", i_values)\n",
    "mean_counts = {}\n",
    "for i in range(1, ((sampled_sequences*2)-1)):\n",
    "    mean = (i_values.count(i))/sim_runs\n",
    "    mean_counts.update({i:mean})\n",
    "# mean_allele_freqs = (i_values.count(i))/sim_runs\n",
    "# print(\"mean allele frequencies, from all sims binned\", mean_counts)\n",
    "# # print(\"average number of mutations in sims\", sum(mutations_in_run)/sim_runs, \"expected mean number of mutations\", expected_mutations)\n",
    "\n",
    "print(\"SAMPLE PRINT STATEMENTS: only apply to last loop over the treesequences\")\n",
    "print(\"list of tree sizes, in units of JUST generations\", genealogy_sizes, \"list of sizes, weighted by how much of the genome they occupy:\", weighted_genealogy_sizes, \"weighting values\", sequence_weights)\n",
    "print(\"value of T, weighted average genealogy length\", T_value)\n",
    "print(\"mean mutation expectation, over genealogies in this ts:\", expected_mean_number_mutations)\n",
    "afs = new_ts.allele_frequency_spectrum(polarised=True, span_normalise=False)\n",
    "afs_branches = new_ts.allele_frequency_spectrum(mode=\"branch\", polarised=True, span_normalise=False)\n",
    "print(\"allele frequency spectrum\", afs, \"total branch lengths above each bin\", afs_branches)\n",
    "# print(afs/new_ts.total_branch_length)\n",
    "# print(\"mean expected lambda output\", mean_expected_lambda)\n",
    "\n",
    "print(\"STATEMENTS THAT APPLY TO ALL TS\")\n",
    "print(\"mean expectation for the number of polymorphisms generated by each ts\", sum(mean_expected_lambda)/sim_runs, \"number of average polymorphisms actually on the trees\", (sum(total_mut_count))/sim_runs)\n",
    "print(\"expected mean site frequency spectrum\", [(theta/i) for i in range(1, sampled_sequences)], \"mean allele frequencies from sims, averaged\", mean_counts)\n",
    "ind = np.arange(len(mean_counts))\n",
    "plt.bar(ind, list(mean_counts.values()))\n",
    "for t in new_ts.trees():\n",
    "  print(t.draw(format='unicode'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.7912818296162087 1791.2818296162088\n",
      "1.822813472845434 1822.813472845434\n",
      "1.8013986092376217 1801.3986092376217\n",
      "1.797466850685748 1797.466850685748\n",
      "4888.190532188829\n"
     ]
    }
   ],
   "source": [
    "#kevin's script for how to deal with units of generations in msprime\n",
    "import msprime\n",
    "\n",
    "nreps = 10000\n",
    "\n",
    "for ploidy in [1, 2, 5, 10]:\n",
    "    mean = 0\n",
    "    # Here is where it (the interface) gets confusing.\n",
    "    # We simulate 10 lineages.\n",
    "    # But there's a GIANT gotcha:\n",
    "    # population_size is number of INDIVIDUALS and not\n",
    "    # the number of LINEAGES.\n",
    "    # Therefore, we have to divied that by PLOIDY\n",
    "    # to keep everything comparable.\n",
    "    for ts in msprime.sim_ancestry(samples =(10/ploidy), population_size=1000/ploidy,\n",
    "                                   ploidy=ploidy, model=\"hudson\",\n",
    "                                   num_replicates=nreps):\n",
    "        assert ts.num_samples == 10 ##I forget this every time, but I think the point is ts.num_samples will return the number of sampled SEQUENCES, while \n",
    "        last_node = ts.num_nodes - 1\n",
    "        tmrca = ts.node(last_node).time\n",
    "        mean += tmrca\n",
    "\n",
    "    # Now, the first value is generations/(ploidy*N),\n",
    "    # And the second is in generations.\n",
    "    print(mean/nreps/1000, mean/nreps)\n",
    "\n",
    "# Divide out the ploidy\n",
    "ts = msprime.sim_ancestry(samples = 5, population_size=1000/2,\n",
    "                          ploidy=ploidy, model=\"hudson\")\n",
    "last_node = ts.num_nodes - 1\n",
    "tmrca = ts.node(last_node).time\n",
    "\n",
    "# needs tskit 0.5.4!!!!!!!!!!!!!!!!!!!!!!!!!\n",
    "ts.draw_svg(path=\"test.svg\", y_axis=True, size=(\n",
    "    600, 600), canvas_size=(800, 800))\n",
    "print(tmrca) #in units of generations- we don't need to divide by the reps above because it's only one rep! Also shows how wide the variance is on that mean- not crazy to get 11,800 gens for the time when the expectation is 1,800 gens, or 1.8 in units gens/ploidy*N \n",
    "\n",
    "#these should all come up with the theoretical expectation: for ploidy and n\n",
    "# print(2*(1-(1/10)))\n",
    "# print(2*(1-1/10)) #me, seeing how 'smart' python is with orders of operation\n",
    "# print(2*(1-1/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nodes in the current tree [0. 1. 2. 3. 4. 5. 6. 0.]\n",
      "mean number of mutations expected on this tree/poisson lambda 0.4709072046998257 number of mutations drawn from poisson distribution 0\n",
      "fraction of the genome this tree occupies 0.545\n",
      "summed branch lengths: 8.64049916880414\n",
      "interval this tree occupies Interval(left=0.0, right=545.0)\n",
      "LOOP BREAK between trees \n",
      "\n",
      "Expected site frequency spectrum [40.0, 20.0, 13.333333333333334, 10.0, 8.0, 6.666666666666667, 5.714285714285714, 5.0, 4.444444444444445]\n",
      "nodes in the current tree [0. 1. 2. 3. 4. 5. 0. 7.]\n",
      "mean number of mutations expected on this tree/poisson lambda 0.9707031794118754 number of mutations drawn from poisson distribution 2\n",
      "fraction of the genome this tree occupies 0.455\n",
      "summed branch lengths: 21.334135811250007\n",
      "interval this tree occupies Interval(left=545.0, right=1000.0)\n",
      "893\n",
      "912\n",
      "LOOP BREAK between trees \n",
      "\n",
      "Expected site frequency spectrum [40.0, 20.0, 13.333333333333334, 10.0, 8.0, 6.666666666666667, 5.714285714285714, 5.0, 4.444444444444445]\n",
      "loop break between trees\n",
      "TableCollection\n",
      "\n",
      "Sequence Length: 1000.0\n",
      "Time units: generations\n",
      "Metadata: b''\n",
      "\n",
      "Individuals\n",
      "╔══╤═════╤════════╤═══════╤════════╗\n",
      "║id│flags│location│parents│metadata║\n",
      "╠══╪═════╪════════╪═══════╪════════╣\n",
      "║0 │    0│        │       │        ║\n",
      "║1 │    0│        │       │        ║\n",
      "╚══╧═════╧════════╧═══════╧════════╝\n",
      "\n",
      "Nodes\n",
      "╔══╤═════╤══════════╤══════════╤═══════════╤════════╗\n",
      "║id│flags│population│individual│time       │metadata║\n",
      "╠══╪═════╪══════════╪══════════╪═══════════╪════════╣\n",
      "║0 │    1│         0│         0│ 0.00000000│        ║\n",
      "║1 │    1│         0│         0│ 0.00000000│        ║\n",
      "║2 │    1│         0│         1│ 0.00000000│        ║\n",
      "║3 │    1│         0│         1│ 0.00000000│        ║\n",
      "║4 │    0│         0│        -1│ 0.22949513│        ║\n",
      "║5 │    0│         0│        -1│ 0.86085404│        ║\n",
      "║6 │    0│         0│        -1│ 4.86542417│        ║\n",
      "║7 │    0│         0│        -1│11.21224249│        ║\n",
      "╚══╧═════╧══════════╧══════════╧═══════════╧════════╝\n",
      "\n",
      "Edges\n",
      "╔══╤════╤═════╤══════╤═════╤════════╗\n",
      "║id│left│right│parent│child│metadata║\n",
      "╠══╪════╪═════╪══════╪═════╪════════╣\n",
      "║0 │   0│ 1000│     4│    0│        ║\n",
      "║1 │   0│ 1000│     4│    1│        ║\n",
      "║2 │   0│ 1000│     5│    2│        ║\n",
      "║3 │   0│ 1000│     5│    3│        ║\n",
      "║4 │   0│  545│     6│    4│        ║\n",
      "║5 │   0│  545│     6│    5│        ║\n",
      "║6 │ 545│ 1000│     7│    4│        ║\n",
      "║7 │ 545│ 1000│     7│    5│        ║\n",
      "╚══╧════╧═════╧══════╧═════╧════════╝\n",
      "\n",
      "Sites\n",
      "╔══╤════════╤═══════════════╤════════╗\n",
      "║id│position│ancestral_state│metadata║\n",
      "╠══╪════════╪═══════════════╪════════╣\n",
      "║0 │     893│              0│        ║\n",
      "║1 │     912│              0│        ║\n",
      "╚══╧════════╧═══════════════╧════════╝\n",
      "\n",
      "Mutations\n",
      "╔══╤════╤════╤═══════════╤═════════════╤══════╤════════╗\n",
      "║id│site│node│time       │derived_state│parent│metadata║\n",
      "╠══╪════╪════╪═══════════╪═════════════╪══════╪════════╣\n",
      "║0 │ 893│   5│11.21224249│            1│    -1│        ║\n",
      "║1 │ 912│   5│ 0.86085404│            1│    -1│        ║\n",
      "╚══╧════╧════╧═══════════╧═════════════╧══════╧════════╝\n",
      "\n",
      "Migrations\n",
      "╔══╤════╤═════╤════╤══════╤════╤════╤════════╗\n",
      "║id│left│right│node│source│dest│time│metadata║\n",
      "╠══╪════╪═════╪════╪══════╪════╪════╪════════╣\n",
      "╚══╧════╧═════╧════╧══════╧════╧════╧════════╝\n",
      "\n",
      "Populations\n",
      "╔══╤════════════════════════════════════╗\n",
      "║id│metadata                            ║\n",
      "╠══╪════════════════════════════════════╣\n",
      "║0 │{'description': '', 'name': 'pop_0'}║\n",
      "╚══╧════════════════════════════════════╝\n",
      "\n",
      "Provenances\n",
      "╔══╤══════════════════════════╤════════════════════════════════════════════════════════════╗\n",
      "║id│timestamp                 │record                                                      ║\n",
      "╠══╪══════════════════════════╪════════════════════════════════════════════════════════════╣\n",
      "║0 │2023-02-27T16:52:50.550967│{\"schema_version\": \"1.0.0\", \"software\": {\"name\": \"msprime...║\n",
      "╚══╧══════════════════════════╧════════════════════════════════════════════════════════════╝\n",
      "\n",
      "   6   \n",
      " ┏━┻━┓ \n",
      " ┃   5 \n",
      " ┃  ┏┻┓\n",
      " 4  ┃ ┃\n",
      "┏┻┓ ┃ ┃\n",
      "0 1 2 3\n",
      "\n",
      "   7   \n",
      " ┏━┻━┓ \n",
      " ┃   5 \n",
      " ┃  ┏┻┓\n",
      " 4  ┃ ┃\n",
      "┏┻┓ ┃ ┃\n",
      "0 1 2 3\n",
      "\n"
     ]
    },
    {
     "ename": "LibraryError",
     "evalue": "Site out of bounds. (TSK_ERR_SITE_OUT_OF_BOUNDS)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLibraryError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 136\u001b[0m\n\u001b[1;32m    132\u001b[0m \u001b[39mfor\u001b[39;00m t \u001b[39min\u001b[39;00m ts\u001b[39m.\u001b[39mtrees():\n\u001b[1;32m    133\u001b[0m   \u001b[39mprint\u001b[39m(t\u001b[39m.\u001b[39mdraw(\u001b[39mformat\u001b[39m\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39municode\u001b[39m\u001b[39m'\u001b[39m))\n\u001b[0;32m--> 136\u001b[0m newtables\u001b[39m.\u001b[39;49msort()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/tskit/tables.py:3524\u001b[0m, in \u001b[0;36mTableCollection.sort\u001b[0;34m(self, edge_start, site_start, mutation_start)\u001b[0m\n\u001b[1;32m   3468\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39msort\u001b[39m(\u001b[39mself\u001b[39m, edge_start\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m, \u001b[39m*\u001b[39m, site_start\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m, mutation_start\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m):\n\u001b[1;32m   3469\u001b[0m     \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m   3470\u001b[0m \u001b[39m    Sorts the tables in place. This ensures that all tree sequence ordering\u001b[39;00m\n\u001b[1;32m   3471\u001b[0m \u001b[39m    requirements listed in the\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3522\u001b[0m \u001b[39m        (default=0; must be one of [0, len(mutations)]).\u001b[39;00m\n\u001b[1;32m   3523\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 3524\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_ll_tables\u001b[39m.\u001b[39;49msort(edge_start, site_start, mutation_start)\n",
      "\u001b[0;31mLibraryError\u001b[0m: Site out of bounds. (TSK_ERR_SITE_OUT_OF_BOUNDS)"
     ]
    }
   ],
   "source": [
    "#make a standalone tree sequence to learn how to do the mutation placement, make sim predictions, etc\n",
    "import msprime\n",
    "import tskit\n",
    "import numpy as np\n",
    "from numpy.random import default_rng\n",
    "rng = default_rng()\n",
    "import random\n",
    "#simulate one population fo now\n",
    "mu = (1e-1) #normally would have this set to 1e-7 *2, but putting deme size down and #individuals up to be able to visualize tree sequences for working on placing mutations issue.\n",
    "rho = (1e-4)*2\n",
    "bases = 1000\n",
    "deme_size = 100\n",
    "ts = msprime.sim_ancestry(recombination_rate=rho, sequence_length = bases, samples=2, random_seed=1234, discrete_genome = True)\n",
    "newtables = ts.dump_tables() #modifying a copy of the original tables in order to be able to overwrite \n",
    "#make a new table collection to hold everything\n",
    "for tree in ts.trees():\n",
    "    nodes_list = np.zeros(tree.tree_sequence.num_nodes) #make a np array for each node to track the branch lengths above it (needed for calculations around placing mutations)\n",
    "    branch_lengths = np.zeros(tree.tree_sequence.num_nodes) #need two np arrays initiated with the same number- one for branch lengths, and one to keep track of the nodes\n",
    "    summed_branch_lengths = 0\n",
    "    interval = tree.interval #find the interval the tree occupies in the genome \n",
    "    span = tree.interval.right-tree.interval.left #find the fraction of the total genome/number of bases\n",
    "    \n",
    "    for node in tree.nodes(): #type error: tree object is not iterable-- this happens w/ you frequently b/c you're forgetting to include the parenthesis to call the method properly, so the code tries to iterate over the method and not the data structure it's referencing \n",
    "        branch_lengths[node] =+ tree.branch_length(node)  #https://stackoverflow.com/questions/43505921/python-trying-to-create-a-dictionary-through-a-for-loop\n",
    "        nodes_list[node] =+ node\n",
    "        child_time = newtables.nodes.time[node]\n",
    "\n",
    "        for index, child_node in enumerate(newtables.edges.child):\n",
    "            if child_node == node:\n",
    "                # print(\"index of edgetable entry\", index, \"child node\", child_node)\n",
    "                # print(\"index of edgetable entry\", index, \"child node\", child_node)\n",
    "                entry = newtables.edges[index]\n",
    "                # print(\"entry that matches a node in this tree\", entry)\n",
    "                # print(\"line of edgetable accessed\", index, \"child node for this loop\", child_node, \"edgetable entry\", entry)\n",
    "                current_tree_interval = (interval.left, interval.right)\n",
    "                if (entry.left, entry.right) == current_tree_interval: \n",
    "                    parent_node = entry.parent\n",
    "                    parent_time = newtables.nodes.time[parent_node]\n",
    "                    branch_length = (parent_time - child_time)    \n",
    "                    summed_branch_lengths = summed_branch_lengths + branch_length #hudson 1990, above equation 1\n",
    "                    ####place mutations on nodes proportional to branch lengths, and proportional to how long of a span of the genome a given tree occupies\n",
    "                    lambda_lineage = mu *branch_length *(span/bases) #my idea to populate the sim with the expected number of mutations given the parameters I'm using: get mu * t in here, after the hudson 2015 paper having lineage mu * branch length be the number of mutations occuring along that lineage. Summing them all up to get expectations for each tree. I'm multiplying by the fraction of the genome, since I believe mu* branch length as the expectation for the # of mutations assumes the whole genome\n",
    "\n",
    "    print(\"nodes in the current tree\", nodes_list)\n",
    "    lambda_total = mu*summed_branch_lengths*(span/bases)\n",
    "    genealogy_mutations = rng.poisson(lam=lambda_total)\n",
    "    print(\"mean number of mutations expected on this tree/poisson lambda\", lambda_total, \"number of mutations drawn from poisson distribution\", genealogy_mutations)\n",
    "    print(\"fraction of the genome this tree occupies\", span/bases)\n",
    "    print(\"summed branch lengths:\", summed_branch_lengths)\n",
    "    proportional_branch_lengths = branch_lengths/summed_branch_lengths \n",
    "    p = np.asarray(proportional_branch_lengths).astype('float64')\n",
    "    p = p/np.sum(p) #not sure if doing this as a workaround to the proportional branch lengths array not summing to 0 is the best idea, but to get it working for now\n",
    "    # print(\"probabilities\",p, \"sums\", sum(p))\n",
    "\n",
    "    #######temporarily hardcoding in some mutations to test the below, changed size = int(number_mutations) to size = 3\n",
    "    nodes_sampled =  rng.choice(nodes_list, p=p, size = genealogy_mutations) #this can sample the same node twice, which IS what you want. however need to figure out a way to determine which parent node was associated with a given branch length, so you can select a site that makes sense given the edges possible for a child:parent node relationship. int is a pretty crude \n",
    "    # nodes_sampled =  np.random.choice(nodes_list, p=p, size = np.random.choice(expected mutations))\n",
    "\n",
    "    nodes_sampled = nodes_sampled.astype(int) ##these need to be integers to be able to use them to access indices/values in the edge table. See if you can make it such when the  variable is created \n",
    "    # print(\"nodes sampled from distribution\", nodes_sampled)\n",
    "    #pick out the spot in the genome that the tree we have iterated over actually occupies (with edges) \n",
    "    print(\"interval this tree occupies\", tree.interval) #access this to pick which sites are possible\n",
    "    # print(\"left side of interval\", interval.left)\n",
    "    for node in nodes_sampled: #since I do a pretty much identical thing above, there's gotta be a way to combine the two, but I'll leave the stunts for later \n",
    "        ######draw a site from the genome that HASN'T ALREADY had a mutation assigned to it. \n",
    "        #use these nodes to write mutations to the mutations table\n",
    "        possible_bases = set(range((int(interval.left)), int(interval.right))) #set them to integers for now? need to figure out how to actually implement using sets \n",
    "        bases_already_mutated = set(int(i) for i in newtables.sites.position)\n",
    "        unmutated_sites =(possible_bases.difference(bases_already_mutated)) #get the difference between two sets- it returns only those bases NOT in bases_already_mutated\n",
    "        # print(\"checking if unmutated sites are constrained to those within the tree\", unmutated_sites)\n",
    "        # print(\"set difference test\", x)\n",
    "        # position = np.random.choice(tuple(unmutated_sites))#suggestion from one stackoverflow entry for this(https://stackoverflow.com/questions/15837729/random-choice-from-set-python) links to another to try to make this process constant time???: https://stackoverflow.com/questions/15993447/python-data-structure-for-efficient-add-remove-and-random-choice\n",
    "        # print(position)\n",
    "        ######add binary mutations in- 0/1, whether it's been mutated or not. Or could use just two nucleotides, for instance\n",
    "        # print(\"SITES THAT DO NOT YET HAVE A MUTATION\", unmutated_sites)\n",
    "        ###suggestion for how to make it faster\n",
    "        position = np.random.randint(low=(interval.left), high=(int(interval.right)), size=1)\n",
    "        position = position[0]\n",
    "        print(position)\n",
    "        if position in unmutated_sites: #size should be 1, since we're only placing one mutation above each loop over a node\n",
    "            newtables.sites.add_row(position=position, ancestral_state=\"0\")\n",
    "\n",
    "        #####get times for the mutations- randomly select them from the interval between parent and child node times \n",
    "        #get child node time\n",
    "        child_time = newtables.nodes.time[node]\n",
    "        # print(\"node ID\", node,  ts.tables.edges.child[node])#https://stackoverflow.com/questions/176918/finding-the-index-of-an-item-in-a-\n",
    "        \n",
    "        for index, child_node in enumerate(newtables.edges.child): #I think the child node is the appropriate one to pick, since we chose nodes BELOW branches. Mutations are associated with the node below them (according to tskit data model page)\n",
    "            # print(index, j) #this print statement coupled with the one below lets you check if it's behaving as intended\n",
    "            if child_node == node:\n",
    "                entry = newtables.edges[index]\n",
    "                # print(\"node\", node, \"edges associated with the node selected\", entry)\n",
    "                current_tree_interval = (interval.left, interval.right)\n",
    "                if (entry.left, entry.right) == current_tree_interval: \n",
    "                    parent_node = entry.parent\n",
    "                    # print(\"tuple matched, parent node\", parent_node)\n",
    "                elif (entry.left, entry.right) == (0, bases): #this happens when the node is not associated with a recombination event \n",
    "                    # print(\"this node has the same parent regardless of tree\") (#do I need to do this above??)\n",
    "                    parent_node=entry.parent\n",
    "\n",
    "        #now, find actual parent time:\n",
    "        parent_time = newtables.nodes.time[parent_node]\n",
    "        # print(\"parent time\", parent_time)\n",
    "        time_span = (child_time, parent_time)\n",
    "        # print(\"time interval between a node and its parent\", time_span)\n",
    "        time=random.choice(time_span) #you get a suspiciously high amount of '0' times. \n",
    "\n",
    "        #####assign all variables to the mutations table \n",
    "        newtables.mutations.add_row(site=position, node=node, time=time, derived_state = \"1\") #this doesn't work as with the above-- perhaps something with the strings. \n",
    "        # print(tskit.unpack_strings(newtables.sites.ancestral_state, newtables.sites.ancestral_state_offset))\n",
    "\n",
    "        # print(np.setdiff1d(nucleotides, newtables.sites.ancestral_state))\n",
    "    print(\"LOOP BREAK between trees\", \"\\n\")\n",
    "\n",
    "\n",
    "    expected_sfs = []\n",
    "    theta = 4*(deme_size)*(mu)\n",
    "    for i in range(1,10):\n",
    "        expected_sfs.append(theta/i)\n",
    "    print(\"Expected site frequency spectrum\", expected_sfs)\n",
    "\n",
    "# mutations expected for whole genome: \n",
    "\n",
    "\n",
    "\n",
    "print(\"loop break between trees\")\n",
    "print(newtables)\n",
    "# simulated_sfs = []\n",
    "# a = ts.allele_frequency_spectrum(sample_sets=[ts.samples(population=1), ts.samples(population=2)], windows=None, mode='site', span_normalise=True, polarised=False)\n",
    "# print(\"allele frequency spectrum function\", a)\n",
    "\n",
    "for t in ts.trees():\n",
    "  print(t.draw(format='unicode'))\n",
    "\n",
    "\n",
    "newtables.sort()\n",
    "\n",
    "# new_ts = newtables.tree_sequence()\n",
    "# for t in new_ts.trees():\n",
    "#   print(t.draw(format='unicode'))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
